<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Differential abundance testing with Milo - Mouse gastrulation example • miloR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Differential abundance testing with Milo - Mouse gastrulation example">
<meta property="og:description" content="miloR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">miloR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/milo_demo.html">Differential abundance testing with Milo</a>
    </li>
    <li>
      <a href="../articles/milo_gastrulation.html">Differential abundance testing with Milo - Mouse gastrulation example</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="milo_gastrulation_files/header-attrs-2.3/header-attrs.js"></script><script src="milo_gastrulation_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Differential abundance testing with Milo - Mouse gastrulation example</h1>
                        <h4 class="author">Emma Dann</h4>
                        <h4 class="author">Mike Morgan</h4>
            
      
      
      <div class="hidden name"><code>milo_gastrulation.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">miloR</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">SingleCellExperiment</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">scater</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">patchwork</span>)</pre></body></html></div>
<div id="load-data" class="section level1">
<h1 class="hasAnchor">
<a href="#load-data" class="anchor"></a>Load data</h1>
<p>For this vignette we will use the mouse gastrulation single-cell data from <a href="https://www.nature.com/articles/s41586-019-0933-9">Pijuan-Sala et al. 2019</a>. The dataset can be downloaded as a <code>SingleCellExperiment</code> object from the <a href="https://bioconductor.org/packages/3.12/data/experiment/html/MouseGastrulationData.html"><code>MouseGastrulationData</code></a> package on Bioconductor. To make computations faster, here we will download just a subset of samples, 4 samples at stage E7 and 4 samples at stage E7.5.</p>
<p>This dataset has already been pre-processed and contains a <code>pca.corrected</code> dimensionality reduction, which was built after batch correction using <a href="https://bioconductor.org/packages/release/bioc/vignettes/batchelor/inst/doc/correction.html"><code>fastMNN</code></a>.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">MouseGastrulationData</span>)
<span class="no">select_samples</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>,  <span class="fl">3</span>,  <span class="fl">6</span>, <span class="fl">15</span>,
                    <span class="co"># 4, 19, </span>
                    <span class="fl">10</span>, <span class="fl">14</span>, <span class="fl">20</span>, <span class="fl">30</span>
                    <span class="co">#31, 32</span>
                    )
<span class="no">embryo_data</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MouseGastrulationData/man/EmbryoAtlasData.html">EmbryoAtlasData</a></span>(<span class="kw">samples</span> <span class="kw">=</span> <span class="no">select_samples</span>)</pre></body></html></div>
<pre><code>## Warning: `select_()` is deprecated as of dplyr 0.7.0.
## Please use `select()` instead.
## This warning is displayed once every 8 hours.
## Call `lifecycle::last_warnings()` to see where this warning was generated.</code></pre>
<pre><code>## Warning: `filter_()` is deprecated as of dplyr 0.7.0.
## Please use `filter()` instead.
## See vignette('programming') for more help
## This warning is displayed once every 8 hours.
## Call `lifecycle::last_warnings()` to see where this warning was generated.</code></pre>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">embryo_data</span></pre></body></html></div>
<pre><code>## class: SingleCellExperiment 
## dim: 29452 14679 
## metadata(0):
## assays(1): counts
## rownames(29452): ENSMUSG00000051951 ENSMUSG00000089699 ...
##   ENSMUSG00000096730 ENSMUSG00000095742
## rowData names(2): ENSEMBL SYMBOL
## colnames(14679): cell_361 cell_362 ... cell_107079 cell_107080
## colData names(16): cell barcode ... celltype colour
## reducedDimNames(2): pca.corrected umap
## spikeNames(0):
## altExpNames(0):</code></pre>
</div>
<div id="visualize-the-data" class="section level1">
<h1 class="hasAnchor">
<a href="#visualize-the-data" class="anchor"></a>Visualize the data</h1>
<p>We recompute PCA and UMAP embedding for this subset of cells to visualize the data.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">embryo_data</span> <span class="kw">&lt;-</span> <span class="no">embryo_data</span>[,<span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span>(<span class="no">embryo_data</span>, <span class="st">"pca.corrected"</span>), <span class="fl">1</span>, <span class="kw">function</span>(<span class="no">x</span>) !<span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">x</span>)))]
<span class="no">embryo_data</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span>(<span class="no">embryo_data</span>, <span class="kw">dimred</span> <span class="kw">=</span> <span class="st">"pca.corrected"</span>, <span class="kw">name</span> <span class="kw">=</span> <span class="st">'umap'</span>)

<span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span>(<span class="no">embryo_data</span>, <span class="kw">colour_by</span><span class="kw">=</span><span class="st">"stage"</span>, <span class="kw">dimred</span> <span class="kw">=</span> <span class="st">"umap"</span>)</pre></body></html></div>
<p><img src="milo_gastrulation_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>We will test for significant differences in abundance of cells between these stages of development, and the associated gene signatures.</p>
</div>
<div id="differential-abundance-testing" class="section level1">
<h1 class="hasAnchor">
<a href="#differential-abundance-testing" class="anchor"></a>Differential abundance testing</h1>
<div id="create-a-milo-object" class="section level2">
<h2 class="hasAnchor">
<a href="#create-a-milo-object" class="anchor"></a>Create a Milo object</h2>
<p>For differential abundance analysis on graph neighbourhoods we first construct a <code>Milo</code> object. This extends the <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html"><code>SingleCellExperiment</code></a> class to store information about neighbourhoods on the KNN graph.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">embryo_milo</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/Milo.html">Milo</a></span>(<span class="no">embryo_data</span>)
<span class="no">embryo_milo</span></pre></body></html></div>
<pre><code>## class: Milo 
## dim: 29452 13002 
## metadata(0):
## assays(1): counts
## rownames(29452): ENSMUSG00000051951 ENSMUSG00000089699 ...
##   ENSMUSG00000096730 ENSMUSG00000095742
## rowData names(2): ENSEMBL SYMBOL
## colnames(13002): cell_361 cell_362 ... cell_107079 cell_107080
## colData names(16): cell barcode ... celltype colour
## reducedDimNames(2): pca.corrected umap
## spikeNames(0):
## altExpNames(0):
## nhoods dimensions(1): 0
## nhoodCounts dimensions(2): 1 1
## nhoodDistances dimension(1): 0
## graph names(0):
## nhoodIndex names(1): 0
## nhoodExpression dimension(2): 1 1
## nhoodReducedDim names(0):
## nhoodGraph names(0):
## nhoodAdjacency dimension(0):</code></pre>
</div>
<div id="construct-knn-graph" class="section level2">
<h2 class="hasAnchor">
<a href="#construct-knn-graph" class="anchor"></a>Construct KNN graph</h2>
<p>We need to add the KNN graph to the Milo object. This is stored in the <code>graph</code> slot, in <a href="https://igraph.org/r/"><code>igraph</code></a> format. The <code>miloR</code> package includes functionality to build and store the graph from the PCA dimensions stored in the <code>reducedDim</code> slot. In this case, we specify that we want to build the graph from the MNN corrected PCA dimensions.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">embryo_milo</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/buildGraph.html">buildGraph</a></span>(<span class="no">embryo_milo</span>, <span class="kw">k</span> <span class="kw">=</span> <span class="fl">30</span>, <span class="kw">d</span> <span class="kw">=</span> <span class="fl">30</span>, <span class="kw">reduced.dim</span> <span class="kw">=</span> <span class="st">"pca.corrected"</span>)</pre></body></html></div>
<p>Alternatively, one can add a precomputed KNN graph (for example constructed with Seurat or scanpy) to the <code>graph</code> slot using the adjacency matrix, through the helper function <code>buildFromAdjacency</code>.</p>
<!-- Alternatively, if you already have a KNN graph (for example constructed with Seurat/scanpy) you can add it from the adjacency matrix. -->
<!-- ```{r} -->
<!-- # ## Build up a mock SNN graph made with Seurat -->
<!-- # pca_df <- reducedDim(traj_milo, "PCA") -->
<!-- # rownames(pca_df) <- traj_milo$cell_id -->
<!-- # snn_graph <- FindNeighbors(pca_df)[["snn"]] -->
<!-- #  -->
<!-- # graph(traj_milo) <-  graph(buildFromAdjacency(snn_graph, k=10)) -->
<!-- ``` -->
</div>
<div id="defining-representative-neighbourhoods-on-the-knn-graph" class="section level2">
<h2 class="hasAnchor">
<a href="#defining-representative-neighbourhoods-on-the-knn-graph" class="anchor"></a>Defining representative neighbourhoods on the KNN graph</h2>
<p>We define the neighbourhood of a cell, the index, as the group of cells connected by an edge in the KNN graph to the index cell. For efficiency, we don’t test for DA in the neighbourhood of every cell, but we sample as indices a subset of representative cells, using a KNN sampling algorithm used by <a href="https://www.nature.com/articles/nmeth.3545">Gut et al. 2015</a>.</p>
<p>For sampling you need to define a few parameters:</p>
<ul>
<li>
<code>prop</code>: the proportion of cells to randomly sample to start with (usually 0.1 - 0.2 is sufficient)</li>
<li>
<code>k</code>: the k to use for KNN refinement (we recommend using the same k used for KNN graph building)</li>
<li>
<code>d</code>: the number of reduced dimensions to use for KNN refinement (we recommend using the same d used for KNN graph building)</li>
<li>
<code>refined</code>: indicates whether you want to use the sampling refinement algorith, or just pick cells at random. The default and recommended way to go is to use refinement. The only situation in which you might consider using <code>random</code> instead, is if you have batch corrected your data with a graph based correction algorithm, such as <a href="https://github.com/Teichlab/bbknn">BBKNN</a>, but the results of DA testing will be suboptimal.</li>
</ul>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">embryo_milo</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/makeNhoods.html">makeNhoods</a></span>(<span class="no">embryo_milo</span>, <span class="kw">prop</span> <span class="kw">=</span> <span class="fl">0.1</span>, <span class="kw">k</span> <span class="kw">=</span> <span class="fl">30</span>, <span class="kw">d</span><span class="kw">=</span><span class="fl">30</span>, <span class="kw">refined</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">reduced_dims</span> <span class="kw">=</span> <span class="st">"pca.corrected"</span>)</pre></body></html></div>
<p>Once we have defined neighbourhoods, it’s good to take a look at how big the neighbourhoods are (i.e. how many cells form each neighbourhood). This affects the power of DA testing. We can check this out using the <code>plotNhoodSizeHist</code> function. Empirically, we found it’s best to have a distribution peaking above 20. Otherwise you might consider rerunning <code>makeNhoods</code> increasing <code>k</code> and/or <code>prop</code>.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="fu"><a href="../reference/plotNhoodSizeHist.html">plotNhoodSizeHist</a></span>(<span class="no">embryo_milo</span>)</pre></body></html></div>
<p><img src="milo_gastrulation_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
<div id="counting-cells-in-neighbourhoods" class="section level2">
<h2 class="hasAnchor">
<a href="#counting-cells-in-neighbourhoods" class="anchor"></a>Counting cells in neighbourhoods</h2>
<p><em>Milo</em> leverages the variation in cell numbers between replicates for the same experimental condition to test for differential abundance. Therefore we have to count how many cells from each sample are in each neighbourhood. We need to use the cell metadata and specify which column contains the sample information.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">embryo_milo</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/countCells.html">countCells</a></span>(<span class="no">embryo_milo</span>, <span class="kw">meta.data</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="fu">colData</span>(<span class="no">embryo_milo</span>)), <span class="kw">sample</span><span class="kw">=</span><span class="st">"sample"</span>)</pre></body></html></div>
<p>This adds to the <code>Milo</code> object a <span class="math inline">\(n \times m\)</span> matrix, where <span class="math inline">\(n\)</span> is the number of neighbourhoods and <span class="math inline">\(m\)</span> is the number of experimental samples. Values indicate the number of cells from each sample counted in a neighbourhood. This count matrix will be used for DA testing.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/methods.html">nhoodCounts</a></span>(<span class="no">embryo_milo</span>))</pre></body></html></div>
<pre><code>## 6 x 8 sparse Matrix of class "dgCMatrix"
##   2 3  6 15 10 14 20 30
## 1 . .  2  8 50  1  2 80
## 2 . .  4  6 28  7  . 45
## 3 . 1  6 10 19  8  2 39
## 4 4 7 20  .  .  . 17  .
## 5 . 1  7  4 27  2  8 42
## 6 2 3 40  4  8  2 26 12</code></pre>
</div>
<div id="defining-experimental-design" class="section level2">
<h2 class="hasAnchor">
<a href="#defining-experimental-design" class="anchor"></a>Defining experimental design</h2>
<p>Now we are all set to test for differential abundance in neighbourhoods. We implement this hypothesis testing in a generalized linear model (GLM) framework, specifically using the Negative Binomial GLM implementation in <a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html"><code>edgeR</code></a>.</p>
<p>We first need to think about our experimental design. The design matrix should match each sample to the experimental condition of interest for DA testing. In this case, we want to detect DA between embryonic stages, stored in the <code>stage</code> column of the dataset <code>colData</code>. We also include the <code>sequencing.batch</code> column in the design matrix. This represents a known technical covariate that we want to account for in DA testing.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">embryo_design</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="fu">colData</span>(<span class="no">embryo_milo</span>))[,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"sample"</span>, <span class="st">"stage"</span>, <span class="st">"sequencing.batch"</span>)]
<span class="co">## Convert batch info from integer to factor</span>
<span class="no">embryo_design</span>$<span class="no">sequencing.batch</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>(<span class="no">embryo_design</span>$<span class="no">sequencing.batch</span>)
<span class="no">embryo_design</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html">distinct</a></span>(<span class="no">embryo_design</span>)
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">embryo_design</span>) <span class="kw">&lt;-</span> <span class="no">embryo_design</span>$<span class="no">sample</span>

<span class="no">embryo_design</span></pre></body></html></div>
<pre><code>##    sample stage sequencing.batch
## 2       2  E7.5                1
## 3       3  E7.5                1
## 6       6  E7.5                1
## 15     15  E7.0                2
## 10     10  E7.0                1
## 14     14  E7.0                2
## 20     20  E7.5                2
## 30     30  E7.0                3</code></pre>
</div>
<div id="computing-neighbourhood-connectivity" class="section level2">
<h2 class="hasAnchor">
<a href="#computing-neighbourhood-connectivity" class="anchor"></a>Computing neighbourhood connectivity</h2>
<p>Milo uses an adaptation of the Spatial FDR correction introduced by <a href="https://bioconductor.org/packages/release/bioc/html/cydar.html">cydar</a>, which accounts for the overlap between neighbourhoods. Specifically, each hypothesis test P-value is weighted by the reciprocal of the kth nearest neighbour distance. To use this statistic we first need to store the distances between nearest neighbors in the Milo object. This is done by the <code>calcNhoodDistance</code> function (N.B. this step is the most time consuming of the analysis workflow and might take a couple of minutes for large datasets).</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="no">embryo_milo</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/calcNhoodDistance.html">calcNhoodDistance</a></span>(<span class="no">embryo_milo</span>, <span class="kw">d</span><span class="kw">=</span><span class="fl">30</span>, <span class="kw">reduced.dim</span> <span class="kw">=</span> <span class="st">"pca.corrected"</span>)</pre></body></html></div>
</div>
<div id="testing" class="section level2">
<h2 class="hasAnchor">
<a href="#testing" class="anchor"></a>Testing</h2>
<p>Now we can do the DA test, explicitly defining our experimental design. In this case, we want to dest for differences between experimental stages, while accounting for the variability between technical batches (You can find more info on how to use formulas to define a testing design in R <a href="https://r4ds.had.co.nz/model-basics.html#formulas-and-model-families">here</a>)</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="no">da_results</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/testNhoods.html">testNhoods</a></span>(<span class="no">embryo_milo</span>, <span class="kw">design</span> <span class="kw">=</span> ~ <span class="no">sequencing.batch</span> + <span class="no">stage</span>, <span class="kw">design.df</span> <span class="kw">=</span> <span class="no">embryo_design</span>)</pre></body></html></div>
<p>This calculates a Fold-change and corrected P-value for each neighbourhood, which indicates wheather there is significant differential abundance between developmental stages.</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="no">da_results</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(<span class="no">SpatialFDR</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>()</pre></body></html></div>
<pre><code>##        logFC   logCPM        F       PValue         FDR Nhood  SpatialFDR
## 7  -7.397680 20.20389 13.18174 2.868468e-04 0.008307757     7 0.007518813
## 15 -7.407804 19.72189 14.72980 1.263205e-04 0.008307757    15 0.007518813
## 53  7.168031 19.37642 12.80220 3.510315e-04 0.008307757    53 0.007518813
## 64 -7.164631 19.67509 14.31767 1.570634e-04 0.008307757    64 0.007518813
## 69 -7.652294 19.96889 14.84507 1.188619e-04 0.008307757    69 0.007518813
## 71 -7.357238 19.56574 15.26142 9.542589e-05 0.008307757    71 0.007518813</code></pre>
</div>
</div>
<div id="inspecting-da-testing-results" class="section level1">
<h1 class="hasAnchor">
<a href="#inspecting-da-testing-results" class="anchor"></a>Inspecting DA testing results</h1>
<p>We can start inspecting the results of our DA analysis from a couple of standard diagnostic plots. We first inspect the distribution of uncorrected P values, to verify that the test was balanced.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="fu">ggplot</span>(<span class="no">da_results</span>, <span class="fu">aes</span>(<span class="no">PValue</span>)) + <span class="fu">geom_histogram</span>(<span class="kw">bins</span><span class="kw">=</span><span class="fl">50</span>)</pre></body></html></div>
<p><img src="milo_gastrulation_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>Then we visualize the test results with a volcano plot (remember that each point here represents a neighbourhood, <em>not</em> a cell).</p>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="fu">ggplot</span>(<span class="no">da_results</span>, <span class="fu">aes</span>(<span class="no">logFC</span>, -<span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="no">SpatialFDR</span>))) +
  <span class="fu">geom_point</span>() +
  <span class="fu">geom_hline</span>(<span class="kw">yintercept</span> <span class="kw">=</span> <span class="fl">1</span>) <span class="co">## Mark significance threshold (10% FDR)</span></pre></body></html></div>
<p><img src="milo_gastrulation_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>Looks like we have detected several neighbourhoods were there is a significant difference in cell abundances between developmental stages.</p>
<p>To visualize DA results relating them to the embedding of single cells, we can build an abstracted graph of neighbourhoods that we can superimpose on the single-cell embedding. Here each node represents a neighbourhood, while edges indicate how many cells two neighbourhoods have in common. Here the layout of nodes is determined by the position of the index cell in the UMAP embedding of all single-cells. The neighbourhoods displaying singificant DA are colored by their log-Fold Change.</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="no">embryo_milo</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/buildNhoodGraph.html">buildNhoodGraph</a></span>(<span class="no">embryo_milo</span>)

<span class="co">## Plot single-cell UMAP</span>
<span class="no">umap_pl</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span>(<span class="no">embryo_milo</span>, <span class="kw">dimred</span> <span class="kw">=</span> <span class="st">"umap"</span>, <span class="kw">colour_by</span><span class="kw">=</span><span class="st">"celltype"</span>, <span class="kw">text_by</span> <span class="kw">=</span> <span class="st">"celltype"</span>, <span class="kw">text_size</span> <span class="kw">=</span> <span class="fl">3</span>) +
  <span class="fu">guides</span>(<span class="kw">fill</span><span class="kw">=</span><span class="st">"none"</span>)

<span class="co">## Plot neighbourhood graph</span>
<span class="no">nh_graph_pl</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/plotNhoodGraphDA.html">plotNhoodGraphDA</a></span>(<span class="no">embryo_milo</span>, <span class="no">da_results</span>, <span class="kw">layout</span><span class="kw">=</span><span class="st">"umap"</span>,<span class="kw">alpha</span><span class="kw">=</span><span class="fl">0.05</span>)

<span class="no">umap_pl</span> + <span class="no">nh_graph_pl</span> +
  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span>(<span class="kw">guides</span><span class="kw">=</span><span class="st">"collect"</span>)</pre></body></html></div>
<p><img src="milo_gastrulation_files/figure-html/unnamed-chunk-16-1.png" width="1440"></p>
<p>We might also be interested in visualizing wheather DA is particularly evident in certain cell types. To do this, we assign a cell type label to each neighbourhood by finding the most abundant cell type within cells in each neighbourhood. We can label neighbourhoods in the results <code>data.frame</code> using the function <code>annotateNhoods</code>. This also saves the fraction of cells harbouring the label.</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="no">da_results</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/annotateNhoods.html">annotateNhoods</a></span>(<span class="no">embryo_milo</span>, <span class="no">da_results</span>, <span class="kw">coldata_col</span> <span class="kw">=</span> <span class="st">"celltype"</span>)
<span class="no">da_results</span></pre></body></html></div>
<pre><code>##           logFC   logCPM            F       PValue         FDR Nhood
## 1   -3.88539263 20.45252 5.799148e+00 1.608673e-02 0.047816065     1
## 2   -4.44088746 19.88934 7.199854e+00 7.326181e-03 0.029250374     2
## 3   -2.54346600 19.83440 3.097632e+00 7.849599e-02 0.137893980     3
## 4    6.83505949 19.26276 1.188472e+01 5.728147e-04 0.008852561     4
## 5   -1.24652000 19.90461 8.204600e-01 3.651097e-01 0.448232716     5
## 6    1.90556301 20.02227 1.753977e+00 1.854663e-01 0.266921141     6
## 7   -7.39767975 20.20389 1.318174e+01 2.868468e-04 0.008307757     7
## 8   -5.51371593 20.05600 1.025192e+01 1.377986e-03 0.012853905     8
## 9   -0.33091891 20.35682 5.808188e-02 8.095682e-01 0.848014792     9
## 10  -5.51682465 19.06571 1.038086e+01 1.285268e-03 0.012443735    10
## 11  -3.42893638 20.79687 4.481716e+00 3.433158e-02 0.078630388    11
## 12  -4.60615246 20.54331 7.368784e+00 6.670009e-03 0.028132910    12
## 13  -0.48808027 20.03086 1.292016e-01 7.192835e-01 0.775733643    13
## 14  -2.93460694 19.26300 4.462344e+00 3.472248e-02 0.079065207    14
## 15  -7.40780426 19.72189 1.472980e+01 1.263205e-04 0.008307757    15
## 16   6.05671986 20.56821 8.462845e+00 3.648197e-03 0.019961042    16
## 17   3.27625552 19.76024 3.398055e+00 6.536012e-02 0.120534242    17
## 18  -5.67952029 20.01764 1.107139e+01 8.860002e-04 0.010340715    18
## 19   0.38456485 20.08531 8.007785e-02 7.772091e-01 0.819532393    19
## 20  -2.37826841 19.72621 3.156472e+00 7.571564e-02 0.133678536    20
## 21  -3.51963280 20.13397 5.360060e+00 2.066218e-02 0.056243377    21
## 22  -2.06040807 19.70255 2.372036e+00 1.236187e-01 0.195042859    22
## 23  -1.75017868 19.82067 1.704277e+00 1.918177e-01 0.272835912    23
## 24  -3.81458170 19.75070 6.703978e+00 9.660720e-03 0.034439053    24
## 25   4.67019632 19.78926 6.682331e+00 9.778497e-03 0.034713666    25
## 26   1.58078432 20.15057 1.218849e+00 2.696646e-01 0.356761252    26
## 27   3.25768037 20.31611 4.074663e+00 4.360913e-02 0.090621903    27
## 28  -3.46826321 20.02479 5.635500e+00 1.765544e-02 0.050358439    28
## 29  -3.07500696 19.51182 4.923335e+00 2.656196e-02 0.065645765    29
## 30  -3.73271327 20.02222 6.036141e+00 1.406539e-02 0.044384131    30
## 31   1.04902960 19.71812 6.396499e-01 4.238933e-01 0.501607111    31
## 32  -3.92687396 20.81276 5.643719e+00 1.757300e-02 0.050358439    32
## 33   3.26040010 19.71295 4.248179e+00 3.936815e-02 0.084995337    33
## 34  -4.44395095 19.71177 8.452772e+00 3.668406e-03 0.019961042    34
## 35   0.04280167 20.02726 9.955960e-04 9.748303e-01 0.979625767    35
## 36  -4.59682956 19.99062 8.118456e+00 4.407975e-03 0.021934865    36
## 37  -4.97412167 19.78654 8.385737e+00 3.805834e-03 0.020084144    37
## 38  -4.41435900 20.10182 4.511696e+00 3.373566e-02 0.078147492    38
## 39  -3.88624370 19.76178 6.919350e+00 8.565167e-03 0.031728357    39
## 40   2.21608855 19.41633 1.974807e+00 1.600297e-01 0.237949848    40
## 41  -3.30302200 19.47520 4.799665e+00 2.853277e-02 0.069856082    41
## 42  -0.23600705 20.05159 2.934236e-02 8.640011e-01 0.892277496    42
## 43  -1.92284348 19.86231 2.062193e+00 1.510845e-01 0.227830087    43
## 44  -0.87264227 19.43917 4.620630e-01 4.967076e-01 0.579423587    44
## 45  -1.85954952 19.68324 2.066217e+00 1.506865e-01 0.227632808    45
## 46  -0.46212360 19.75639 1.157267e-01 7.337372e-01 0.789323324    46
## 47   2.96189639 20.22944 3.422195e+00 6.441194e-02 0.119744848    47
## 48  -2.53483450 19.67278 3.624198e+00 5.702916e-02 0.109099741    48
## 49  -6.17201406 19.79322 1.231080e+01 4.561685e-04 0.008398930    49
## 50   3.21230491 19.05774 3.664014e+00 5.568319e-02 0.108068509    50
## 51  -1.63837359 19.75579 1.576608e+00 2.093361e-01 0.293345930    51
## 52   0.07180693 20.23360 2.758996e-03 9.581125e-01 0.968341504    52
## 53   7.16803143 19.37642 1.280220e+01 3.510315e-04 0.008307757    53
## 54   3.44516108 19.02378 3.706136e+00 5.429588e-02 0.105616635    54
## 55  -3.37832879 19.93274 5.096239e+00 2.404082e-02 0.061325680    55
## 56  -1.39165591 19.58253 5.869042e-01 4.436718e-01 0.524283472    56
## 57  -2.52119305 19.63784 3.307260e+00 6.906213e-02 0.125193481    57
## 58  -1.48331444 19.16424 1.074751e+00 2.999482e-01 0.387205867    58
## 59   1.94688243 20.66532 1.570967e+00 2.101526e-01 0.293524644    59
## 60   3.26801439 20.62898 3.722746e+00 5.375894e-02 0.104811488    60
## 61  -3.99440596 20.48595 6.128885e+00 1.334732e-02 0.042743065    61
## 62  -0.39889404 20.07530 3.995430e-02 8.415819e-01 0.875491779    62
## 63  -4.07744441 19.42118 7.404181e+00 6.540308e-03 0.028035736    63
## 64  -7.16463090 19.67509 1.431767e+01 1.570634e-04 0.008307757    64
## 65   2.83705228 19.95735 3.418673e+00 6.454937e-02 0.119744848    65
## 66  -5.09291936 20.22496 9.310816e+00 2.295599e-03 0.015901222    66
## 67  -3.57139938 20.19653 5.789586e+00 1.617431e-02 0.047848986    67
## 68  -1.09980577 19.38877 6.887074e-01 4.066625e-01 0.486624247    68
## 69  -7.65229374 19.96889 1.484507e+01 1.188619e-04 0.008307757    69
## 70  -4.30089468 19.74194 7.748063e+00 5.406665e-03 0.024899883    70
## 71  -7.35723850 19.56574 1.526142e+01 9.542589e-05 0.008307757    71
## 72  -3.89635113 19.97808 5.983327e+00 1.449196e-02 0.045062583    72
## 73   4.41318556 19.47181 6.805016e+00 9.129922e-03 0.033242281    73
## 74   0.54275986 19.41269 1.874821e-01 6.650480e-01 0.731939686    74
## 75   0.47511060 19.65552 9.894985e-02 7.531128e-01 0.808126039    75
## 76  -3.44743693 19.73945 5.832939e+00 1.578113e-02 0.047510690    76
## 77  -5.79160022 20.28433 8.573621e+00 3.433278e-03 0.019441541    77
## 78  -4.92224147 20.07391 8.773323e+00 3.077800e-03 0.018337662    78
## 79  -0.12560263 19.38323 8.111262e-03 9.282428e-01 0.949415209    79
## 80  -6.77650921 20.63496 1.136908e+01 7.550898e-04 0.009662974    80
## 81  -3.73817170 18.93222 5.203089e+00 2.260823e-02 0.060383117    81
## 82  -5.65616320 19.80857 1.016213e+01 1.446523e-03 0.012944125    82
## 83   0.75631044 20.27966 2.645962e-01 6.070133e-01 0.679599650    83
## 84  -4.26942496 19.17398 7.337648e+00 6.786268e-03 0.028268389    84
## 85  -3.12701095 19.65336 5.135962e+00 2.349754e-02 0.060459533    85
## 86   2.35157057 20.64713 2.152534e+00 1.424271e-01 0.217469263    86
## 87  -3.78001922 19.91544 6.234369e+00 1.257628e-02 0.041692556    87
## 88  -4.47392087 20.02789 6.835141e+00 8.977517e-03 0.032969157    88
## 89   4.56440969 19.83395 6.122592e+00 1.339483e-02 0.042743065    89
## 90  -1.10557495 19.50767 7.100565e-01 3.994842e-01 0.481415186    90
## 91  -3.17640265 20.26997 4.678012e+00 3.062041e-02 0.073696578    91
## 92  -1.93349857 20.33878 1.826522e+00 1.766286e-01 0.256804697    92
## 93  -1.52005406 19.92591 1.137327e+00 2.862934e-01 0.374113526    93
## 94  -7.28873398 19.57320 1.522643e+01 9.720190e-05 0.008307757    94
## 95  -1.26554262 19.52421 9.890616e-01 3.200424e-01 0.403964680    95
## 96   2.28589772 19.57175 2.772263e+00 9.600238e-02 0.160380446    96
## 97  -2.86291451 21.02965 3.034972e+00 8.157819e-02 0.142136236    97
## 98  -3.53782052 20.07589 5.561952e+00 1.841110e-02 0.051643691    98
## 99  -0.52089282 19.56142 1.425982e-01 7.057351e-01 0.764023252    99
## 100  1.39673396 19.06396 1.089729e+00 2.966053e-01 0.384475088   100
## 101  1.23114527 19.67118 8.545833e-01 3.553244e-01 0.439480517   101
## 102 -1.84368164 20.18256 1.660911e+00 1.975678e-01 0.278227718   102
## 103  0.96625186 20.20552 4.621599e-01 4.966624e-01 0.579423587   103
## 104  2.87792737 19.32069 3.336426e+00 6.784906e-02 0.124050212   104
## 105 -4.66750792 19.33818 8.486566e+00 3.601052e-03 0.019937538   105
## 106 -2.62710223 19.25405 3.656169e+00 5.594570e-02 0.108101849   106
## 107 -1.80264740 20.04800 1.662062e+00 1.974126e-01 0.278227718   107
## 108 -4.31405447 19.55106 8.175686e+00 4.271361e-03 0.021683653   108
## 109 -3.62116724 19.83813 5.941917e+00 1.483573e-02 0.045467768   109
## 110 -2.25738793 20.28144 2.439724e+00 1.183899e-01 0.188538656   110
## 111  7.54102277 19.63448 1.357695e+01 2.325402e-04 0.008307757   111
## 112 -5.04838131 19.95410 8.637249e+00 3.315701e-03 0.018959578   112
## 113 -2.65763269 19.97077 3.520043e+00 6.071626e-02 0.113943282   113
## 114  2.10169203 20.23619 1.545016e+00 2.139582e-01 0.296410375   114
## 115  4.14117697 19.09927 5.972046e+00 1.458479e-02 0.045186327   115
## 116 -4.70025791 19.66278 6.133734e+00 1.331083e-02 0.042743065   116
## 117  0.82933335 20.20629 2.483568e-01 6.182666e-01 0.688579307   117
## 118 -3.79045785 19.52259 6.561490e+00 1.046357e-02 0.036536721   118
## 119 -3.26658475 19.65052 5.154699e+00 2.324571e-02 0.060459533   119
## 120 -2.75744063 19.97240 3.800543e+00 5.131788e-02 0.101681003   120
## 121 -4.50332364 20.48568 7.421056e+00 6.479383e-03 0.028022511   121
## 122 -3.87931693 19.90413 5.517894e+00 1.887971e-02 0.052395807   122
## 123  6.09723773 19.82050 1.009790e+01 1.497662e-03 0.012989204   123
## 124 -0.73047157 19.68813 3.061101e-01 5.801134e-01 0.655512721   124
## 125 -6.19179982 20.44575 1.147911e+01 7.118162e-04 0.009626467   125
## 126 -1.49784700 20.16443 1.060766e+00 3.031136e-01 0.390700102   126
## 127  4.71503852 19.93595 7.601092e+00 5.864367e-03 0.026048693   127
## 128 -1.54649100 19.52049 1.029615e+00 3.103211e-01 0.396987344   128
## 129 -5.43306706 19.52875 1.042001e+01 1.258392e-03 0.012323562   129
## 130 -3.87325229 19.66698 6.956447e+00 8.389716e-03 0.031489153   130
## 131  2.95633473 19.24743 4.166589e+00 4.130559e-02 0.087761499   131
## 132 -6.09350836 19.47152 8.860803e+00 2.934093e-03 0.018026682   132
## 133 -3.06482156 20.40400 4.142714e+00 4.189130e-02 0.088332829   133
## 134  0.65584675 20.34764 1.802640e-01 6.711733e-01 0.735958322   134
## 135  0.12281507 19.00724 1.411445e-02 9.054375e-01 0.930558155   135
## 136 -3.20585223 19.42183 5.381109e+00 2.041474e-02 0.055747957   136
## 137  4.46204459 19.17065 7.032276e+00 8.042438e-03 0.030799946   137
## 138  2.93082901 19.95474 2.454303e+00 1.172960e-01 0.187497542   138
## 139 -5.04293692 19.55513 9.274298e+00 2.341691e-03 0.015960967   139
## 140 -2.16282831 20.18617 2.350941e+00 1.253002e-01 0.195882225   140
## 141 -3.30132622 19.50660 5.142866e+00 2.340443e-02 0.060459533   141
## 142 -1.57605476 20.26683 1.315754e+00 2.514360e-01 0.340577894   142
## 143  6.84371575 19.20651 1.185545e+01 5.818585e-04 0.008852561   143
## 144 -3.67501511 20.27838 5.592415e+00 1.809414e-02 0.051216633   144
## 145  4.90752734 19.92427 7.028030e+00 8.061488e-03 0.030799946   145
## 146 -0.35708180 19.24583 8.652164e-02 7.686646e-01 0.816586281   146
## 147 -0.27303085 20.67787 3.148325e-02 8.591771e-01 0.889451871   147
## 148 -2.35422205 19.78432 2.516857e+00 1.127277e-01 0.181901460   148
## 149 -1.12019124 19.43087 5.321712e-01 4.657456e-01 0.549605592   149
## 150 -2.92964515 20.20854 3.416074e+00 6.465097e-02 0.119744848   150
## 151 -1.99241833 20.18373 2.038916e+00 1.534105e-01 0.230928842   151
## 152 -5.35955867 20.68892 8.795302e+00 3.041034e-03 0.018246204   152
## 153  4.13894769 19.87672 4.994216e+00 2.549675e-02 0.063816927   153
## 154 -2.61310928 19.74189 2.928928e+00 8.709622e-02 0.148575625   154
## 155 -0.73621319 19.76538 2.655120e-01 6.063918e-01 0.679599650   155
## 156 -3.23917699 19.38358 4.602105e+00 3.200319e-02 0.075396175   156
## 157 -0.52654481 20.20791 1.335015e-01 7.148515e-01 0.771930922   157
## 158  0.67417129 20.09174 2.353754e-01 6.275975e-01 0.698058874   158
## 159 -3.18951327 18.97258 4.458538e+00 3.479983e-02 0.079065207   159
## 160  3.31655289 19.86151 4.009157e+00 4.533287e-02 0.093327851   160
## 161 -7.78159818 19.99469 1.524050e+01 9.648375e-05 0.008307757   161
## 162 -0.05995922 19.70516 1.895415e-03 9.652765e-01 0.972852763   162
## 163  1.44967380 19.83294 8.515264e-01 3.561862e-01 0.439812536   163
## 164 -2.99568049 19.82943 4.161979e+00 4.141802e-02 0.087781477   164
## 165  1.39878538 19.61469 1.145139e+00 2.846448e-01 0.373488926   165
## 166 -4.16737274 20.01066 7.042969e+00 7.994664e-03 0.030799946   166
## 167  3.93448678 19.71336 4.992481e+00 2.552227e-02 0.063816927   167
## 168  0.61132514 19.79966 1.865985e-01 6.657902e-01 0.731939686   168
## 169 -2.70762575 19.72450 3.791657e+00 5.159067e-02 0.101984342   169
## 170 -1.98010935 19.98297 1.654490e+00 1.984363e-01 0.278989608   170
## 171 -3.91753672 19.84425 5.914636e+00 1.506677e-02 0.045682887   171
## 172 -7.72185391 19.84145 1.517545e+01 9.985004e-05 0.008307757   172
## 173 -4.36641814 20.10897 7.186385e+00 7.381256e-03 0.029250374   173
## 174  5.32101551 19.51739 8.707605e+00 3.190448e-03 0.018746633   174
## 175 -2.78313803 19.34041 3.884186e+00 4.882369e-02 0.098239843   175
## 176  0.44598093 20.37953 9.073686e-02 7.632608e-01 0.811869439   176
## 177  0.83534420 19.53975 4.319782e-01 5.110642e-01 0.592417320   177
## 178 -0.85626582 19.10505 4.439869e-01 5.052495e-01 0.588077297   178
## 179  2.38019715 19.90366 2.374019e+00 1.234619e-01 0.195042859   179
## 180  3.58749911 19.89642 3.964846e+00 4.653949e-02 0.095316452   180
## 181 -3.61562820 20.36675 5.727100e+00 1.675885e-02 0.049130810   181
## 182 -0.83186836 19.51101 4.092044e-01 5.224166e-01 0.603933460   182
## 183  3.52316934 19.82022 4.922036e+00 2.658191e-02 0.065645765   183
## 184 -2.87821383 20.21170 3.960420e+00 4.666185e-02 0.095337887   184
## 185  4.48672141 19.58484 6.992584e+00 8.222343e-03 0.031020382   185
## 186  4.24550992 19.29768 6.652560e+00 9.942887e-03 0.035084792   186
## 187  5.13216683 19.49239 8.757875e+00 3.103911e-03 0.018364807   187
## 188 -3.29672867 20.53592 4.506125e+00 3.384557e-02 0.078147492   188
## 189 -2.79139881 19.79648 3.976037e+00 4.623158e-02 0.094913999   189
## 190 -0.55885325 19.44750 1.917656e-01 6.614789e-01 0.729081573   190
## 191 -4.50735144 19.37400 9.288693e+00 2.323411e-03 0.015960967   191
## 192  2.48544894 20.04080 2.746319e+00 9.757065e-02 0.161661092   192
## 193 -1.76817959 20.35835 1.563996e+00 2.111669e-01 0.294191386   193
## 194  7.51790411 19.63836 1.351506e+01 2.403045e-04 0.008307757   194
## 195 -6.04167648 20.41897 1.050806e+01 1.199994e-03 0.011888317   195
## 196 -2.52136718 19.95240 3.253018e+00 7.138037e-02 0.128303964   196
## 197  1.26504828 20.20439 7.143035e-01 3.980782e-01 0.481415186   197
## 198 -2.94923287 20.65078 3.638874e+00 5.652906e-02 0.108675425   198
## 199 -5.16940563 19.89401 9.087251e+00 2.592940e-03 0.016864007   199
## 200  4.99568427 19.34166 8.038891e+00 4.605339e-03 0.022550280   200
## 201 -6.37643108 20.26595 1.141430e+01 7.369934e-04 0.009662974   201
## 202  3.11025612 19.92126 2.802374e+00 9.421658e-02 0.158016785   202
## 203 -2.31839926 19.63401 2.896732e+00 8.885056e-02 0.150498359   203
## 204  1.79263140 19.02491 1.470766e+00 2.253089e-01 0.309618079   204
## 205 -6.74875595 18.83765 1.192341e+01 5.610819e-04 0.008852561   205
## 206 -1.93508037 19.53413 2.018476e+00 1.554865e-01 0.232770965   206
## 207 -1.10318847 19.84374 7.913484e-02 7.784904e-01 0.819868713   207
## 208 -1.83721224 19.82139 1.841461e+00 1.748696e-01 0.255555615   208
## 209 -4.26147787 19.30880 8.443466e+00 3.687178e-03 0.019961042   209
## 210  7.57267029 19.75659 1.287165e+01 3.382885e-04 0.008307757   210
## 211 -7.33084188 19.58876 1.536521e+01 9.034685e-05 0.008307757   211
## 212  1.93136216 20.27199 1.704630e+00 1.917717e-01 0.272835912   212
## 213  1.34000333 19.40800 8.543095e-01 3.554015e-01 0.439480517   213
## 214 -4.02251242 20.26960 6.413410e+00 1.137070e-02 0.038906960   214
## 215  1.76491750 20.43331 1.277960e+00 2.583583e-01 0.346647609   215
## 216 -2.92987177 20.11249 3.877968e+00 4.900462e-02 0.098239843   216
## 217  4.39489836 19.21791 6.851457e+00 8.896063e-03 0.032811454   217
## 218 -2.71990919 20.72747 2.986385e+00 8.405802e-02 0.144710833   218
## 219  1.65460375 19.74604 1.222301e+00 2.689877e-01 0.356419112   219
## 220 -4.16027431 19.84455 7.258736e+00 7.090285e-03 0.029042898   220
## 221 -1.27313002 20.22365 8.590168e-01 3.540796e-01 0.439480517   221
## 222  5.11385301 19.19750 8.485208e+00 3.603733e-03 0.019937538   222
## 223 -3.88777031 20.16097 6.090595e+00 1.363910e-02 0.043198919   223
## 224  1.34169382 19.25323 9.808870e-01 3.220502e-01 0.405897633   224
## 225 -0.57589569 19.78448 1.568447e-01 6.921031e-01 0.750218660   225
## 226 -0.55151969 20.14622 1.595243e-01 6.896199e-01 0.748479196   226
## 227 -3.97070282 19.53920 7.478303e+00 6.276989e-03 0.027566983   227
## 228  3.64413973 19.66734 4.278358e+00 3.867589e-02 0.084406132   228
## 229  0.02051759 19.21529 9.445820e-04 9.754834e-01 0.979625767   229
## 230 -2.66425454 19.43526 4.065571e+00 4.384420e-02 0.090888711   230
## 231 -1.30748007 19.11545 1.017657e+00 3.131471e-01 0.398499996   231
## 232  3.26613581 19.59932 3.808906e+00 5.106255e-02 0.101410945   232
## 233 -1.02917894 19.61846 6.630643e-01 4.155371e-01 0.495157505   233
## 234 -5.92244819 20.41745 1.083001e+01 1.008853e-03 0.010905501   234
## 235  5.65388991 19.74931 9.131556e+00 2.531056e-03 0.016673795   235
## 236 -0.84844073 19.46386 3.853484e-01 5.347962e-01 0.615738283   236
## 237 -7.57221903 19.96639 1.419573e+01 1.675304e-04 0.008307757   237
## 238 -6.40344165 19.39531 1.132285e+01 7.740613e-04 0.009698533   238
## 239  2.43975757 19.61739 2.986058e+00 8.407496e-02 0.144710833   239
## 240 -0.03359610 20.28932 5.128229e-04 9.819343e-01 0.984244734   240
## 241 -6.71994633 20.50954 1.166732e+01 6.435178e-04 0.009137953   241
## 242 -2.69870572 19.72536 3.652290e+00 5.607598e-02 0.108101849   242
## 243 -2.55630662 20.31483 3.260844e+00 7.104079e-02 0.128234641   243
## 244  4.68022333 19.42362 7.453551e+00 6.363695e-03 0.027725246   244
## 245  5.03310357 19.26675 8.223875e+00 4.159674e-03 0.021456890   245
## 246 -4.91605645 19.42234 9.478544e+00 2.095438e-03 0.014877609   246
## 247 -5.15958885 20.42430 9.192814e+00 2.447955e-03 0.016328359   247
## 248 -2.47847370 20.32735 3.012212e+00 8.272983e-02 0.143805851   248
## 249  1.91306451 20.03965 1.416359e+00 2.340865e-01 0.318597024   249
## 250 -3.51707148 20.56712 5.178778e+00 2.292620e-02 0.060459533   250
## 251 -3.09842380 19.48932 4.277997e+00 3.868409e-02 0.084406132   251
## 252  2.96563293 19.90473 3.882501e+00 4.887265e-02 0.098239843   252
## 253 -2.68779576 19.28627 3.911310e+00 4.804263e-02 0.097924221   253
## 254  4.00592238 20.10230 5.658154e+00 1.742918e-02 0.050337845   254
## 255 -1.98489107 19.81571 2.194670e+00 1.385813e-01 0.212741000   255
## 256  4.93745923 19.91305 7.188080e+00 7.374303e-03 0.029250374   256
## 257 -5.44259640 19.93147 9.893199e+00 1.673196e-03 0.013425612   257
## 258 -2.56027186 19.62362 3.347364e+00 6.740003e-02 0.123494244   258
## 259  0.41983395 19.90543 8.224328e-02 7.742975e-01 0.819504957   259
## 260 -3.19093615 19.70185 5.076683e+00 2.431305e-02 0.061650942   260
## 261  2.24529902 19.34336 1.674394e+00 1.957588e-01 0.277514901   261
## 262  0.66380256 19.01778 2.615798e-01 6.090700e-01 0.681007416   262
## 263 -3.67923963 20.07686 5.951604e+00 1.475456e-02 0.045382251   263
## 264 -5.43290595 20.26680 9.242685e+00 2.382350e-03 0.016109227   264
## 265 -3.87627643 19.36843 6.947936e+00 8.429639e-03 0.031500230   265
## 266 -1.21994541 19.79165 6.583113e-01 4.172134e-01 0.496460635   266
## 267 -0.71793897 19.91457 2.876367e-01 5.917740e-01 0.665160222   267
## 268 -2.72606100 19.84406 3.330590e+00 6.808997e-02 0.124224091   268
## 269  5.60408012 19.63276 9.648546e+00 1.910602e-03 0.014534221   269
## 270 -5.68271019 19.93096 1.061065e+01 1.135406e-03 0.011516262   270
## 271 -1.43119787 19.23217 7.637082e-01 3.822321e-01 0.465231075   271
## 272 -5.02058273 19.55931 9.013364e+00 2.699580e-03 0.017409953   272
## 273 -5.45431785 20.59719 8.811805e+00 3.013723e-03 0.018210580   273
## 274 -0.89773342 20.89442 3.384142e-01 5.607845e-01 0.637164562   274
## 275  7.29951756 19.65694 1.283741e+01 3.445117e-04 0.008307757   275
## 276  2.12015318 19.47302 2.414186e+00 1.203333e-01 0.190919883   276
## 277 -6.72731883 19.35935 1.300753e+01 3.146907e-04 0.008307757   277
## 278  3.47137592 19.04480 4.498884e+00 3.398899e-02 0.078266534   278
## 279 -7.48354060 19.97579 1.420318e+01 1.668712e-04 0.008307757   279
## 280 -2.80988702 19.41190 4.207620e+00 4.031906e-02 0.086311151   280
## 281  5.76607456 19.74434 9.522055e+00 2.046468e-03 0.014767399   281
## 282  0.20124362 19.18271 1.725781e-02 8.954912e-01 0.922561655   282
## 283 -4.96608444 19.87432 8.662101e+00 3.270896e-03 0.018902846   283
## 284  1.15665440 19.17899 7.700143e-01 3.802745e-01 0.464174599   284
## 285 -2.08753027 20.32017 2.214064e+00 1.368505e-01 0.210463244   285
## 286  4.26454152 19.57989 6.155707e+00 1.314677e-02 0.042740882   286
## 287 -1.92025358 20.16919 1.893967e+00 1.688456e-01 0.248456781   287
## 288 -6.97301573 19.81402 1.255209e+01 4.010708e-04 0.008307757   288
## 289 -2.14649050 19.69841 2.613677e+00 1.060384e-01 0.172413602   289
## 290 -0.90326809 20.15182 4.363999e-01 5.089099e-01 0.590723710   290
## 291  1.44027278 19.74441 9.000260e-01 3.428420e-01 0.427674068   291
## 292  2.40366578 19.23503 2.858460e+00 9.098591e-02 0.153809511   292
## 293  3.11199883 20.10673 3.903166e+00 4.827576e-02 0.097930824   293
## 294  1.73253186 19.46965 1.503443e+00 2.202271e-01 0.303123606   294
## 295 -5.23095442 19.80366 9.938108e+00 1.632980e-03 0.013425612   295
## 296 -1.94291734 19.65182 1.729588e+00 1.885520e-01 0.270448342   296
## 297 -3.68270907 19.74674 5.393680e+00 2.026843e-02 0.055526369   297
## 298  2.52773386 20.16020 2.358714e+00 1.246777e-01 0.195763320   298
## 299  2.67708705 19.57564 2.445519e+00 1.179538e-01 0.188195887   299
## 300 -2.62831399 20.18455 3.281275e+00 7.016246e-02 0.126918075   300
## 301 -5.48449858 18.98820 7.187091e+00 7.378361e-03 0.029250374   301
## 302  3.97798308 18.87673 5.284994e+00 2.157019e-02 0.057974135   302
## 303  0.13484344 20.13416 8.613541e-03 9.260606e-01 0.948321698   303
## 304  7.24720422 19.68009 1.249153e+01 4.142342e-04 0.008307757   304
## 305  3.33046284 19.90464 3.254708e+00 7.130689e-02 0.128303964   305
## 306 -0.72464819 19.02467 3.082182e-01 5.788121e-01 0.654910862   306
## 307  1.47938933 19.46921 1.143496e+00 2.849906e-01 0.373488926   307
## 308 -3.44923294 19.55948 5.920073e+00 1.502043e-02 0.045682887   308
## 309  6.26452038 18.98496 9.604900e+00 1.956417e-03 0.014626600   309
## 310 -4.11065940 19.11147 6.931120e+00 8.509096e-03 0.031658294   310
## 311 -3.28836102 20.18094 3.871875e+00 4.918261e-02 0.098365216   311
## 312 -2.87292360 20.12635 3.746839e+00 5.299010e-02 0.103849981   312
## 313 -4.26104078 20.24458 7.109982e+00 7.701802e-03 0.030100622   313
## 314 -0.44645469 20.00689 1.068695e-01 7.437568e-01 0.799093077   314
## 315 -1.69560129 19.65824 1.416782e+00 2.340167e-01 0.318597024   315
## 316 -0.86397780 19.39256 4.622136e-01 4.966374e-01 0.579423587   316
## 317  4.27800245 19.10571 6.472787e+00 1.099764e-02 0.037935170   317
## 318 -3.17508757 19.94578 4.443802e+00 3.510098e-02 0.079537331   318
## 319 -3.59456903 20.38653 5.578015e+00 1.824326e-02 0.051467746   319
## 320 -1.59546689 19.80776 1.454867e+00 2.278324e-01 0.311578111   320
## 321 -4.93172574 20.26026 8.848178e+00 2.954403e-03 0.018026682   321
## 322 -2.31377066 19.66012 3.073537e+00 7.966604e-02 0.139374671   322
## 323 -4.20977965 20.79635 5.814870e+00 1.594379e-02 0.047669269   323
## 324 -6.59294542 19.52363 1.231528e+01 4.550799e-04 0.008398930   324
## 325  3.14458758 19.96990 4.050312e+00 4.424172e-02 0.091490151   325
## 326 -1.35006686 19.85551 1.017549e+00 3.131727e-01 0.398499996   326
## 327  7.35107892 19.69202 1.273205e+01 3.643943e-04 0.008307757   327
## 328  3.42355686 19.44477 4.558506e+00 3.282676e-02 0.076625747   328
## 329  4.75169010 19.65686 7.335024e+00 6.796162e-03 0.028268389   329
## 330 -2.95718178 19.69544 4.604733e+00 3.195425e-02 0.075396175   330
## 331  0.39105617 19.33166 9.456230e-02 7.584741e-01 0.810752978   331
## 332  6.89496627 19.35934 1.202403e+01 5.316881e-04 0.008805029   332
## 333  2.43596503 20.36142 2.508197e+00 1.133482e-01 0.182557066   333
## 334 -4.86412429 19.76864 8.430215e+00 3.714077e-03 0.019961042   334
## 335 -4.33137157 20.12382 4.578983e+00 3.243723e-02 0.076133667   335
## 336 -1.38057923 19.47393 1.090961e+00 2.963324e-01 0.384475088   336
## 337 -1.41852409 19.63862 1.170415e+00 2.793925e-01 0.368486740   337
## 338  1.66905951 19.71419 1.197520e+00 2.738953e-01 0.361796518   338
## 339 -5.86955331 20.39999 1.012773e+01 1.473686e-03 0.012944125   339
## 340 -2.66652868 19.35288 3.904570e+00 4.823547e-02 0.097930824   340
## 341  0.24964012 20.06914 2.950676e-02 8.636244e-01 0.892277496   341
## 342  7.51731665 19.78181 1.255750e+01 3.999174e-04 0.008307757   342
## 343 -3.42272187 19.49014 6.152269e+00 1.317230e-02 0.042740882   343
## 344 -3.97958991 20.88239 5.523697e+00 1.881729e-02 0.052395807   344
## 345 -2.81140014 19.38388 4.275729e+00 3.873568e-02 0.084406132   345
## 346 -0.38910498 19.18869 9.155435e-02 7.622288e-01 0.811869439   346
## 347 -3.58408822 19.53526 6.427533e+00 1.128080e-02 0.038755019   347
## 348  0.73936344 19.50250 2.029689e-01 6.523637e-01 0.721836224   348
## 349 -4.69186599 19.48196 7.047652e+00 7.973829e-03 0.030799946   349
## 350 -2.19282937 20.30778 2.399722e+00 1.214497e-01 0.192332926   350
## 351  4.60608273 19.64663 6.809083e+00 9.109194e-03 0.033242281   351
## 352  1.69518805 19.10056 1.607639e+00 2.049107e-01 0.287617640   352
## 353  1.96909724 20.17261 1.763834e+00 1.842360e-01 0.265599046   353
## 354  1.43777500 19.19365 1.027587e+00 3.107980e-01 0.397001335   354
## 355 -2.13392175 19.64165 2.471616e+00 1.160115e-01 0.186142755   355
## 356 -1.60201484 20.04480 1.363843e+00 2.429543e-01 0.329613169   356
## 357  2.83627690 19.88828 3.607444e+00 5.760580e-02 0.109632179   357
## 358 -2.45151189 19.50465 3.493634e+00 6.169091e-02 0.115264594   358
## 359  1.23435833 20.34953 6.443275e-01 4.222041e-01 0.500999807   359
## 360 -0.40329435 19.83552 9.072835e-02 7.632716e-01 0.811869439   360
## 361 -4.45087549 19.54712 8.875192e+00 2.911116e-03 0.018026682   361
## 362  0.43398080 20.37232 7.433762e-02 7.851394e-01 0.824831984   362
## 363  3.79973247 19.90719 5.427804e+00 1.987671e-02 0.054628882   363
## 364 -2.83535160 19.90672 2.676955e+00 1.019026e-01 0.166642955   364
## 365  2.25758782 19.90567 2.263996e+00 1.325046e-01 0.205121798   365
## 366  4.55437747 19.97760 5.961260e+00 1.467411e-02 0.045298343   366
## 367  2.96541835 20.57349 3.370029e+00 6.647967e-02 0.122334083   367
## 368 -3.59986170 20.76658 4.940224e+00 2.630402e-02 0.065338253   368
## 369 -5.91189680 19.81348 1.117685e+01 8.371841e-04 0.010046209   369
## 370  2.01547467 19.08966 2.001139e+00 1.572724e-01 0.234669218   370
## 371  3.54439178 19.76399 4.080830e+00 4.345045e-02 0.090512916   371
## 372 -2.30429122 20.20952 2.585350e+00 1.079492e-01 0.174853066   372
## 373  0.09835567 19.66355 5.436728e-03 9.412262e-01 0.956950781   373
## 374  5.78237703 19.74061 9.604276e+00 1.957080e-03 0.014626600   374
## 375 -0.94592694 19.70340 4.611469e-01 4.971346e-01 0.579423587   375
## 376 -5.63452163 19.70281 1.082571e+01 1.011191e-03 0.010905501   376
## 377 -1.98495717 19.11949 1.228128e+00 2.678495e-01 0.355463870   377
## 378  0.78283096 19.86894 2.938940e-01 5.877713e-01 0.661533907   378
## 379  2.00066629 20.47926 1.711767e+00 1.908446e-01 0.272818063   379
## 380 -3.09763562 20.06430 4.618811e+00 3.169337e-02 0.075396175   380
## 381 -2.52698258 20.04676 3.237496e+00 7.205900e-02 0.129251091   381
## 382 -3.53526893 19.26256 6.007165e+00 1.429780e-02 0.044621714   382
## 383  1.53222662 19.69851 8.951580e-01 3.441504e-01 0.428052814   383
## 384  4.05452415 19.22588 6.258921e+00 1.240347e-02 0.041391738   384
## 385 -3.48320684 19.44935 5.190074e+00 2.277789e-02 0.060459533   385
## 386 -0.79368791 19.67950 3.489905e-01 5.547243e-01 0.633941947   386
## 387 -1.72430517 19.96230 1.532070e+00 2.158872e-01 0.298597282   387
## 388 -1.37416116 19.60381 1.148948e+00 2.838454e-01 0.373204147   388
## 389 -2.30672119 20.88755 2.175849e+00 1.402845e-01 0.214582352   389
## 390  1.29088174 19.89491 7.552960e-01 3.848658e-01 0.467768466   390
## 391 -0.81442311 19.52699 4.054473e-01 5.243321e-01 0.605326508   391
## 392 -5.33073680 19.74082 9.352497e+00 2.244114e-03 0.015801529   392
## 393  1.39471201 20.36175 8.246528e-01 3.638875e-01 0.447376843   393
## 394 -0.10948063 20.18711 6.047552e-03 9.380188e-01 0.955781889   394
## 395 -2.02894284 20.18002 1.704465e+00 1.917932e-01 0.272835912   395
## 396 -4.04178153 19.47650 7.297312e+00 6.939968e-03 0.028564504   396
## 397  2.48320544 19.16536 2.725438e+00 9.885319e-02 0.162916352   397
## 398  4.43657652 20.35299 5.168799e+00 2.305806e-02 0.060459533   398
## 399  3.93073413 19.57831 5.234971e+00 2.219814e-02 0.059474271   399
## 400  3.34261770 19.28528 5.185536e+00 2.283735e-02 0.060459533   400
## 401  1.22337514 19.27054 6.945886e-01 4.046664e-01 0.486252277   401
## 402  0.53462300 19.55600 1.627420e-01 6.866698e-01 0.746227900   402
## 403 -2.18485904 20.18573 2.262238e+00 1.326551e-01 0.205121798   403
## 404  5.74967638 19.70658 9.730354e+00 1.827637e-03 0.014285746   404
## 405  2.41962298 19.72933 2.927147e+00 8.719227e-02 0.148575625   405
## 406  3.56906969 19.74121 4.329032e+00 3.754208e-02 0.083732595   406
## 407  3.94342947 20.17568 4.995462e+00 2.547842e-02 0.063816927   407
## 408  3.20234114 19.78205 3.128452e+00 7.702622e-02 0.135591605   408
## 409  7.32001087 19.53199 1.298127e+01 3.191180e-04 0.008307757   409
## 410  2.71617582 20.22923 3.166117e+00 7.527009e-02 0.133678536   410
## 411 -2.96097232 19.43832 4.600425e+00 3.203452e-02 0.075396175   411
## 412  1.62150558 19.61562 1.255717e+00 2.625422e-01 0.350604991   412
## 413  1.28583211 20.35630 7.655293e-01 3.816653e-01 0.465205785   413
## 414 -6.38363803 20.72823 1.066553e+01 1.102308e-03 0.011453245   414
## 415 -1.13775606 19.94996 6.777026e-01 4.104366e-01 0.490451532   415
## 416  1.55332482 19.45297 1.088263e+00 2.969303e-01 0.384475088   416
## 417 -3.94059190 20.11210 6.109483e+00 1.349436e-02 0.042899972   417
## 418 -1.36196849 19.38907 7.835403e-01 3.761230e-01 0.459765884   418
## 419 -7.85154208 20.19247 1.482414e+01 1.201821e-04 0.008307757   419
## 420  2.11777359 19.22133 2.016128e+00 1.557271e-01 0.232770965   420
## 421 -2.37844070 20.25635 2.761385e+00 9.665655e-02 0.161157296   421
## 422 -1.89217951 19.63087 1.858660e+00 1.728696e-01 0.253066828   422
## 423 -2.55485762 19.76440 2.098868e+00 1.475003e-01 0.223890948   423
## 424  3.76434573 19.67860 4.991159e+00 2.554175e-02 0.063816927   424
## 425  2.25799996 20.57870 1.952605e+00 1.623973e-01 0.239796401   425
## 426 -1.15878916 19.13670 8.076977e-01 3.688653e-01 0.452191682   426
## 427 -0.83034364 20.15433 3.572389e-01 5.500834e-01 0.630782063   427
## 428  3.04233031 20.10324 3.652141e+00 5.608101e-02 0.108101849   428
## 429 -5.19169512 19.90434 8.843398e+00 2.962131e-03 0.018026682   429
## 430  3.27148552 19.72415 4.609058e+00 3.187386e-02 0.075396175   430
## 431 -0.51861072 19.31727 1.635316e-01 6.859511e-01 0.746227900   431
## 432  1.69232319 19.28068 1.266921e+00 2.604244e-01 0.348870399   432
## 433 -2.85013310 19.18182 3.864820e+00 4.938954e-02 0.098547757   433
## 434 -5.27089525 19.30705 1.014150e+01 1.462752e-03 0.012944125   434
## 435 -1.47567382 19.17349 1.304934e+00 2.533941e-01 0.342685399   435
## 436 -1.66813549 20.28323 1.418125e+00 2.337952e-01 0.318597024   436
## 437 -2.96326079 20.01566 4.156427e+00 4.155384e-02 0.087850801   437
## 438  2.29874742 19.87214 2.530637e+00 1.117479e-01 0.180662584   438
## 439 -0.39028229 19.60326 8.009369e-02 7.771877e-01 0.819532393   439
## 440 -2.56335702 20.00395 3.178221e+00 7.471492e-02 0.133173872   440
## 441 -7.15257749 19.70694 1.311773e+01 2.967762e-04 0.008307757   441
## 442 -2.95855373 20.28289 4.178628e+00 4.101353e-02 0.087358823   442
## 443 -3.97948946 19.47903 5.708543e+00 1.693661e-02 0.049130810   443
## 444 -0.92114338 19.52174 5.293462e-01 4.669318e-01 0.550243315   444
## 445 -1.85346634 19.47647 1.961892e+00 1.614022e-01 0.238740754   445
## 446 -3.16770681 20.57391 4.122434e+00 4.239566e-02 0.088968232   446
## 447  4.45877269 19.23755 6.174501e+00 1.300809e-02 0.042740882   447
## 448 -2.61062404 20.81125 2.834251e+00 9.236540e-02 0.155217594   448
## 449  0.18333290 20.81321 1.528149e-02 9.016248e-01 0.927758819   449
## 450  7.10344497 19.38183 1.260622e+01 3.896639e-04 0.008307757   450
## 451  6.54151996 19.11531 1.075966e+01 1.047799e-03 0.011021292   451
## 452  4.72480630 19.62456 7.324800e+00 6.834845e-03 0.028268389   452
## 453  6.13600598 19.79951 1.023501e+01 1.390637e-03 0.012853905   453
## 454  6.69471577 19.42439 1.102017e+01 9.107373e-04 0.010345975   454
## 455 -4.18237452 19.57592 7.837335e+00 5.146607e-03 0.024226017   455
## 456  2.56303229 19.95294 2.756344e+00 9.696136e-02 0.161349770   456
## 457 -4.04526314 19.32924 7.328735e+00 6.819928e-03 0.028268389   457
## 458  1.51205447 19.28804 1.117733e+00 2.904820e-01 0.377926008   458
## 459 -1.33095172 19.65768 9.906989e-01 3.196423e-01 0.403964680   459
## 460  4.26101010 19.15647 6.396690e+00 1.147807e-02 0.039117274   460
## 461 -1.19013979 19.46126 8.393993e-01 3.596335e-01 0.442987780   461
## 462 -4.35416880 20.27363 6.648522e+00 9.965399e-03 0.035084792   462
## 463  7.52920132 19.61298 1.357199e+01 2.331531e-04 0.008307757   463
## 464 -3.95681790 19.99821 4.862974e+00 2.750555e-02 0.067535240   464
## 465  5.00183871 19.35378 7.989304e+00 4.732873e-03 0.022911408   465
## 466 -2.29970684 20.03598 2.703678e+00 1.002093e-01 0.164189020   466
## 467 -0.77955451 19.92261 3.047163e-01 5.809769e-01 0.655618963   467
## 468 -5.79489236 19.65016 1.181568e+01 5.943730e-04 0.008884312   468
## 469 -1.25664395 19.35034 9.282944e-01 3.353749e-01 0.419587938   469
## 470 -2.48848634 19.14802 3.180272e+00 7.462129e-02 0.133173872   470
## 471  2.08374739 20.17495 1.838085e+00 1.752654e-01 0.255695449   471
## 472  0.94035210 20.61805 3.978611e-01 5.282382e-01 0.609010805   472
## 473  0.39634163 19.14384 9.660189e-02 7.559653e-01 0.809148744   473
## 474 -2.52606838 19.70020 2.908557e+00 8.820178e-02 0.149697051   474
## 475 -3.97300212 19.90555 5.112833e+00 2.381229e-02 0.060925149   475
## 476  1.22070226 19.93308 7.907878e-01 3.739248e-01 0.457735514   476
## 477  0.37085485 19.78803 5.769426e-02 8.101925e-01 0.848014792   477
## 478 -3.99537225 20.18740 6.733746e+00 9.501131e-03 0.034012452   478
## 479  7.07938078 19.58683 1.224223e+01 4.731792e-04 0.008398930   479
## 480 -4.30883914 20.37081 7.140578e+00 7.571760e-03 0.029728755   480
## 481 -0.38421726 19.27945 8.234304e-02 7.741644e-01 0.819504957   481
## 482  1.72692329 19.75612 1.249342e+00 2.637570e-01 0.351676038   482
## 483 -2.08391182 19.65699 1.748744e+00 1.861235e-01 0.267415156   483
## 484 -4.03873135 19.42796 6.019077e+00 1.420178e-02 0.044581953   484
## 485 -1.87954113 19.52534 1.962040e+00 1.613863e-01 0.238740754   485
## 486  2.28130141 19.74039 2.304968e+00 1.290537e-01 0.200279983   486
## 487 -1.33112523 19.43428 9.680815e-01 3.252290e-01 0.408694796   487
## 488 -3.07689020 19.27292 3.446904e+00 6.345661e-02 0.118304232   488
## 489 -2.10226175 20.39407 2.235278e+00 1.349849e-01 0.207969561   489
## 490 -0.30428506 20.00924 4.345911e-02 8.348756e-01 0.870641435   490
## 491 -3.45805966 20.07419 5.331195e+00 2.100655e-02 0.056998674   491
## 492 -0.10886341 20.16310 5.163450e-03 9.427198e-01 0.957326907   492
## 493  0.29573373 19.36711 3.873081e-02 8.439947e-01 0.876931038   493
## 494  1.24206207 19.74412 7.414879e-01 3.892452e-01 0.471745293   494
## 495 -2.24590180 19.79898 2.716840e+00 9.938656e-02 0.163469791   495
## 496 -1.39385150 19.44732 1.049530e+00 3.056880e-01 0.393118709   496
## 497  3.92420629 20.01219 5.633780e+00 1.767274e-02 0.050358439   497
## 498 -4.57965544 19.31590 8.424801e+00 3.725124e-03 0.019961042   498
## 499 -1.07674567 19.68668 6.923452e-01 4.054261e-01 0.486252277   499
## 500 -2.40079188 19.55648 2.996961e+00 8.351136e-02 0.144617240   500
## 501 -5.07650340 19.63206 9.937891e+00 1.633171e-03 0.013425612   501
## 502 -1.85286523 19.49649 1.859667e+00 1.727533e-01 0.253066828   502
## 503 -1.99491453 19.59426 2.327792e+00 1.271749e-01 0.198085879   503
## 504  3.79215780 19.63899 6.323194e+00 1.196257e-02 0.040444873   504
## 505  2.35520224 19.20144 2.357626e+00 1.247647e-01 0.195763320   505
## 506 -1.41402525 20.00871 1.077500e+00 2.993311e-01 0.386995574   506
## 507 -1.88211212 19.40389 2.085424e+00 1.488028e-01 0.225186536   507
## 508  0.64941828 19.46323 1.772238e-01 6.737965e-01 0.736027475   508
## 509 -1.49659622 19.77641 9.696198e-01 3.248449e-01 0.408694796   509
## 510 -3.05359716 19.57541 4.865221e+00 2.746981e-02 0.067535240   510
## 511 -4.54554801 20.04143 7.239985e+00 7.164552e-03 0.029081694   511
## 512 -3.74407448 19.51327 5.132605e+00 2.354296e-02 0.060459533   512
## 513 -4.42922996 19.18678 6.755111e+00 9.388257e-03 0.033750189   513
## 514  7.01324557 19.41611 1.218758e+01 4.871933e-04 0.008471198   514
## 515 -1.76133103 19.00029 1.716382e+00 1.902477e-01 0.272421976   515
## 516  3.38125288 19.64017 4.713828e+00 2.998974e-02 0.072467140   516
## 517  2.13884498 19.14742 2.424598e+00 1.195367e-01 0.190009845   517
## 518  1.13067433 20.35887 6.399153e-01 4.237972e-01 0.501607111   518
## 519  4.14964241 19.66056 5.931340e+00 1.492487e-02 0.045577016   519
## 520 -3.13701828 19.37794 3.083410e+00 7.918438e-02 0.138817053   520
## 521 -2.74822459 20.08604 3.745832e+00 5.302200e-02 0.103849981   521
## 522 -1.49573382 19.20360 1.278041e+00 2.583432e-01 0.346647609   522
## 523  5.67028585 19.49456 9.861876e+00 1.701838e-03 0.013425612   523
## 524 -4.21173427 19.63960 7.859676e+00 5.083545e-03 0.024196537   524
## 525 -5.93026602 19.97928 1.027676e+01 1.359607e-03 0.012853905   525
## 526  3.67207548 19.72034 5.180692e+00 2.290101e-02 0.060459533   526
## 527 -4.44046089 19.99738 6.991479e+00 8.227408e-03 0.031020382   527
## 528 -1.57670168 19.18493 1.126565e+00 2.885843e-01 0.376529620   528
## 529 -1.66968631 19.77475 1.467117e+00 2.258851e-01 0.309910044   529
## 530  6.64323117 19.24970 1.124704e+01 8.062109e-04 0.009812738   530
## 531 -1.77927765 19.26702 1.039559e+00 3.079964e-01 0.394993627   531
## 532 -2.16610354 19.21886 2.743910e+00 9.771768e-02 0.161661092   532
## 533 -0.43045883 20.31251 9.384200e-02 7.593672e-01 0.810752978   533
## 534  0.05022031 20.07456 8.919248e-04 9.761764e-01 0.979625767   534
## 535 -3.24021984 19.47585 5.708153e+00 1.694037e-02 0.049130810   535
## 536 -4.65961025 19.76532 8.944965e+00 2.802271e-03 0.017817427   536
## 537 -1.89445471 19.54471 1.928636e+00 1.649985e-01 0.243215717   537
## 538 -2.88478110 19.66518 4.751679e+00 2.933794e-02 0.071213461   538
## 539 -0.82968049 20.20447 3.382418e-01 5.608843e-01 0.637164562   539
## 540 -0.63713912 19.71247 2.494727e-01 6.174789e-01 0.688579307   540
## 541 -4.25766224 20.14325 7.632741e+00 5.762574e-03 0.025977318   541
## 542  3.78363710 19.37174 6.128518e+00 1.335009e-02 0.042743065   542
## 543 -2.72815412 20.05141 3.605606e+00 5.766945e-02 0.109632179   543
## 544  5.15749494 19.14343 8.567068e+00 3.445625e-03 0.019441541   544
## 545  3.15380199 19.56683 3.403352e+00 6.515079e-02 0.120408842   545
## 546  0.88464439 19.61330 3.674781e-01 5.444225e-01 0.625132041   546
## 547  5.56358714 19.61872 9.536684e+00 2.030267e-03 0.014767399   547
## 548  7.08406832 19.41000 1.243593e+01 4.267048e-04 0.008307757   548
## 549 -2.11408833 19.76852 2.028157e+00 1.544993e-01 0.232157627   549
## 550 -2.80850431 20.26009 3.739746e+00 5.321522e-02 0.103989365   550
## 551  2.71794286 19.79959 2.182500e+00 1.396799e-01 0.214041922   551
## 552 -3.16086021 19.31643 4.425054e+00 3.548811e-02 0.079989064   552
## 553 -3.84103086 19.89902 5.305322e+00 2.132029e-02 0.057483817   553
## 554 -3.62435393 19.69074 5.714999e+00 1.687455e-02 0.049130810   554
## 555  2.87284012 19.43211 3.537969e+00 6.006404e-02 0.113218049   555
## 556  2.98602241 19.61799 4.413993e+00 3.571857e-02 0.080296107   556
## 557 -0.10106770 19.85473 4.518976e-03 9.464079e-01 0.959015049   557
## 558  0.55348845 19.19259 1.646451e-01 6.849409e-01 0.746227900   558
## 559  6.67940881 19.16478 1.128182e+01 7.912951e-04 0.009770775   559
## 560 -0.39539441 19.65642 9.750945e-02 7.548582e-01 0.808980125   560
## 561 -0.55063758 19.32100 1.249906e-01 7.237055e-01 0.779515860   561
## 562  5.32659980 20.09296 7.958452e+00 4.814034e-03 0.023172639   562
## 563 -5.52192060 19.82044 1.052873e+01 1.186685e-03 0.011888317   563
## 564 -1.45435501 19.58077 1.259023e+00 2.619152e-01 0.350316726   564
## 565 -3.89964473 20.25150 6.535526e+00 1.061708e-02 0.036921430   565
## 566  6.93577087 19.48968 1.169519e+01 6.339800e-04 0.009137953   566
## 567 -6.58798906 20.14990 1.226352e+01 4.678299e-04 0.008398930   567
## 568 -3.05441729 19.79599 4.485341e+00 3.425894e-02 0.078630388   568
## 569 -0.10519294 20.07365 5.865925e-03 9.389547e-01 0.955781889   569
## 570  1.36218864 19.74144 6.701633e-01 4.130519e-01 0.492885465   570
## 571  0.89187868 19.68988 3.333893e-01 5.637083e-01 0.639519897   571
## 572 -0.08404025 19.27380 6.171612e-03 9.373875e-01 0.955781889   572
## 573 -2.05549844 19.97760 1.828523e+00 1.763918e-01 0.256804697   573
## 574 -5.36707680 20.32481 9.891619e+00 1.674629e-03 0.013425612   574
## 575 -2.78705798 19.92605 3.881114e+00 4.891301e-02 0.098239843   575
## 576 -7.11889983 19.66141 1.426954e+01 1.611133e-04 0.008307757   576
## 577  0.65595511 19.19314 1.771876e-01 6.738280e-01 0.736027475   577
## 578 -1.43909153 19.97035 1.048555e+00 3.059128e-01 0.393118709   578
## 579 -2.84553837 19.66002 4.305645e+00 3.806096e-02 0.084010199   579
## 580  5.63572566 19.36710 1.002587e+01 1.557191e-03 0.013267267   580
## 581 -2.85373696 19.41239 4.634341e+00 3.140816e-02 0.075167845   581
## 582  6.34341140 19.00023 9.892581e+00 1.673756e-03 0.013425612   582
## 583 -3.53121760 19.62349 5.324635e+00 2.108565e-02 0.057031665   583
## 584 -3.04026293 19.74501 4.606343e+00 3.192430e-02 0.075396175   584
## 585 -0.34437530 20.07419 4.638422e-02 8.294918e-01 0.866087010   585
## 586 -0.34157855 19.11013 4.169596e-02 8.382126e-01 0.873052749   586
## 587 -2.45380418 19.98656 2.991952e+00 8.376981e-02 0.144710833   587
## 588  1.60332095 19.97053 1.288329e+00 2.564360e-01 0.345187817   588
## 589  5.58375924 19.58488 9.648936e+00 1.910197e-03 0.014534221   589
## 590  4.83028963 19.64529 7.665954e+00 5.657687e-03 0.025640157   590
## 591 -0.45748184 19.35274 1.347506e-01 7.135793e-01 0.771534927   591
## 592 -0.57511108 19.45530 1.775737e-01 6.734933e-01 0.736027475   592
## 593 -2.17875091 19.90822 2.347441e+00 1.255817e-01 0.195962657   593
## 594  4.95289615 19.20204 7.946796e+00 4.845064e-03 0.023190978   594
## 595  6.61511531 19.14470 1.104477e+01 8.987691e-04 0.010345975   595
## 596  7.51835969 19.68731 1.325287e+01 2.762060e-04 0.008307757   596
## 597 -3.08394352 19.55164 5.039775e+00 2.483559e-02 0.062789098   597
## 598 -2.55653754 19.66521 2.836607e+00 9.223017e-02 0.155217594   598
## 599 -0.23982791 19.01407 3.611324e-02 8.492931e-01 0.880289247   599
## 600 -2.68744101 19.17285 4.104174e+00 4.285525e-02 0.089711720   600
## 601  1.26803531 19.37606 4.662882e-01 4.947462e-01 0.579423587   601
## 602 -1.63508631 19.71635 1.560582e+00 2.116659e-01 0.294191386   602
## 603 -1.30893908 19.30390 9.491486e-01 3.300053e-01 0.414086133   603
## 604 -2.90137854 19.75101 4.246583e+00 3.940512e-02 0.084995337   604
## 605 -0.81441622 20.06127 3.447094e-01 5.571623e-01 0.634628715   605
## 606 -4.30038584 20.08198 7.239113e+00 7.168023e-03 0.029081694   606
## 607 -0.09828688 20.18996 4.480963e-03 9.466334e-01 0.959015049   607
## 608  2.50872976 19.81263 2.775845e+00 9.578801e-02 0.160336714   608
## 609  1.09352921 19.12580 6.912996e-01 4.057809e-01 0.486252277   609
## 610 -4.71099896 19.18961 7.528336e+00 6.105394e-03 0.026952311   610
## 611  7.62459954 19.65576 1.372361e+01 2.151362e-04 0.008307757   611
## 612 -5.26581304 20.12736 9.188971e+00 2.453087e-03 0.016328359   612
## 613 -1.20553641 19.11260 8.568451e-01 3.546886e-01 0.439480517   613
## 614  7.26645030 19.43792 1.282698e+01 3.464303e-04 0.008307757   614
## 615 -4.49237939 19.57929 7.779983e+00 5.312175e-03 0.024732093   615
## 616  4.03938187 19.45260 5.637889e+00 1.763143e-02 0.050358439   616
## 617 -2.22804325 19.81385 2.748672e+00 9.742725e-02 0.161661092   617
## 618  1.94421971 20.16959 1.516179e+00 2.182836e-01 0.301422359   618
## 619 -2.02603852 19.55802 2.352857e+00 1.251465e-01 0.195882225   619
## 620 -1.05386165 19.62338 7.122876e-01 3.987447e-01 0.481415186   620
## 621  0.80539364 19.72136 3.513491e-01 5.533897e-01 0.633720496   621
## 622  6.69917269 19.19518 1.139070e+01 7.463816e-04 0.009662974   622
## 623 -1.41190434 19.22335 1.141662e+00 2.853771e-01 0.373488926   623
## 624 -4.85758002 19.39157 8.496336e+00 3.581815e-03 0.019937538   624
## 625  2.68760634 19.32903 3.009386e+00 8.287403e-02 0.143805851   625
## 626 -1.97166764 19.39768 2.123143e+00 1.451807e-01 0.221277215   626
## 627 -3.06743483 19.88127 4.335015e+00 3.741052e-02 0.083658170   627
## 628  6.73972842 19.11076 1.135730e+01 7.598817e-04 0.009662974   628
## 629 -1.41256749 19.51458 1.010065e+00 3.149587e-01 0.399917778   629
## 630 -1.77840517 19.80222 1.682547e+00 1.946742e-01 0.276437406   630
## 631  4.74994160 19.72792 7.380818e+00 6.625622e-03 0.028084725   631
## 632 -2.57397026 19.53172 3.621812e+00 5.711090e-02 0.109099741   632
## 633 -2.31776589 19.78329 2.923329e+00 8.739860e-02 0.148629951   633
## 634 -1.80835305 19.74363 1.572544e+00 2.099240e-01 0.293524644   634
## 635 -3.09827381 19.68119 5.131395e+00 2.355935e-02 0.060459533   635
## 636 -3.38154079 20.21428 4.766119e+00 2.909320e-02 0.070821150   636
## 637 -4.87741159 19.98858 8.050691e+00 4.575510e-03 0.022533725   637
## 638  7.88824768 20.14949 1.242570e+01 4.290391e-04 0.008307757   638
## 639  0.26347111 20.25665 3.633165e-02 8.488436e-01 0.880289247   639
## 640  2.28690978 20.48208 2.020170e+00 1.553133e-01 0.232770965   640
## 641  1.03633066 19.51768 4.424502e-01 5.059872e-01 0.588132468   641
## 642 -5.97824860 19.78761 1.160052e+01 6.669650e-04 0.009165390   642
## 643 -4.37396950 19.64529 8.389857e+00 3.797237e-03 0.020084144   643
## 644 -3.96279569 19.49877 6.991261e+00 8.228411e-03 0.031020382   644
## 645  1.59652266 19.67417 1.379456e+00 2.402759e-01 0.326499247   645
## 646 -1.22355162 19.67216 8.973279e-01 3.435664e-01 0.427951108   646
## 647 -1.31824723 19.73967 1.016701e+00 3.133744e-01 0.398499996   647
## 648  0.30251634 19.22092 5.404731e-02 8.161786e-01 0.853232076   648
## 649  3.79796590 19.75002 5.560452e+00 1.842686e-02 0.051643691   649
## 650  3.10772715 19.61388 3.318398e+00 6.859619e-02 0.124880241   650
## 651 -6.86759942 19.52087 1.314868e+01 2.919335e-04 0.008307757   651
## 652  1.71950148 20.56803 1.117457e+00 2.905417e-01 0.377926008   652
## 653 -3.36917120 19.14360 5.714801e+00 1.687645e-02 0.049130810   653
## 654  2.03272467 19.58641 1.794610e+00 1.804548e-01 0.261031441   654
## 655 -2.33739994 19.42641 3.059768e+00 8.034315e-02 0.140271241   655
## 656  4.04310147 20.05540 5.608143e+00 1.793269e-02 0.050928853   656
## 657  1.20855611 20.22819 7.054655e-01 4.010122e-01 0.482573985   657
## 658 -4.60542447 19.44217 8.365005e+00 3.849386e-03 0.020120719   658
## 659  3.39098800 19.20349 4.319083e+00 3.776189e-02 0.083764091   659
## 660 -2.86534717 19.86116 4.008540e+00 4.534945e-02 0.093327851   660
## 661  1.55611820 19.32411 1.295502e+00 2.551165e-01 0.344467988   661
## 662  0.90683273 19.10440 3.483791e-01 5.550712e-01 0.633941947   662
## 663  2.83906053 19.36933 2.927934e+00 8.714983e-02 0.148575625   663
## 664 -1.73580014 19.21107 1.801163e+00 1.796614e-01 0.260325687   664
## 665  0.14290723 19.97865 1.037163e-02 9.188887e-01 0.943244788   665
## 666  4.66673591 19.89957 6.574787e+00 1.038584e-02 0.036414533   666
## 667 -1.34677464 19.88064 7.104939e-01 3.993391e-01 0.481415186   667
## 668 -5.42277715 19.41053 1.061691e+01 1.131581e-03 0.011516262   668
## 669  0.63987583 19.57877 2.049457e-01 6.507869e-01 0.721027861   669
## 670 -2.91125684 19.38524 4.315061e+00 3.785114e-02 0.083764091   670
## 671 -1.04774859 19.61138 6.558842e-01 4.180733e-01 0.496790030   671
## 672  7.50466155 19.52154 1.336320e+01 2.604818e-04 0.008307757   672
## 673 -1.39270138 19.87170 1.007070e+00 3.156769e-01 0.400233267   673
## 674 -2.49367936 19.81222 3.155059e+00 7.578116e-02 0.133678536   674
## 675 -5.09257601 19.58823 1.008358e+01 1.509309e-03 0.012989204   675
## 676 -0.06802293 19.46769 2.312659e-03 9.616472e-01 0.970762384   676
## 677 -1.18529317 20.07743 7.496981e-01 3.866328e-01 0.469246614   677
## 678 -0.60384276 19.60901 2.074411e-01 6.488095e-01 0.719773019   678
## 679 -3.47455050 19.89065 5.842075e+00 1.569954e-02 0.047432662   679
## 680  3.13542752 19.28869 4.234253e+00 3.969195e-02 0.085397831   680
## 681  4.55002470 19.59227 7.218691e+00 7.249855e-03 0.029250374   681
## 682 -3.22680860 20.15826 4.511331e+00 3.374286e-02 0.078147492   682
## 683  5.16643850 19.90148 8.004186e+00 4.694224e-03 0.022854163   683
## 684 -2.51864917 19.76249 3.356311e+00 6.703510e-02 0.123090321   684
## 685  6.86510020 19.24169 1.197030e+01 5.471869e-04 0.008805029   685
## 686 -0.90964497 19.54560 5.177950e-01 4.718331e-01 0.555251147   686
## 687  3.46635096 19.16933 5.165120e+00 2.310687e-02 0.060459533   687
## 688 -4.78094744 20.12329 8.379532e+00 3.818816e-03 0.020084144   688
## 689  1.60868808 20.20364 9.470308e-01 3.305453e-01 0.414153808   689
## 690 -2.75944650 19.96827 3.773905e+00 5.214023e-02 0.102832119   690
## 691 -3.16762531 19.60928 4.656122e+00 3.101265e-02 0.074430369   691
## 692  3.08536107 19.22210 4.465871e+00 3.465095e-02 0.079065207   692
## 693 -3.11511280 20.13026 4.436411e+00 3.525306e-02 0.079670040   693
## 694 -4.64000779 19.31901 7.599318e+00 5.870128e-03 0.026048693   694
## 695  4.53989900 19.70916 7.033841e+00 8.035428e-03 0.030799946   695
## 696 -2.21069114 19.83687 2.609896e+00 1.062913e-01 0.172495559   696
## 697  5.57918977 19.64426 9.530164e+00 2.037472e-03 0.014767399   697
## 698  4.78756571 19.50163 7.701630e+00 5.547191e-03 0.025273832   698
## 699  3.23569059 19.68975 3.635793e+00 5.663367e-02 0.108675425   699
## 700  0.71826901 19.61258 3.110949e-01 5.770457e-01 0.653780513   700
## 701 -3.54272480 19.22899 6.194623e+00 1.286128e-02 0.042472135   701
## 702 -5.25171525 20.01025 7.449470e+00 6.378108e-03 0.027725246   702
## 703  0.73877014 20.01825 2.544271e-01 6.140075e-01 0.685628317   703
## 704  7.16913583 19.37833 1.279869e+01 3.516885e-04 0.008307757   704
## 705  2.30148378 19.87003 1.460602e+00 2.269182e-01 0.310826912   705
## 706 -2.83539430 19.35938 4.269508e+00 3.887756e-02 0.084499188   706
## 707  5.04094135 19.15533 8.214761e+00 4.180568e-03 0.021456890   707
## 708 -3.27570903 19.78029 4.950028e+00 2.615548e-02 0.065159278   708
## 709 -2.86401002 19.34955 4.276137e+00 3.872639e-02 0.084406132   709
## 710  3.28613862 19.46131 4.509876e+00 3.377153e-02 0.078147492   710
## 711  5.02942996 19.33844 8.172941e+00 4.277813e-03 0.021683653   711
## 712 -4.51671216 20.02924 8.110152e+00 4.428165e-03 0.021934865   712
## 713 -4.39704539 20.17330 5.814658e+00 1.594571e-02 0.047669269   713
## 714 -2.89863507 20.36837 3.889774e+00 4.866170e-02 0.098239843   714
## 715 -0.01489751 19.94726 1.039249e-04 9.918668e-01 0.991866818   715
## 716  5.36056107 19.54764 8.902957e+00 2.867297e-03 0.017962775   716
## 717  2.66453287 19.44594 3.549260e+00 5.965709e-02 0.112700305   717
## 718 -2.82612498 19.58786 3.837859e+00 5.018888e-02 0.099908707   718
## 719  1.71633321 19.28768 1.553886e+00 2.126484e-01 0.295075576   719
## 720 -2.53241389 19.53071 3.531624e+00 6.029404e-02 0.113400706   720
## 721 -2.18665343 20.13235 2.482784e+00 1.151912e-01 0.185175245   721
## 722  1.93835120 19.88624 1.787814e+00 1.812820e-01 0.261783518   722
## 723 -8.28969368 20.48692 1.524090e+01 9.646327e-05 0.008307757   723
## 724 -3.21264415 19.55607 5.141425e+00 2.342382e-02 0.060459533   724
## 725 -7.69183397 19.90302 1.506308e+01 1.059457e-04 0.008307757   725
## 726  2.82909124 20.15238 2.616329e+00 1.058614e-01 0.172413602   726
## 727 -5.55771134 19.48099 1.088519e+01 9.793244e-04 0.010905501   727
## 728 -3.78894765 19.32461 6.312930e+00 1.203188e-02 0.040518442   728
## 729  7.05689446 19.30243 1.254749e+01 4.020574e-04 0.008307757   729
## 730 -6.86808771 19.21871 1.331909e+01 2.666578e-04 0.008307757   730
## 731  3.82738905 19.34037 5.077927e+00 2.429564e-02 0.061650942   731
## 732 -3.35576754 19.65364 5.461372e+00 1.949901e-02 0.053938810   732
## 733 -6.01343852 19.97547 1.109558e+01 8.745532e-04 0.010340715   733
## 734  1.66218071 19.32247 1.288199e+00 2.564600e-01 0.345187817   734
## 735 -7.12961006 19.74130 1.378713e+01 2.080111e-04 0.008307757   735
## 736  2.41432171 19.95574 2.725346e+00 9.885887e-02 0.162916352   736
## 737  2.05289867 19.39291 1.822705e+00 1.770813e-01 0.257024273   737
## 738  5.43380748 19.48607 9.122110e+00 2.544124e-03 0.016673795   738
## 739 -5.27055633 19.32314 1.026310e+01 1.369680e-03 0.012853905   739
## 740 -0.85455829 19.58226 4.634485e-01 4.960629e-01 0.579423587   740
## 741 -4.85128642 20.38252 7.762569e+00 5.363514e-03 0.024835400   741
## 742  4.21092538 20.15194 6.254136e+00 1.243695e-02 0.041391738   742
## 743 -2.36196283 19.67494 3.155029e+00 7.578255e-02 0.133678536   743
## 744  5.28753969 19.86726 8.118004e+00 4.409071e-03 0.021934865   744
## 745  2.32370742 19.35334 2.637622e+00 1.044521e-01 0.170485060   745
## 746 -5.22722420 19.78045 9.507621e+00 2.062583e-03 0.014767399   746
## 747  2.72019666 19.81870 1.998289e+00 1.575683e-01 0.234699574   747
## 748 -2.10727824 19.82771 2.311887e+00 1.284809e-01 0.199754959   748
## 749  6.07369031 19.88190 9.871460e+00 1.693022e-03 0.013425612   749
## 750 -5.19976811 20.22434 6.268473e+00 1.233690e-02 0.041382054   750
## 751 -2.49335284 19.91736 3.308945e+00 6.899145e-02 0.125193481   751
## 752 -5.05144944 19.71587 9.666422e+00 1.892153e-03 0.014534221   752
## 753 -3.43750401 19.61122 5.520562e+00 1.885099e-02 0.052395807   753
## 754 -3.15232063 19.71211 5.147976e+00 2.333575e-02 0.060459533   754
## 755  2.33641127 19.59038 2.962473e+00 8.530825e-02 0.146338129   755
## 756 -7.85655241 19.99539 1.502785e+01 1.079329e-04 0.008307757   756
## 757 -1.45449325 19.25259 1.038259e+00 3.082990e-01 0.394993627   757
## 758 -6.83257853 19.26734 1.365534e+01 2.230683e-04 0.008307757   758
## 759 -4.28617501 20.01207 7.381949e+00 6.621467e-03 0.028084725   759
## 760 -3.18579679 20.21146 4.138757e+00 4.198920e-02 0.088332829   760
## 761  7.50877359 19.39198 1.338141e+01 2.579750e-04 0.008307757   761
## 762 -1.23688255 19.78635 8.388259e-01 3.597976e-01 0.442987780   762
## 763  2.68748088 19.27235 3.506236e+00 6.122377e-02 0.114643185   763
## 764 -2.31716811 19.55031 2.366542e+00 1.240542e-01 0.195368230   764
## 765 -6.76086455 19.82233 1.162094e+01 6.597075e-04 0.009165390   765
## 766 -0.37863319 19.52638 8.052742e-02 7.766012e-01 0.819532393   766
## 767 -2.39411116 19.75273 2.961418e+00 8.536391e-02 0.146338129   767
## 768  3.37115476 19.47269 4.223340e+00 3.994766e-02 0.085731497   768
## 769  0.79524085 19.95391 2.975587e-01 5.854527e-01 0.659795878   769
## 770 -2.43613125 19.36003 3.160335e+00 7.553686e-02 0.133678536   770
## 771 -5.58561797 20.29128 9.941971e+00 1.629566e-03 0.013425612   771
## 772  2.00997289 19.21031 2.244405e+00 1.341911e-01 0.207120971   772
## 773 -5.24107755 19.67106 1.021855e+01 1.403067e-03 0.012853905   773
## 774 -4.74987018 19.95826 8.163070e+00 4.301100e-03 0.021683653   774
## 775  1.53711881 19.34325 1.231717e+00 2.671517e-01 0.355090843   775
## 776 -4.65334289 19.97686 7.622152e+00 5.796428e-03 0.025992403   776
## 777 -3.06176423 19.61092 4.793517e+00 2.863463e-02 0.069904587   777
## 778 -3.84605296 19.69494 6.509650e+00 1.077237e-02 0.037309195   778
## 779 -0.06282833 20.01157 1.817027e-03 9.660017e-01 0.972852763   779
## 780  7.28522927 19.36478 1.309850e+01 2.998269e-04 0.008307757   780
## 781 -2.69626250 19.71952 3.768491e+00 5.230907e-02 0.102926854   781
## 782 -3.13479195 19.82583 4.342836e+00 3.723929e-02 0.083494410   782
## 783  0.72863098 19.73678 1.989509e-01 6.555973e-01 0.724364071   783
## 784 -4.01665905 19.08369 7.401992e+00 6.548253e-03 0.028035736   784
## 785 -4.40194915 20.17313 7.703605e+00 5.541136e-03 0.025273832   785
## 786 -0.35791548 19.83451 7.007553e-02 7.912428e-01 0.830220321   786
## 787 -5.45672378 19.51325 9.545109e+00 2.020995e-03 0.014767399   787
## 788 -3.74868080 19.69937 6.766849e+00 9.326825e-03 0.033750189   788
## 789 -1.50909624 19.72037 1.164298e+00 2.806524e-01 0.369576287   789
## 790 -4.43143111 19.95748 7.808265e+00 5.229860e-03 0.024482641   790
## 791 -1.82373633 20.07912 1.667858e+00 1.966332e-01 0.278051921   791
## 792  4.70130813 19.73602 7.176157e+00 7.423363e-03 0.029281043   792
## 793 -6.31235065 18.86985 1.086211e+01 9.915690e-04 0.010905501   793
## 794  3.50069723 19.35235 4.180367e+00 4.097153e-02 0.087358823   794
## 795 -0.35356342 19.54666 7.742973e-02 7.808281e-01 0.821315519   795
## 796 -2.35256503 19.35036 3.231211e+00 7.233575e-02 0.129474914   796
## 797 -5.30076970 19.65760 1.077360e+01 1.039963e-03 0.011021292   797
## 798  0.85037017 19.82193 3.776645e-01 5.388968e-01 0.619622281   798
## 799  0.58185559 19.53700 2.174220e-01 6.410411e-01 0.712082127   799
## 800 -2.92253966 19.07821 4.564738e+00 3.270770e-02 0.076557577   800
## 801 -1.89339362 19.09679 1.666688e+00 1.967903e-01 0.278051921   801
## 802  7.02082614 19.11887 1.203223e+01 5.293640e-04 0.008805029   802
## 803  3.80790185 19.86694 5.796924e+00 1.610705e-02 0.047816065   803
## 804 -6.82985773 19.84604 1.173327e+01 6.211826e-04 0.009124959   804
## 805 -0.78256912 19.69730 3.456763e-01 5.566099e-01 0.634628715   805
## 806  5.23847294 19.40489 8.681775e+00 3.235864e-03 0.018883263   806
## 807 -0.18143916 19.95499 1.921079e-02 8.897721e-01 0.917779411   807
## 808 -3.03999913 19.65066 4.711832e+00 3.002453e-02 0.072467140   808
## 809 -2.31252674 19.58622 2.841925e+00 9.192569e-02 0.155090474   809
## 810 -1.89327677 19.67648 2.114228e+00 1.460277e-01 0.222170759   810
## 811 -5.06523313 19.66632 9.330396e+00 2.271266e-03 0.015861626   811
## 812 -3.09542312 19.70659 4.318600e+00 3.777261e-02 0.083764091   812
## 813 -1.61962081 19.36752 1.512838e+00 2.187915e-01 0.301634878   813
## 814  7.00733960 19.47211 1.196844e+01 5.477307e-04 0.008805029   814
## 815  1.65658930 20.21682 1.244956e+00 2.645967e-01 0.352244333   815
## 816  1.22947472 19.66418 9.025386e-01 3.421693e-01 0.427460774   816
## 817 -3.87511435 19.21609 5.706781e+00 1.695359e-02 0.049130810   817
## 818  2.65376367 20.27989 2.708775e+00 9.988984e-02 0.163981008   818
## 819 -2.81875417 20.49003 3.576629e+00 5.868267e-02 0.111105855   819
## 820 -6.30414800 18.90583 1.014541e+01 1.459665e-03 0.012944125   820
## 821 -3.83647279 19.66050 6.757342e+00 9.376549e-03 0.033750189   821
## 822 -1.43550068 19.81014 9.949767e-01 3.185999e-01 0.403338931   822
## 823 -4.87239529 19.53709 8.906028e+00 2.862493e-03 0.017962775   823
## 824 -2.10298078 19.48952 2.456160e+00 1.171575e-01 0.187497542   824
## 825 -3.38014637 19.64325 4.253886e+00 3.923623e-02 0.084995337   825
## 826  3.40723111 19.76253 3.602544e+00 5.777564e-02 0.109632179   826
## 827 -4.58883434 19.39038 8.655023e+00 3.283593e-03 0.018902846   827
## 828  1.06569688 19.78179 2.743817e-01 6.004416e-01 0.674013506   828
## 829 -5.09016996 19.40994 9.001071e+00 2.717751e-03 0.017409953   829
## 830  1.86377046 19.46659 1.962340e+00 1.613543e-01 0.238740754   830
## 831  1.92797733 19.59220 2.096964e+00 1.476839e-01 0.223890948   831
## 832  4.70389521 19.23971 7.839794e+00 5.139628e-03 0.024226017   832
## 833  0.08061805 20.00674 3.240463e-03 9.546082e-01 0.965945613   833
## 834 -0.69184897 19.35681 1.980242e-01 6.563487e-01 0.724364071   834
## 835 -0.40082899 20.14681 8.326705e-02 7.729354e-01 0.819504957   835
## 836 -3.54175694 20.14401 5.437933e+00 1.976194e-02 0.054489238   836
## 837  4.01398909 19.70201 6.165038e+00 1.307773e-02 0.042740882   837
## 838 -3.12327933 20.42563 4.283837e+00 3.855158e-02 0.084406132   838
## 839  0.01571321 18.91450 1.412747e-04 9.905173e-01 0.991681285   839
## 840  4.27813915 19.27908 6.376309e+00 1.161037e-02 0.039410509   840
## 841 -4.47636101 19.75139 8.334921e+00 3.913493e-03 0.020331072   841
## 842 -1.77269388 19.41063 1.886102e+00 1.697325e-01 0.249331226   842
## 843  4.15637295 19.49903 6.149421e+00 1.319349e-02 0.042740882   843
## 844 -3.49022687 19.53949 6.015228e+00 1.423274e-02 0.044581953   844
## 845 -0.65534116 20.60643 1.671288e-01 6.827019e-01 0.744765703   845
## 846 -0.14292204 19.78575 9.302809e-03 9.231680e-01 0.946497158   846
## 847  0.68296378 19.55088 1.822779e-01 6.694498e-01 0.735014516   847
## 848 -1.64157757 19.29038 1.561165e+00 2.115805e-01 0.294191386   848
## 849 -0.91366212 19.94915 4.728495e-01 4.917260e-01 0.577862776   849
## 850  1.02746797 19.45931 4.181122e-01 5.179241e-01 0.599553386   850
## 851 -0.11717416 19.48841 7.782362e-03 9.297088e-01 0.949774492   851
## 852  3.30427867 19.22089 4.094057e+00 4.311213e-02 0.090028278   852
##      SpatialFDR                       celltype celltype_fraction
## 1   0.046943396                       Epiblast         1.0000000
## 2   0.028662204               Nascent mesoderm         0.9777778
## 3   0.137210177                       Epiblast         0.8705882
## 4   0.008542646            Pharyngeal mesoderm         0.4791667
## 5   0.446838288                       Epiblast         0.9780220
## 6   0.264747245                   ExE ectoderm         1.0000000
## 7   0.007518813                       Epiblast         1.0000000
## 8   0.012256985                   ExE ectoderm         1.0000000
## 9   0.848278188                   ExE endoderm         1.0000000
## 10  0.011959481                   ExE endoderm         1.0000000
## 11  0.078059081                   ExE ectoderm         1.0000000
## 12  0.027474822                       Epiblast         1.0000000
## 13  0.776603496                   ExE ectoderm         1.0000000
## 14  0.078498387                       Epiblast         0.9615385
## 15  0.007518813               Primitive Streak         0.6153846
## 16  0.019351974           Rostral neurectoderm         0.7551020
## 17  0.119456004            Blood progenitors 1         0.8947368
## 18  0.009950537                       Epiblast         1.0000000
## 19  0.819406693                   ExE ectoderm         1.0000000
## 20  0.132962868      Anterior Primitive Streak         0.6410256
## 21  0.055288990               Primitive Streak         1.0000000
## 22  0.193239837               Primitive Streak         0.9078947
## 23  0.270364918              Visceral endoderm         1.0000000
## 24  0.033859162               Nascent mesoderm         0.9875000
## 25  0.034156038                     Mesenchyme         1.0000000
## 26  0.353956779                   ExE ectoderm         1.0000000
## 27  0.089869246                   ExE endoderm         1.0000000
## 28  0.049597312                       Epiblast         0.6632653
## 29  0.065102052               Primitive Streak         0.9538462
## 30  0.043655593                   ExE ectoderm         1.0000000
## 31  0.499501181                       Epiblast         0.9610390
## 32  0.049597312                       Epiblast         1.0000000
## 33  0.084296490           Rostral neurectoderm         0.5540541
## 34  0.019351974                       Epiblast         0.9350649
## 35  0.979765204                   ExE endoderm         1.0000000
## 36  0.021348394                       Epiblast         0.9693878
## 37  0.019478789               Primitive Streak         0.9125000
## 38  0.077531195                       Epiblast         0.9900000
## 39  0.031147997                       Epiblast         0.8750000
## 40  0.235623688                   ExE ectoderm         1.0000000
## 41  0.069340777      Anterior Primitive Streak         1.0000000
## 42  0.892168706                   ExE ectoderm         1.0000000
## 43  0.225363623                   ExE ectoderm         1.0000000
## 44  0.577973598                   ExE endoderm         1.0000000
## 45  0.225227373               Primitive Streak         1.0000000
## 46  0.790944452              Visceral endoderm         0.9750000
## 47  0.118561485                   ExE endoderm         1.0000000
## 48  0.108179776              Visceral endoderm         0.9864865
## 49  0.007823087                       Epiblast         0.9259259
## 50  0.107070166                  Def. endoderm         0.7317073
## 51  0.290535257                   ExE ectoderm         1.0000000
## 52  0.968775753                   ExE endoderm         1.0000000
## 53  0.007518813                 Mixed mesoderm         0.6296296
## 54  0.104559541               Surface ectoderm         0.8205128
## 55  0.060602026                   ExE endoderm         1.0000000
## 56  0.522117265                   ExE ectoderm         1.0000000
## 57  0.124337467              Visceral endoderm         0.6164384
## 58  0.384972086                   ExE endoderm         1.0000000
## 59  0.290535257                       Epiblast         0.7839506
## 60  0.103685383                   ExE ectoderm         1.0000000
## 61  0.042107569                       Epiblast         0.9931034
## 62  0.875797585                       Epiblast         0.8989899
## 63  0.027422265               Primitive Streak         0.9661017
## 64  0.007518813                       Epiblast         1.0000000
## 65  0.118561485                       Epiblast         0.9361702
## 66  0.015231082                       Epiblast         0.5423729
## 67  0.046963480                   ExE ectoderm         1.0000000
## 68  0.485150702                 Mixed mesoderm         0.8620690
## 69  0.007518813                       Epiblast         1.0000000
## 70  0.024236085      Anterior Primitive Streak         0.6582278
## 71  0.007518813                       Epiblast         1.0000000
## 72  0.044305343                       Epiblast         0.9468085
## 73  0.032721155                     Mesenchyme         0.7758621
## 74  0.731609960                   ExE endoderm         1.0000000
## 75  0.809974625 Haematoendothelial progenitors         0.9295775
## 76  0.046697109                       Epiblast         0.9746835
## 77  0.018874915                       Epiblast         1.0000000
## 78  0.017617633                       Epiblast         1.0000000
## 79  0.950055376              Parietal endoderm         1.0000000
## 80  0.009351811                       Epiblast         0.9871795
## 81  0.059432172                       Epiblast         0.9736842
## 82  0.012296241                       Epiblast         0.9024390
## 83  0.678623078                   ExE endoderm         1.0000000
## 84  0.027653832                       Epiblast         1.0000000
## 85  0.059707805               Primitive Streak         1.0000000
## 86  0.215418560                   ExE ectoderm         1.0000000
## 87  0.041002951               Primitive Streak         0.9777778
## 88  0.032395090               Nascent mesoderm         0.9500000
## 89  0.042133733                     Mesenchyme         0.9873418
## 90  0.479941168                   ExE endoderm         0.7538462
## 91  0.073123881                       Epiblast         1.0000000
## 92  0.254655772                       Epiblast         1.0000000
## 93  0.371642717               Nascent mesoderm         0.9890110
## 94  0.007518813                       Epiblast         0.5147059
## 95  0.402556097              Visceral endoderm         1.0000000
## 96  0.159339502                   ExE endoderm         1.0000000
## 97  0.141402493                       Epiblast         0.7222222
## 98  0.050763040                       Epiblast         1.0000000
## 99  0.765041457                   ExE endoderm         1.0000000
## 100 0.382298621               Nascent mesoderm         0.4285714
## 101 0.437925008                   ExE endoderm         1.0000000
## 102 0.275578686                       Epiblast         1.0000000
## 103 0.577973598                   ExE endoderm         1.0000000
## 104 0.122938981            Blood progenitors 2         0.9800000
## 105 0.019351974               Primitive Streak         0.7500000
## 106 0.107070166               Primitive Streak         0.8823529
## 107 0.275578686                       Epiblast         1.0000000
## 108 0.021073172                       Epiblast         0.9846154
## 109 0.044649492                   ExE ectoderm         1.0000000
## 110 0.186771538               Nascent mesoderm         0.8943089
## 111 0.007518813           Rostral neurectoderm         0.9104478
## 112 0.018372953                       Epiblast         1.0000000
## 113 0.112865194               Nascent mesoderm         1.0000000
## 114 0.293659647 Haematoendothelial progenitors         0.8947368
## 115 0.044462831 Haematoendothelial progenitors         0.9047619
## 116 0.042107569                       Epiblast         1.0000000
## 117 0.687409270                   ExE ectoderm         1.0000000
## 118 0.035978139                       Epiblast         1.0000000
## 119 0.059707805                       Epiblast         1.0000000
## 120 0.100729032               Nascent mesoderm         1.0000000
## 121 0.027371804                       Epiblast         0.5000000
## 122 0.051500490               Nascent mesoderm         0.9890110
## 123 0.012356317                Caudal epiblast         0.5875000
## 124 0.654066888               Nascent mesoderm         0.9054054
## 125 0.009271338                       Epiblast         0.9117647
## 126 0.388600359                   ExE endoderm         1.0000000
## 127 0.025338981                       Epiblast         0.8764045
## 128 0.395118953                 Mixed mesoderm         0.7846154
## 129 0.011807459                       Epiblast         0.8484848
## 130 0.030920570               Primitive Streak         0.9864865
## 131 0.087121977               Nascent mesoderm         0.8333333
## 132 0.017387992                       Epiblast         0.8448276
## 133 0.087625468                       Epiblast         1.0000000
## 134 0.735583240                   ExE ectoderm         1.0000000
## 135 0.930601549                       Epiblast         0.7750000
## 136 0.054796297                       Epiblast         0.9830508
## 137 0.030229397                     Mesenchyme         0.8636364
## 138 0.185615043            Blood progenitors 2         0.9882353
## 139 0.015354574                       Epiblast         1.0000000
## 140 0.194294092               Primitive Streak         1.0000000
## 141 0.059707805                   ExE endoderm         1.0000000
## 142 0.338005065               Primitive Streak         0.9918033
## 143 0.008542646              Paraxial mesoderm         0.5000000
## 144 0.050396769                       Epiblast         0.9837398
## 145 0.030229397                       Epiblast         0.6250000
## 146 0.817230680                       Epiblast         0.5000000
## 147 0.889451264                       Epiblast         0.9940828
## 148 0.180242526                   ExE ectoderm         1.0000000
## 149 0.547380306               Nascent mesoderm         0.9666667
## 150 0.118561485                       Epiblast         0.8750000
## 151 0.228498899               Nascent mesoderm         0.9911504
## 152 0.017554614               Primitive Streak         0.9532164
## 153 0.063318872                     Mesenchyme         1.0000000
## 154 0.147790816                   ExE ectoderm         1.0000000
## 155 0.678623078                   ExE endoderm         1.0000000
## 156 0.074875298              Visceral endoderm         0.9655172
## 157 0.772894754                       Epiblast         1.0000000
## 158 0.697001164               Nascent mesoderm         0.9313725
## 159 0.078503869               Primitive Streak         0.9743590
## 160 0.092435664           Rostral neurectoderm         0.6506024
## 161 0.007518813                       Epiblast         0.8958333
## 162 0.973838559      Anterior Primitive Streak         0.6184211
## 163 0.438292382               Nascent mesoderm         0.5555556
## 164 0.087148303               Primitive Streak         0.9882353
## 165 0.370946311                   ExE endoderm         1.0000000
## 166 0.030229397               Primitive Streak         1.0000000
## 167 0.063318872                  Def. endoderm         0.7866667
## 168 0.731609960                   ExE endoderm         1.0000000
## 169 0.101027750               Primitive Streak         1.0000000
## 170 0.276423162               Nascent mesoderm         0.9578947
## 171 0.044945864                   ExE endoderm         1.0000000
## 172 0.007518813                       Epiblast         1.0000000
## 173 0.028662204                       Epiblast         1.0000000
## 174 0.018051416                     Mesenchyme         0.9508197
## 175 0.097228643               Nascent mesoderm         0.7857143
## 176 0.812415754                   ExE ectoderm         1.0000000
## 177 0.590754438                   ExE ectoderm         1.0000000
## 178 0.586720816               Primitive Streak         0.8181818
## 179 0.193239837                   ExE ectoderm         1.0000000
## 180 0.094432660            Blood progenitors 1         0.6511628
## 181 0.048417513                       Epiblast         1.0000000
## 182 0.602440369                   ExE endoderm         1.0000000
## 183 0.065102052                   ExE endoderm         1.0000000
## 184 0.094470266               Primitive Streak         0.9741379
## 185 0.030460100                            Gut         0.8000000
## 186 0.034566585            Blood progenitors 1         0.8039216
## 187 0.017658633           Rostral neurectoderm         0.3833333
## 188 0.077531195                       Epiblast         1.0000000
## 189 0.093950168                   ExE ectoderm         1.0000000
## 190 0.728763605                       Epiblast         0.7704918
## 191 0.015342990                   ExE endoderm         1.0000000
## 192 0.160227177                   ExE ectoderm         1.0000000
## 193 0.291203114               Primitive Streak         1.0000000
## 194 0.007518813                Caudal epiblast         0.7761194
## 195 0.011409289                       Epiblast         1.0000000
## 196 0.127461888               Primitive Streak         0.6702128
## 197 0.479941168                   ExE ectoderm         1.0000000
## 198 0.107654384                       Epiblast         0.9939394
## 199 0.016223865               Primitive Streak         0.6777778
## 200 0.021989902           Rostral neurectoderm         0.6862745
## 201 0.009351811                       Epiblast         0.9430894
## 202 0.157025916            Blood progenitors 2         1.0000000
## 203 0.149781121      Anterior Primitive Streak         0.7123288
## 204 0.306845272                 Mixed mesoderm         0.4750000
## 205 0.008501284                   ExE endoderm         1.0000000
## 206 0.230277340               Primitive Streak         0.9696970
## 207 0.819819707                   ExE ectoderm         1.0000000
## 208 0.253573267      Anterior Primitive Streak         0.6941176
## 209 0.019351974               Primitive Streak         0.7547170
## 210 0.007518813           Rostral neurectoderm         0.8648649
## 211 0.007518813               Primitive Streak         0.8840580
## 212 0.270364918                   ExE ectoderm         1.0000000
## 213 0.437925008                 Mixed mesoderm         0.6896552
## 214 0.038213098                       Epiblast         0.9918699
## 215 0.343794715                   ExE endoderm         1.0000000
## 216 0.097228643                       Epiblast         0.9908257
## 217 0.032214679                   ExE endoderm         1.0000000
## 218 0.143865203                       Epiblast         1.0000000
## 219 0.353758684               Nascent mesoderm         0.7567568
## 220 0.028374755               Primitive Streak         1.0000000
## 221 0.437925008                       Epiblast         1.0000000
## 222 0.019351974                Caudal epiblast         0.7500000
## 223 0.042507146                       Epiblast         0.9907407
## 224 0.404454467                   ExE endoderm         1.0000000
## 225 0.750925047                   ExE endoderm         1.0000000
## 226 0.748995383                   ExE ectoderm         1.0000000
## 227 0.026863267      Anterior Primitive Streak         0.6363636
## 228 0.083846973               Nascent mesoderm         0.5142857
## 229 0.979765204                       Epiblast         1.0000000
## 230 0.090126170               Primitive Streak         0.6666667
## 231 0.396961198               Nascent mesoderm         0.6222222
## 232 0.100451859            Blood progenitors 1         0.8030303
## 233 0.493780186               Primitive Streak         0.9718310
## 234 0.010456710                       Epiblast         0.9926471
## 235 0.016046162           Rostral neurectoderm         0.7027027
## 236 0.614187685                   ExE ectoderm         1.0000000
## 237 0.007518813               Primitive Streak         0.7187500
## 238 0.009384075                       Epiblast         0.9636364
## 239 0.143865203                   ExE ectoderm         1.0000000
## 240 0.984360261                   ExE ectoderm         1.0000000
## 241 0.008794174                       Epiblast         1.0000000
## 242 0.107070166                       Epiblast         0.9615385
## 243 0.127343762               Primitive Streak         0.7096774
## 244 0.027111029                            Gut         0.8750000
## 245 0.020870573                      Notochord         0.5600000
## 246 0.014224436                       Epiblast         1.0000000
## 247 0.015691311                       Epiblast         1.0000000
## 248 0.143045418                       Epiblast         0.9921875
## 249 0.316003147                   ExE ectoderm         1.0000000
## 250 0.059600245                       Epiblast         0.9934211
## 251 0.083846973                   ExE ectoderm         1.0000000
## 252 0.097228643                   ExE ectoderm         1.0000000
## 253 0.097004881               Nascent mesoderm         0.8490566
## 254 0.049596294                   ExE endoderm         1.0000000
## 255 0.210837132                       Epiblast         0.9761905
## 256 0.028662204                 Mixed mesoderm         0.8470588
## 257 0.012753616                       Epiblast         1.0000000
## 258 0.122332218               Primitive Streak         0.5277778
## 259 0.819406693                   ExE endoderm         1.0000000
## 260 0.061009174                   ExE endoderm         1.0000000
## 261 0.275131086                  Def. endoderm         0.9259259
## 262 0.680422572              Visceral endoderm         0.6923077
## 263 0.044602752               Primitive Streak         1.0000000
## 264 0.015490378                       Epiblast         1.0000000
## 265 0.030943421               Nascent mesoderm         1.0000000
## 266 0.494911846                   ExE ectoderm         1.0000000
## 267 0.663736596               Primitive Streak         0.9444444
## 268 0.123170585                   ExE endoderm         1.0000000
## 269 0.013847381                            Gut         0.8676471
## 270 0.011103489               Primitive Streak         0.7956989
## 271 0.464105596               Nascent mesoderm         0.6200000
## 272 0.016766954                       Epiblast         1.0000000
## 273 0.017542373                       Epiblast         0.9113924
## 274 0.635456761                   ExE ectoderm         1.0000000
## 275 0.007518813                 Mixed mesoderm         0.5147059
## 276 0.189319968                   ExE endoderm         1.0000000
## 277 0.007518813               Nascent mesoderm         0.9824561
## 278 0.077756959                  Def. endoderm         0.5750000
## 279 0.007518813                       Epiblast         0.8085106
## 280 0.085696480                   ExE ectoderm         1.0000000
## 281 0.014128687                     Mesenchyme         1.0000000
## 282 0.922797312                       Epiblast         0.3913043
## 283 0.018294544                       Epiblast         1.0000000
## 284 0.463054645                   ExE ectoderm         1.0000000
## 285 0.208670220               Primitive Streak         0.6904762
## 286 0.042107569                     Mesenchyme         1.0000000
## 287 0.246280657               Primitive Streak         1.0000000
## 288 0.007732912               Primitive Streak         0.9764706
## 289 0.170850625                       Epiblast         0.9342105
## 290 0.589072386                   ExE ectoderm         1.0000000
## 291 0.426303514                   ExE endoderm         1.0000000
## 292 0.152971074                   ExE ectoderm         1.0000000
## 293 0.097004881                   ExE ectoderm         1.0000000
## 294 0.300323004               Nascent mesoderm         0.6779661
## 295 0.012753616                   ExE ectoderm         1.0000000
## 296 0.268234719                   ExE ectoderm         1.0000000
## 297 0.054596878                   ExE endoderm         1.0000000
## 298 0.194251750                   ExE endoderm         1.0000000
## 299 0.186377405 Haematoendothelial progenitors         0.8923077
## 300 0.126062530                       Epiblast         0.9909910
## 301 0.028662204                       Epiblast         0.9459459
## 302 0.057046947                  Def. endoderm         0.6764706
## 303 0.948508847                   ExE endoderm         1.0000000
## 304 0.007732912                 Mixed mesoderm         0.4117647
## 305 0.127461888            Blood progenitors 1         0.7674419
## 306 0.653347655               Primitive Streak         0.7317073
## 307 0.370946311               Nascent mesoderm         0.6557377
## 308 0.044936575               Primitive Streak         0.6323529
## 309 0.014003839                      Notochord         0.7105263
## 310 0.031102077               Nascent mesoderm         1.0000000
## 311 0.097349134                       Epiblast         0.8990826
## 312 0.102760330                   ExE ectoderm         1.0000000
## 313 0.029473487                       Epiblast         0.9666667
## 314 0.800595812                   ExE ectoderm         1.0000000
## 315 0.316003147               Nascent mesoderm         0.8750000
## 316 0.577973598              Visceral endoderm         0.8448276
## 317 0.037333645              Visceral endoderm         0.9761905
## 318 0.078935983                       Epiblast         0.9789474
## 319 0.050637801               Primitive Streak         0.9402985
## 320 0.308913657               Nascent mesoderm         1.0000000
## 321 0.017387992                       Epiblast         1.0000000
## 322 0.138758394               Primitive Streak         0.9864865
## 323 0.046851761                       Epiblast         0.8907104
## 324 0.007823087               Primitive Streak         0.8636364
## 325 0.090635096                   ExE ectoderm         1.0000000
## 326 0.396961198               Nascent mesoderm         0.9425287
## 327 0.007621193           Rostral neurectoderm         0.8169014
## 328 0.076181487              Visceral endoderm         0.9824561
## 329 0.027653832                     Mesenchyme         0.9852941
## 330 0.074875298               Primitive Streak         0.8947368
## 331 0.811831827                   ExE ectoderm         1.0000000
## 332 0.008395562                            Gut         0.8888889
## 333 0.180834661                   ExE ectoderm         1.0000000
## 334 0.019351974                       Epiblast         0.7804878
## 335 0.075611417                       Epiblast         0.8823529
## 336 0.382298621               Nascent mesoderm         0.9682540
## 337 0.365768470      Anterior Primitive Streak         0.8333333
## 338 0.358934884                   ExE ectoderm         1.0000000
## 339 0.012296241                       Epiblast         0.9318182
## 340 0.097004881               Primitive Streak         1.0000000
## 341 0.892168706                   ExE endoderm         1.0000000
## 342 0.007732912            Pharyngeal mesoderm         0.5066667
## 343 0.042107569               Primitive Streak         1.0000000
## 344 0.051500490                       Epiblast         1.0000000
## 345 0.083846973               Primitive Streak         0.9482759
## 346 0.812415754               Nascent mesoderm         0.6875000
## 347 0.038095098                       Epiblast         0.9848485
## 348 0.721183528                   ExE ectoderm         1.0000000
## 349 0.030229397                       Epiblast         1.0000000
## 350 0.190635564                       Epiblast         0.9920635
## 351 0.032721155                     Mesenchyme         1.0000000
## 352 0.285010827               Nascent mesoderm         0.9534884
## 353 0.263589525                   ExE ectoderm         1.0000000
## 354 0.395172127               Nascent mesoderm         0.9787234
## 355 0.184245418                       Epiblast         0.9178082
## 356 0.327118755               Nascent mesoderm         1.0000000
## 357 0.108671209                       Epiblast         0.5465116
## 358 0.114262344               Primitive Streak         1.0000000
## 359 0.499113847                   ExE ectoderm         1.0000000
## 360 0.812415754                   ExE ectoderm         1.0000000
## 361 0.017387992                       Epiblast         1.0000000
## 362 0.824822042                   ExE ectoderm         1.0000000
## 363 0.053661914               Nascent mesoderm         0.5294118
## 364 0.165062946                   ExE ectoderm         1.0000000
## 365 0.203390180                   ExE ectoderm         1.0000000
## 366 0.044555243           Rostral neurectoderm         0.6344086
## 367 0.121183753                   ExE ectoderm         1.0000000
## 368 0.064830383               Primitive Streak         0.8135593
## 369 0.009699044                       Epiblast         1.0000000
## 370 0.232306725               Nascent mesoderm         0.6190476
## 371 0.089774884                 Mixed mesoderm         0.8051948
## 372 0.173307136                       Epiblast         1.0000000
## 373 0.956798936                   ExE endoderm         1.0000000
## 374 0.014003839                   ExE endoderm         1.0000000
## 375 0.577973598      Anterior Primitive Streak         0.6666667
## 376 0.010456710               Primitive Streak         0.9866667
## 377 0.352706988                       Epiblast         0.6666667
## 378 0.660077704                   ExE endoderm         1.0000000
## 379 0.270364918                   ExE ectoderm         1.0000000
## 380 0.074875298               Primitive Streak         0.5490196
## 381 0.128336380                   ExE ectoderm         1.0000000
## 382 0.043886432                   ExE ectoderm         1.0000000
## 383 0.426782466 Haematoendothelial progenitors         1.0000000
## 384 0.040692708                Caudal epiblast         0.4680851
## 385 0.059600245      Anterior Primitive Streak         0.8852459
## 386 0.632400811 Haematoendothelial progenitors         0.6081081
## 387 0.295800572               Primitive Streak         0.9684211
## 388 0.370406900               Primitive Streak         1.0000000
## 389 0.212679152                   ExE ectoderm         1.0000000
## 390 0.466491713                   ExE endoderm         1.0000000
## 391 0.604140395              Visceral endoderm         1.0000000
## 392 0.015099039               Primitive Streak         1.0000000
## 393 0.446114677                 Mixed mesoderm         0.8031496
## 394 0.955596716                   ExE ectoderm         1.0000000
## 395 0.270364918                       Epiblast         0.9739130
## 396 0.027939176               Primitive Streak         0.9841270
## 397 0.161488815                 Mixed mesoderm         0.5333333
## 398 0.059707805                     Mesenchyme         0.8606557
## 399 0.058593921                     Mesenchyme         1.0000000
## 400 0.059600245                 Mixed mesoderm         0.7058824
## 401 0.484686455               Nascent mesoderm         0.9591837
## 402 0.747007933               Primitive Streak         0.5454545
## 403 0.203390180                       Epiblast         0.9565217
## 404 0.013569411           Rostral neurectoderm         0.7123288
## 405 0.147790816                   ExE ectoderm         1.0000000
## 406 0.083144243 Haematoendothelial progenitors         0.6233766
## 407 0.063318872                   ExE endoderm         1.0000000
## 408 0.134926303            Blood progenitors 2         0.9864865
## 409 0.007518813           Rostral neurectoderm         0.8196721
## 410 0.132948264                   ExE ectoderm         1.0000000
## 411 0.074875298                       Epiblast         1.0000000
## 412 0.347826277                   ExE endoderm         1.0000000
## 413 0.464013013                   ExE ectoderm         1.0000000
## 414 0.011053297                       Epiblast         0.9600000
## 415 0.488900316               Primitive Streak         1.0000000
## 416 0.382298621                   ExE ectoderm         1.0000000
## 417 0.042229501                       Epiblast         0.9816514
## 418 0.458460821                 Mixed mesoderm         0.9137931
## 419 0.007518813                       Epiblast         1.0000000
## 420 0.230334974 Haematoendothelial progenitors         0.7083333
## 421 0.160062088                       Epiblast         0.5371901
## 422 0.251058667              Visceral endoderm         1.0000000
## 423 0.221569517                   ExE ectoderm         1.0000000
## 424 0.063318872 Haematoendothelial progenitors         0.8611111
## 425 0.237569287                   ExE endoderm         1.0000000
## 426 0.450751662               Primitive Streak         0.7391304
## 427 0.628856706               Primitive Streak         1.0000000
## 428 0.107070166                   ExE ectoderm         1.0000000
## 429 0.017387992                       Epiblast         0.8804348
## 430 0.074875298                   ExE ectoderm         1.0000000
## 431 0.747007933                   ExE endoderm         1.0000000
## 432 0.346043550                   ExE endoderm         1.0000000
## 433 0.097559252      Anterior Primitive Streak         0.9791667
## 434 0.012296241               Nascent mesoderm         1.0000000
## 435 0.340149034               Primitive Streak         0.6170213
## 436 0.316003147                   ExE endoderm         1.0000000
## 437 0.087238247               Nascent mesoderm         1.0000000
## 438 0.179030407                   ExE ectoderm         1.0000000
## 439 0.819406693               Nascent mesoderm         0.8529412
## 440 0.132313976                   ExE endoderm         1.0000000
## 441 0.007518813                       Epiblast         1.0000000
## 442 0.086711383                       Epiblast         1.0000000
## 443 0.048417513                   ExE ectoderm         1.0000000
## 444 0.548019409               Nascent mesoderm         0.8125000
## 445 0.236485894               Nascent mesoderm         0.9032258
## 446 0.088203607                       Epiblast         0.7243590
## 447 0.042107569               Surface ectoderm         0.5217391
## 448 0.154179368                       Epiblast         0.7634409
## 449 0.928049586                   ExE ectoderm         1.0000000
## 450 0.007732912                Caudal epiblast         0.6415094
## 451 0.010648618                      Notochord         0.6511628
## 452 0.027653832                            Gut         0.5000000
## 453 0.012256985           Rostral neurectoderm         0.8076923
## 454 0.010007434                            Gut         0.6379310
## 455 0.023625907                       Epiblast         0.8970588
## 456 0.160153667                   ExE ectoderm         1.0000000
## 457 0.027653832               Primitive Streak         0.6603774
## 458 0.375611560               Primitive Streak         0.3725490
## 459 0.402438438               Nascent mesoderm         0.7432432
## 460 0.038435166                 Mixed mesoderm         0.9111111
## 461 0.441625161                   ExE ectoderm         1.0000000
## 462 0.034566585                       Epiblast         0.9674797
## 463 0.007518813                       Epiblast         0.6119403
## 464 0.066985828                       Epiblast         1.0000000
## 465 0.022368916                 Mixed mesoderm         0.7735849
## 466 0.162621968                       Epiblast         1.0000000
## 467 0.654094253                   ExE ectoderm         1.0000000
## 468 0.008560849               Primitive Streak         0.8356164
## 469 0.418244518              Visceral endoderm         1.0000000
## 470 0.132313976 Haematoendothelial progenitors         0.7391304
## 471 0.253584534                   ExE ectoderm         1.0000000
## 472 0.607647996                   ExE ectoderm         1.0000000
## 473 0.810998597               Primitive Streak         0.9130435
## 474 0.148965068                   ExE endoderm         1.0000000
## 475 0.060172559                       Epiblast         0.8522727
## 476 0.456340917                   ExE ectoderm         1.0000000
## 477 0.848278188               Nascent mesoderm         0.7125000
## 478 0.033424077                       Epiblast         0.9912281
## 479 0.007823087            Pharyngeal mesoderm         0.4375000
## 480 0.029140488               Primitive Streak         0.8496241
## 481 0.819406693                   ExE ectoderm         1.0000000
## 482 0.348973471                   ExE ectoderm         1.0000000
## 483 0.265314368      Anterior Primitive Streak         0.9452055
## 484 0.043836438               Primitive Streak         1.0000000
## 485 0.236485894               Primitive Streak         1.0000000
## 486 0.198754272                   ExE ectoderm         1.0000000
## 487 0.407067721                       Epiblast         0.9508197
## 488 0.117271543               Primitive Streak         0.5098039
## 489 0.206181396                   ExE ectoderm         1.0000000
## 490 0.870558791                   ExE ectoderm         1.0000000
## 491 0.056015959               Primitive Streak         0.7238095
## 492 0.956967102                   ExE ectoderm         1.0000000
## 493 0.877302660                   ExE ectoderm         1.0000000
## 494 0.470218412                   ExE ectoderm         1.0000000
## 495 0.162023471               Primitive Streak         0.5662651
## 496 0.391133384                 Mixed mesoderm         0.5573770
## 497 0.049597312                   ExE endoderm         1.0000000
## 498 0.019351974               Primitive Streak         0.5740741
## 499 0.484686455               Primitive Streak         1.0000000
## 500 0.143865203                 Mixed mesoderm         0.6417910
## 501 0.012753616                       Epiblast         1.0000000
## 502 0.251058667                 Mixed mesoderm         0.8281250
## 503 0.196594781                   ExE endoderm         1.0000000
## 504 0.039750446                   ExE ectoderm         1.0000000
## 505 0.194251750                            PGC         0.6808511
## 506 0.384670051                       Epiblast         0.7676768
## 507 0.222844975               Primitive Streak         0.9830508
## 508 0.736001060                   ExE ectoderm         1.0000000
## 509 0.407067721               Nascent mesoderm         0.9506173
## 510 0.066985828               Primitive Streak         0.9710145
## 511 0.028374755                       Epiblast         0.9803922
## 512 0.059707805                       Epiblast         0.9206349
## 513 0.033171441                       Epiblast         0.9795918
## 514 0.007937614          Intermediate mesoderm         0.4259259
## 515 0.270165624               Primitive Streak         1.0000000
## 516 0.071951281               Nascent mesoderm         0.7794118
## 517 0.188370117                   ExE ectoderm         1.0000000
## 518 0.499501181                   ExE ectoderm         1.0000000
## 519 0.044799164                   ExE mesoderm         0.6617647
## 520 0.138167611                   ExE ectoderm         1.0000000
## 521 0.102760330                   ExE ectoderm         1.0000000
## 522 0.343794715                       Epiblast         0.7959184
## 523 0.012753616           Rostral neurectoderm         0.5500000
## 524 0.023624079                       Epiblast         0.9863014
## 525 0.012256985                       Epiblast         1.0000000
## 526 0.059600245               Surface ectoderm         0.4722222
## 527 0.030460100                       Epiblast         0.5454545
## 528 0.374122212                       Epiblast         0.8260870
## 529 0.307083816                   ExE ectoderm         1.0000000
## 530 0.009486143                   ExE endoderm         1.0000000
## 531 0.392974387                       Epiblast         0.5200000
## 532 0.160227177                       Epiblast         0.9795918
## 533 0.811831827                   ExE ectoderm         1.0000000
## 534 0.979765204                   ExE ectoderm         1.0000000
## 535 0.048417513              Visceral endoderm         0.9677419
## 536 0.017165229               Primitive Streak         0.9875000
## 537 0.241077142      Anterior Primitive Streak         0.7727273
## 538 0.070711951               Primitive Streak         0.8904110
## 539 0.635456761                   ExE ectoderm         1.0000000
## 540 0.687409270               Primitive Streak         0.9350649
## 541 0.025262508                       Epiblast         0.9716981
## 542 0.042107569                       Epiblast         0.8000000
## 543 0.108671209               Primitive Streak         1.0000000
## 544 0.018874915               Somitic mesoderm         0.3095238
## 545 0.119286938                     Mesenchyme         0.7187500
## 546 0.623250183                   ExE endoderm         1.0000000
## 547 0.014128687          Intermediate mesoderm         0.4153846
## 548 0.007732912           Rostral neurectoderm         0.7500000
## 549 0.229792933               Nascent mesoderm         0.6750000
## 550 0.102871584                   ExE ectoderm         1.0000000
## 551 0.212219293 Haematoendothelial progenitors         0.8375000
## 552 0.079383699 Haematoendothelial progenitors         0.4814815
## 553 0.056458107                       Epiblast         1.0000000
## 554 0.048417513      Anterior Primitive Streak         0.9733333
## 555 0.112198799               Nascent mesoderm         0.9473684
## 556 0.079717092                   ExE ectoderm         1.0000000
## 557 0.959186324              Visceral endoderm         1.0000000
## 558 0.746768700                            PGC         0.4166667
## 559 0.009466651                     Mesenchyme         0.5909091
## 560 0.810684273               Primitive Streak         0.8630137
## 561 0.780800805              Parietal endoderm         1.0000000
## 562 0.022635044           Rostral neurectoderm         0.7800000
## 563 0.011409289                       Epiblast         0.9518072
## 564 0.347649339                   ExE endoderm         0.9710145
## 565 0.036308326                       Epiblast         0.9500000
## 566 0.008794174                 Mixed mesoderm         0.6140351
## 567 0.007823087                       Epiblast         1.0000000
## 568 0.078059081               Primitive Streak         0.8072289
## 569 0.955596716                   ExE endoderm         1.0000000
## 570 0.491541585 Haematoendothelial progenitors         1.0000000
## 571 0.637754178                       Epiblast         0.9726027
## 572 0.955596716                       Epiblast         0.5576923
## 573 0.254655772               Nascent mesoderm         0.7340426
## 574 0.012753616                       Epiblast         1.0000000
## 575 0.097228643               Primitive Streak         0.9000000
## 576 0.007518813               Nascent mesoderm         1.0000000
## 577 0.736001060                 Mixed mesoderm         0.5106383
## 578 0.391133384               Primitive Streak         0.9354839
## 579 0.083501973               Primitive Streak         1.0000000
## 580 0.012604594           Rostral neurectoderm         0.9622642
## 581 0.074634545                   ExE ectoderm         1.0000000
## 582 0.012753616                 Mixed mesoderm         0.5405405
## 583 0.056071662               Nascent mesoderm         0.9166667
## 584 0.074875298                       Epiblast         0.9875000
## 585 0.866243776                   ExE ectoderm         1.0000000
## 586 0.873339905              Parietal endoderm         1.0000000
## 587 0.143865203                   ExE ectoderm         1.0000000
## 588 0.342463317                   ExE ectoderm         1.0000000
## 589 0.013847381          Intermediate mesoderm         0.3750000
## 590 0.024959065                     Mesenchyme         1.0000000
## 591 0.772704954               Nascent mesoderm         0.6964286
## 592 0.736001060      Anterior Primitive Streak         0.6721311
## 593 0.194432524               Nascent mesoderm         0.9890110
## 594 0.022683729                     Mesenchyme         0.7608696
## 595 0.009972022                 Mixed mesoderm         0.7727273
## 596 0.007518813           Rostral neurectoderm         0.9142857
## 597 0.062152201               Primitive Streak         1.0000000
## 598 0.154179368                       Epiblast         0.9859155
## 599 0.880463955                       Epiblast         1.0000000
## 600 0.088925754               Primitive Streak         0.6808511
## 601 0.577973598 Haematoendothelial progenitors         0.7090909
## 602 0.291203114               Primitive Streak         0.9473684
## 603 0.412508555      Anterior Primitive Streak         0.6981132
## 604 0.084296490               Primitive Streak         1.0000000
## 605 0.633282034              Visceral endoderm         1.0000000
## 606 0.028374755                       Epiblast         1.0000000
## 607 0.959186324                   ExE endoderm         1.0000000
## 608 0.159339502                   ExE endoderm         1.0000000
## 609 0.484686455               Primitive Streak         0.7272727
## 610 0.026246302                       Epiblast         0.9782609
## 611 0.007518813           Rostral neurectoderm         0.7391304
## 612 0.015691311               Nascent mesoderm         1.0000000
## 613 0.437925008               Primitive Streak         0.9777778
## 614 0.007518813           Rostral neurectoderm         1.0000000
## 615 0.024087495               Nascent mesoderm         0.7246377
## 616 0.049597312                 Mixed mesoderm         0.3859649
## 617 0.160227177                   ExE ectoderm         1.0000000
## 618 0.298542928                   ExE ectoderm         1.0000000
## 619 0.194294092                       Epiblast         0.9558824
## 620 0.479941168               Nascent mesoderm         1.0000000
## 621 0.631611837                       Epiblast         0.9736842
## 622 0.009351811                     Mesenchyme         0.9777778
## 623 0.370977869               Nascent mesoderm         0.8800000
## 624 0.019351974      Anterior Primitive Streak         0.9482759
## 625 0.143057916               Nascent mesoderm         0.7843137
## 626 0.219269932               Nascent mesoderm         0.9491525
## 627 0.082970294                       Epiblast         0.9222222
## 628 0.009351811           Rostral neurectoderm         0.4285714
## 629 0.398509931               Nascent mesoderm         0.9843750
## 630 0.273913625               Nascent mesoderm         1.0000000
## 631 0.027447403                       Epiblast         0.5600000
## 632 0.108179776               Primitive Streak         1.0000000
## 633 0.147869545               Primitive Streak         0.7804878
## 634 0.290535257                       Epiblast         0.9125000
## 635 0.059707805               Primitive Streak         1.0000000
## 636 0.070307838                       Epiblast         1.0000000
## 637 0.021940820                       Epiblast         1.0000000
## 638 0.007732912           Rostral neurectoderm         0.7641509
## 639 0.880463955                   ExE endoderm         1.0000000
## 640 0.230277340                   ExE ectoderm         1.0000000
## 641 0.586794318                       Epiblast         0.4838710
## 642 0.008824876                       Epiblast         0.9876543
## 643 0.019478789               Nascent mesoderm         1.0000000
## 644 0.030460100               Primitive Streak         0.9531250
## 645 0.323997612                   ExE endoderm         1.0000000
## 646 0.426513835               Primitive Streak         0.9866667
## 647 0.396961198                   ExE endoderm         1.0000000
## 648 0.853533293               Nascent mesoderm         0.8958333
## 649 0.050763040                   ExE ectoderm         1.0000000
## 650 0.123873070            Blood progenitors 2         0.9218750
## 651 0.007518813                       Epiblast         1.0000000
## 652 0.375611560                   ExE endoderm         1.0000000
## 653 0.048417513              Visceral endoderm         1.0000000
## 654 0.259210303               Nascent mesoderm         0.6119403
## 655 0.139600924                       Epiblast         0.8833333
## 656 0.050140567                   ExE endoderm         1.0000000
## 657 0.480941107                   ExE ectoderm         1.0000000
## 658 0.019493041               Primitive Streak         0.9333333
## 659 0.083256148                       Epiblast         0.6086957
## 660 0.092435664               Primitive Streak         1.0000000
## 661 0.341853555                   ExE ectoderm         1.0000000
## 662 0.632400811                            PGC         0.6046512
## 663 0.147790816            Blood progenitors 1         0.6181818
## 664 0.258454704                   ExE ectoderm         1.0000000
## 665 0.943287973                   ExE endoderm         1.0000000
## 666 0.035893747                     Mesenchyme         0.9523810
## 667 0.479941168                   ExE ectoderm         1.0000000
## 668 0.011103489               Primitive Streak         0.8983051
## 669 0.720265079               Nascent mesoderm         0.5820896
## 670 0.083256148      Anterior Primitive Streak         0.8771930
## 671 0.494991132                   ExE ectoderm         1.0000000
## 672 0.007518813                Caudal epiblast         0.3606557
## 673 0.398707202                       Epiblast         0.9662921
## 674 0.132962868              Visceral endoderm         1.0000000
## 675 0.012356317               Primitive Streak         0.9710145
## 676 0.971173155                   ExE ectoderm         1.0000000
## 677 0.467822155               Primitive Streak         0.5619048
## 678 0.718780942                   ExE endoderm         0.6619718
## 679 0.046665286               Primitive Streak         0.9886364
## 680 0.084769906                 Mixed mesoderm         0.5400000
## 681 0.028625872                            Gut         0.7692308
## 682 0.077531195                       Epiblast         1.0000000
## 683 0.022276328           Rostral neurectoderm         0.8352941
## 684 0.121969489      Anterior Primitive Streak         0.6296296
## 685 0.008405047                 Mixed mesoderm         0.6808511
## 686 0.553197264                   ExE endoderm         1.0000000
## 687 0.059707805                            Gut         0.8863636
## 688 0.019478789                       Epiblast         0.9814815
## 689 0.412592858                   ExE endoderm         1.0000000
## 690 0.101850646               Primitive Streak         1.0000000
## 691 0.073899437               Nascent mesoderm         0.8142857
## 692 0.078498387               Surface ectoderm         0.6250000
## 693 0.079030909                       Epiblast         1.0000000
## 694 0.025338981                       Epiblast         0.9019608
## 695 0.030229397                Caudal epiblast         0.7777778
## 696 0.171015418                   ExE endoderm         1.0000000
## 697 0.014128687                   ExE endoderm         1.0000000
## 698 0.024568295           Rostral neurectoderm         0.5000000
## 699 0.107695439            Blood progenitors 2         0.9855072
## 700 0.652067262               Primitive Streak         0.8840580
## 701 0.041818648              Visceral endoderm         1.0000000
## 702 0.027111029                       Epiblast         0.9891304
## 703 0.684751494                   ExE ectoderm         1.0000000
## 704 0.007518813                Caudal epiblast         0.7962963
## 705 0.308140993 Haematoendothelial progenitors         0.9880952
## 706 0.083922676               Nascent mesoderm         1.0000000
## 707 0.020870573                Caudal epiblast         0.5333333
## 708 0.064668952                       Epiblast         0.7272727
## 709 0.083846973                   ExE ectoderm         1.0000000
## 710 0.077531195                   ExE ectoderm         1.0000000
## 711 0.021073172                   ExE endoderm         1.0000000
## 712 0.021348394               Nascent mesoderm         1.0000000
## 713 0.046851761                       Epiblast         0.9818182
## 714 0.097228643                       Epiblast         0.9696970
## 715 0.991866818                   ExE ectoderm         1.0000000
## 716 0.017342805                 Mixed mesoderm         0.6031746
## 717 0.111662963               Primitive Streak         0.3833333
## 718 0.098908975               Nascent mesoderm         0.7571429
## 719 0.292152648               Nascent mesoderm         0.8431373
## 720 0.112315018               Primitive Streak         0.5303030
## 721 0.183374613                       Epiblast         0.7407407
## 722 0.259794984                   ExE ectoderm         1.0000000
## 723 0.007518813                       Epiblast         1.0000000
## 724 0.059707805               Primitive Streak         0.9843750
## 725 0.007518813                       Epiblast         1.0000000
## 726 0.170850625                   ExE mesoderm         0.4038462
## 727 0.010456710                       Epiblast         0.7777778
## 728 0.039793008                   ExE ectoderm         1.0000000
## 729 0.007732912          Intermediate mesoderm         0.4693878
## 730 0.007518813               Primitive Streak         0.7200000
## 731 0.061009174                  Def. endoderm         0.8888889
## 732 0.052970137                       Epiblast         1.0000000
## 733 0.009950537                       Epiblast         0.7083333
## 734 0.342463317                   ExE ectoderm         1.0000000
## 735 0.007518813                       Epiblast         0.9367089
## 736 0.161488815                   ExE ectoderm         1.0000000
## 737 0.254979195                 Mixed mesoderm         0.6000000
## 738 0.016046162           Rostral neurectoderm         0.7166667
## 739 0.012256985                       Epiblast         1.0000000
## 740 0.577973598                   ExE ectoderm         1.0000000
## 741 0.024164156                       Epiblast         0.6567164
## 742 0.040692708                       Epiblast         0.7047619
## 743 0.132962868                   ExE ectoderm         1.0000000
## 744 0.021348394               Nascent mesoderm         0.5679012
## 745 0.168887583                   ExE endoderm         1.0000000
## 746 0.014128687                       Epiblast         1.0000000
## 747 0.232472211            Blood progenitors 2         0.5000000
## 748 0.198320792               Nascent mesoderm         1.0000000
## 749 0.012753616                       Epiblast         0.5000000
## 750 0.040661745                       Epiblast         0.9908257
## 751 0.124337467               Primitive Streak         0.9777778
## 752 0.013847381               Primitive Streak         0.7564103
## 753 0.051500490                 Mixed mesoderm         0.5492958
## 754 0.059707805                       Epiblast         1.0000000
## 755 0.145491907                       Epiblast         0.8805970
## 756 0.007518813                       Epiblast         0.9693878
## 757 0.392974387               Nascent mesoderm         0.9000000
## 758 0.007518813                       Epiblast         1.0000000
## 759 0.027447403               Primitive Streak         0.7575758
## 760 0.087625468               Primitive Streak         0.9385965
## 761 0.007518813                Caudal epiblast         0.6666667
## 762 0.441625161               Nascent mesoderm         0.9876543
## 763 0.113629450                Caudal epiblast         0.4897959
## 764 0.193708080      Anterior Primitive Streak         0.9090909
## 765 0.008824876               Primitive Streak         0.8536585
## 766 0.819406693                   ExE ectoderm         1.0000000
## 767 0.145491907                   ExE endoderm         1.0000000
## 768 0.085191217            Blood progenitors 1         0.6500000
## 769 0.658178883                   ExE endoderm         1.0000000
## 770 0.132962868                   ExE endoderm         0.8245614
## 771 0.012753616                       Epiblast         1.0000000
## 772 0.205389393           Rostral neurectoderm         0.5531915
## 773 0.012256985                       Epiblast         1.0000000
## 774 0.021073172                       Epiblast         0.9148936
## 775 0.352344676 Haematoendothelial progenitors         0.9814815
## 776 0.025262508                       Epiblast         0.9479167
## 777 0.069427808                   ExE endoderm         1.0000000
## 778 0.036665613                       Epiblast         1.0000000
## 779 0.973838559                   ExE endoderm         1.0000000
## 780 0.007518813           Rostral neurectoderm         1.0000000
## 781 0.101952580               Primitive Streak         1.0000000
## 782 0.082865130                       Epiblast         1.0000000
## 783 0.723851914                   ExE ectoderm         1.0000000
## 784 0.027422265               Nascent mesoderm         1.0000000
## 785 0.024568295                       Epiblast         1.0000000
## 786 0.830428225               Nascent mesoderm         0.9647059
## 787 0.014128687                       Epiblast         1.0000000
## 788 0.033171441               Primitive Streak         1.0000000
## 789 0.366841590               Nascent mesoderm         0.9090909
## 790 0.023834359                       Epiblast         0.9894737
## 791 0.275578686                   ExE endoderm         1.0000000
## 792 0.028725880                   ExE mesoderm         0.5675676
## 793 0.010456710                       Epiblast         0.8857143
## 794 0.086711383                     Mesenchyme         0.9615385
## 795 0.821440314               Nascent mesoderm         0.9846154
## 796 0.128518752               Primitive Streak         0.9821429
## 797 0.010616546                       Epiblast         1.0000000
## 798 0.617747573                   ExE ectoderm         1.0000000
## 799 0.710681549                   ExE ectoderm         1.0000000
## 800 0.076033372               Primitive Streak         0.8604651
## 801 0.275578686               Primitive Streak         0.7555556
## 802 0.008395562                 Mixed mesoderm         0.4285714
## 803 0.046943396                   ExE ectoderm         1.0000000
## 804 0.008781951                       Epiblast         1.0000000
## 805 0.633138491                   ExE ectoderm         1.0000000
## 806 0.018223180                 Mixed mesoderm         0.7818182
## 807 0.917984353                   ExE endoderm         1.0000000
## 808 0.071951281               Primitive Streak         0.9324324
## 809 0.154102387                       Epiblast         1.0000000
## 810 0.220051533                   ExE ectoderm         1.0000000
## 811 0.015200187                   ExE ectoderm         1.0000000
## 812 0.083256148                 Mixed mesoderm         0.7012987
## 813 0.298739235               Nascent mesoderm         0.6666667
## 814 0.008405047                 Mixed mesoderm         0.6949153
## 815 0.349386308                       Epiblast         0.9827586
## 816 0.425995143                       Epiblast         0.8219178
## 817 0.048417513                       Epiblast         0.9795918
## 818 0.162510179                       Epiblast         0.8319328
## 819 0.110128191                       Epiblast         1.0000000
## 820 0.012296241                       Epiblast         0.9729730
## 821 0.033171441               Primitive Streak         0.9861111
## 822 0.401654460                   ExE ectoderm         1.0000000
## 823 0.017342805                       Epiblast         0.9531250
## 824 0.185615043                       Epiblast         1.0000000
## 825 0.084296490                       Epiblast         1.0000000
## 826 0.108703536            Blood progenitors 1         0.8947368
## 827 0.018310396                   ExE endoderm         1.0000000
## 828 0.672766911 Haematoendothelial progenitors         1.0000000
## 829 0.016766954               Primitive Streak         0.8166667
## 830 0.236485894               Nascent mesoderm         1.0000000
## 831 0.221569517                   ExE ectoderm         1.0000000
## 832 0.023625907           Rostral neurectoderm         0.4255319
## 833 0.966166572                   ExE endoderm         1.0000000
## 834 0.724134243              Parietal endoderm         1.0000000
## 835 0.819406693                   ExE ectoderm         1.0000000
## 836 0.053496063               Primitive Streak         0.9909910
## 837 0.042107569                   ExE endoderm         1.0000000
## 838 0.083846973                       Epiblast         0.9928058
## 839 0.991866818      Anterior Primitive Streak         0.3611111
## 840 0.038726982               Nascent mesoderm         0.6530612
## 841 0.019691482               Primitive Streak         0.9000000
## 842 0.247144370               Primitive Streak         1.0000000
## 843 0.042107569                Caudal epiblast         0.5593220
## 844 0.043836438                   ExE ectoderm         1.0000000
## 845 0.744893154               Nascent mesoderm         0.9220779
## 846 0.946557149                   ExE ectoderm         1.0000000
## 847 0.734859246      Anterior Primitive Streak         0.5454545
## 848 0.291203114               Primitive Streak         0.7924528
## 849 0.575910055                   ExE endoderm         1.0000000
## 850 0.598079879                 Mixed mesoderm         0.5762712
## 851 0.950265959                   ExE ectoderm         1.0000000
## 852 0.089235012               Primitive Streak         0.6666667</code></pre>
<p>While neighbourhoods tend to be homogeneous, we can define a threshold for <code>celltype_fraction</code> to exclude neighbourhoods that are a mix of cell types.</p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="fu">ggplot</span>(<span class="no">da_results</span>, <span class="fu">aes</span>(<span class="no">celltype_fraction</span>)) + <span class="fu">geom_histogram</span>(<span class="kw">bins</span><span class="kw">=</span><span class="fl">50</span>)</pre></body></html></div>
<p><img src="milo_gastrulation_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="no">da_results</span>$<span class="no">celltype</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="no">da_results</span>$<span class="no">celltype_fraction</span> <span class="kw">&lt;</span> <span class="fl">0.7</span>, <span class="st">"Mixed"</span>, <span class="no">da_results</span>$<span class="no">celltype</span>)</pre></body></html></div>
<p>Now we can visualize the distribution of DA Fold Changes in different cell types</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="fu"><a href="../reference/plotDAbeeswarm.html">plotDAbeeswarm</a></span>(<span class="no">da_results</span>, <span class="kw">group.by</span> <span class="kw">=</span> <span class="st">"celltype"</span>)</pre></body></html></div>
<p><img src="milo_gastrulation_files/figure-html/unnamed-chunk-20-1.png" width="672"></p>
<p>This is already quite informative: we can see that certain early development cell types, such as epiblast and primitive streak, are enriched in the earliest time stage, while others are enriched later in development, such as ectoderm cells. Interestingly, we also see plenty of DA neighbourhood with a mixed label. This could indicate that transitional states show changes in abundance in time.</p>
</div>
<div id="identifying-signatures-of-da-subpopulations" class="section level1">
<h1 class="hasAnchor">
<a href="#identifying-signatures-of-da-subpopulations" class="anchor"></a>Identifying signatures of DA subpopulations</h1>
<p>Once we have identified neighbourhoods with significant DA, we might be interested in learning more about the gene expression signatures that define these subpopulations. Here the analyst might get creative, depending on the specific characteristics of their dataset and the biological question of interest. In the <code>miloR</code> package, we provide functionality to explore neighbourhood markers through the function <code>findNhoodMarkers</code>. This groups significantly DA neighbourhoods that show concordant Fold-Change and performs a test for differential expression between cells in those neighbourhoods.</p>
<p>In practice, it might be convenient to subset a selected number of neighbourhoods of interest for gene-level downstream analysis. For the sake of demonstration, here we focus on identifying signatures of DA subpopulations in the endoderm development lineage.</p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">logcounts</a></span>(<span class="no">embryo_milo</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log1p</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span>(<span class="no">embryo_milo</span>))
<span class="no">dge_smp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/findNhoodMarkers.html">findNhoodMarkers</a></span>(<span class="no">embryo_milo</span>, <span class="no">da_results</span>,
                                     <span class="kw">assay</span> <span class="kw">=</span> <span class="st">"counts"</span>, <span class="kw">gene.offset</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">da.fdr</span> <span class="kw">=</span> <span class="fl">0.1</span>,
                                     <span class="kw">aggregate.samples</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">sample_col</span> <span class="kw">=</span> <span class="st">"sample"</span>,
                                     <span class="kw">subset.nhoods</span> <span class="kw">=</span> <span class="no">da_results</span>$<span class="no">celltype</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Anterior Primitive Streak"</span>, <span class="st">"Def. endoderm"</span>, <span class="st">"Gut"</span>, <span class="st">"Visceral endoderm"</span>)
                                     )

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">dge_smp</span>)</pre></body></html></div>
<pre><code>##                       logFC_1 adj.P.Val_1   logFC_2 adj.P.Val_2   logFC_3
## ENSMUSG00000051951  0.0000000           1 0.0000000           1 0.0000000
## ENSMUSG00000089699  0.0000000           1 0.0000000           1 0.0000000
## ENSMUSG00000102343  0.0000000           1 0.0000000           1 0.0000000
## ENSMUSG00000025900  0.0000000           1 0.0000000           1 0.0000000
## ENSMUSG00000025902 -0.1955242           1 0.9332859           1 0.5272697
## ENSMUSG00000104328  0.0000000           1 0.0000000           1 0.0000000
##                    adj.P.Val_3   logFC_4 adj.P.Val_4             GeneID
## ENSMUSG00000051951           1  0.000000    1.000000 ENSMUSG00000051951
## ENSMUSG00000089699           1  0.000000    1.000000 ENSMUSG00000089699
## ENSMUSG00000102343           1  0.000000    1.000000 ENSMUSG00000102343
## ENSMUSG00000025900           1  0.000000    1.000000 ENSMUSG00000025900
## ENSMUSG00000025902           1 -1.844589    0.260445 ENSMUSG00000025902
## ENSMUSG00000104328           1  0.000000    1.000000 ENSMUSG00000104328</code></pre>
<p>This identifies n marker genes at FDR 10% that distinguish two main groups within the epiblast neighbourhoods, one significantly depleted in the early stage and one significantly enriched. We can visualize expression of the detected marker genes using the function <code>plotNhoodExpressionDA</code>. This shows the average expression in each neighbourhood, ranked by log-Fold Change in the DA test. Note that the gene x nhood expression matrix can be pre-computed and stored using the <code>calcNhoodExpression</code> function, to avoid repeating the computation every time you need to plot.</p>
<p>In this case we mainly identified negative markers of the epiblast neighbourhoods enriched with age.</p>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="no">markers</span> <span class="kw">&lt;-</span> <span class="no">dge_smp</span>[<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">dge_smp</span>$<span class="no">adj.P.Val_1</span> <span class="kw">&lt;</span> <span class="fl">0.1</span> ), <span class="st">"GeneID"</span>]
<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">logcounts</a></span>(<span class="no">embryo_milo</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log1p</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span>(<span class="no">embryo_milo</span>))
<span class="no">embryo_milo</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/calcNhoodExpression.html">calcNhoodExpression</a></span>(<span class="no">embryo_milo</span>, <span class="kw">subset.row</span><span class="kw">=</span><span class="no">markers</span>)

<span class="fu"><a href="../reference/plotNhoodExpressionDA.html">plotNhoodExpressionDA</a></span>(<span class="no">embryo_milo</span>, <span class="no">da_results</span>, <span class="kw">features</span> <span class="kw">=</span> <span class="no">markers</span>,
                      <span class="kw">subset.nhoods</span> <span class="kw">=</span> <span class="no">da_results</span>$<span class="no">celltype</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Anterior Primitive Streak"</span>, <span class="st">"Def. endoderm"</span>, <span class="st">"Gut"</span>, <span class="st">"Visceral endoderm"</span>),
                      <span class="kw">assay</span><span class="kw">=</span><span class="st">"logcounts"</span>,
                      <span class="kw">scale_to_1</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">cluster_features</span> <span class="kw">=</span> <span class="fl">TRUE</span>
                      )</pre></body></html></div>
<p><img src="milo_gastrulation_files/figure-html/unnamed-chunk-22-1.png" width="864"></p>
<details><p><summary><strong>Session Info</strong></summary></p>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</pre></body></html></div>
<pre><code>## R version 3.6.3 (2020-02-29)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.3 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
## LAPACK: /usr/lib/x86_64-linux-gnu/openblas/liblapack.so.3
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] MouseGastrulationData_1.0.0 patchwork_1.0.1            
##  [3] dplyr_1.0.2                 scater_1.14.6              
##  [5] ggplot2_3.3.2               SingleCellExperiment_1.8.0 
##  [7] SummarizedExperiment_1.16.1 DelayedArray_0.12.3        
##  [9] BiocParallel_1.20.1         matrixStats_0.56.0         
## [11] Biobase_2.46.0              GenomicRanges_1.38.0       
## [13] GenomeInfoDb_1.22.1         IRanges_2.20.2             
## [15] S4Vectors_0.24.4            BiocGenerics_0.32.0        
## [17] miloR_0.1.0                 edgeR_3.28.1               
## [19] limma_3.42.2                BiocStyle_2.14.4           
## 
## loaded via a namespace (and not attached):
##   [1] ggbeeswarm_0.6.0              colorspace_1.4-1             
##   [3] ellipsis_0.3.1                rprojroot_1.3-2              
##   [5] XVector_0.26.0                BiocNeighbors_1.4.2          
##   [7] fs_1.5.0                      rstudioapi_0.11              
##   [9] farver_2.0.3                  graphlayouts_0.7.0           
##  [11] ggrepel_0.8.2                 bit64_4.0.2                  
##  [13] AnnotationDbi_1.48.0          interactiveDisplayBase_1.24.0
##  [15] splines_3.6.3                 codetools_0.2-16             
##  [17] knitr_1.29                    polyclip_1.10-0              
##  [19] dbplyr_1.4.4                  uwot_0.1.8                   
##  [21] ggforce_0.3.2                 shiny_1.5.0                  
##  [23] BiocManager_1.30.10           compiler_3.6.3               
##  [25] httr_1.4.2                    backports_1.1.9              
##  [27] fastmap_1.0.1                 assertthat_0.2.1             
##  [29] Matrix_1.2-18                 later_1.1.0.1                
##  [31] tweenr_1.0.1                  BiocSingular_1.2.2           
##  [33] htmltools_0.5.0               tools_3.6.3                  
##  [35] rsvd_1.0.3                    igraph_1.2.5                 
##  [37] gtable_0.3.0                  glue_1.4.1                   
##  [39] GenomeInfoDbData_1.2.2        rappdirs_0.3.1               
##  [41] Rcpp_1.0.5                    pkgdown_1.5.1                
##  [43] vctrs_0.3.2                   ExperimentHub_1.12.0         
##  [45] DelayedMatrixStats_1.8.0      ggraph_2.0.3                 
##  [47] xfun_0.16                     stringr_1.4.0                
##  [49] mime_0.9                      lifecycle_0.2.0              
##  [51] irlba_2.3.3                   gtools_3.8.2                 
##  [53] statmod_1.4.34                AnnotationHub_2.18.0         
##  [55] zlibbioc_1.32.0               MASS_7.3-51.5                
##  [57] scales_1.1.1                  tidygraph_1.2.0              
##  [59] promises_1.1.1                yaml_2.2.1                   
##  [61] curl_4.3                      memoise_1.1.0                
##  [63] gridExtra_2.3                 stringi_1.4.6                
##  [65] RSQLite_2.2.0                 BiocVersion_3.10.1           
##  [67] desc_1.2.0                    rlang_0.4.7                  
##  [69] pkgconfig_2.0.3               bitops_1.0-6                 
##  [71] evaluate_0.14                 lattice_0.20-40              
##  [73] purrr_0.3.4                   labeling_0.3                 
##  [75] cowplot_1.0.0                 bit_4.0.4                    
##  [77] tidyselect_1.1.0              RcppAnnoy_0.0.16             
##  [79] magrittr_1.5                  bookdown_0.20                
##  [81] R6_2.4.1                      generics_0.0.2               
##  [83] DBI_1.1.0                     pillar_1.4.6                 
##  [85] withr_2.2.0                   RCurl_1.98-1.2               
##  [87] tibble_3.0.3                  crayon_1.3.4                 
##  [89] BiocFileCache_1.10.2          rmarkdown_2.3                
##  [91] viridis_0.5.1                 locfit_1.5-9.4               
##  [93] grid_3.6.3                    blob_1.2.1                   
##  [95] digest_0.6.25                 xtable_1.8-4                 
##  [97] tidyr_1.1.1                   httpuv_1.5.4                 
##  [99] munsell_0.5.0                 beeswarm_0.2.3               
## [101] viridisLite_0.3.0             vipor_0.4.5</code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Mike Morgan, Emma Dann.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
