<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Differential abundance testing with Milo - Mouse gastrulation example • miloR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Differential abundance testing with Milo - Mouse gastrulation example">
<meta property="og:description" content="miloR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">miloR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/milo_demo.html">Differential abundance testing with Milo</a>
    </li>
    <li>
      <a href="../articles/milo_gastrulation.html">Differential abundance testing with Milo - Mouse gastrulation example</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="milo_gastrulation_files/header-attrs-2.3/header-attrs.js"></script><script src="milo_gastrulation_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Differential abundance testing with Milo - Mouse gastrulation example</h1>
                        <h4 class="author">Emma Dann</h4>
                        <h4 class="author">Mike Morgan</h4>
            
      
      
      <div class="hidden name"><code>milo_gastrulation.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">miloR</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">SingleCellExperiment</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">scater</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">patchwork</span>)</pre></body></html></div>
<div id="load-data" class="section level1">
<h1 class="hasAnchor">
<a href="#load-data" class="anchor"></a>Load data</h1>
<p>For this vignette we will use the mouse gastrulation single-cell data from <a href="https://www.nature.com/articles/s41586-019-0933-9">Pijuan-Sala et al. 2019</a>. The dataset can be downloaded as a <code>SingleCellExperiment</code> object from the <a href="https://bioconductor.org/packages/3.12/data/experiment/html/MouseGastrulationData.html"><code>MouseGastrulationData</code></a> package on Bioconductor. To make computations faster, here we will download just a subset of samples, 4 samples at stage E7 and 4 samples at stage E7.5.</p>
<p>This dataset has already been pre-processed and contains a <code>pca.corrected</code> dimensionality reduction, which was built after batch correction using <a href="https://bioconductor.org/packages/release/bioc/vignettes/batchelor/inst/doc/correction.html"><code>fastMNN</code></a>.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">MouseGastrulationData</span>)
<span class="no">select_samples</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>,  <span class="fl">3</span>,  <span class="fl">6</span>, <span class="fl">15</span>,
                    <span class="co"># 4, 19, </span>
                    <span class="fl">10</span>, <span class="fl">14</span>, <span class="fl">20</span>, <span class="fl">30</span>
                    <span class="co">#31, 32</span>
                    )
<span class="no">embryo_data</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MouseGastrulationData/man/EmbryoAtlasData.html">EmbryoAtlasData</a></span>(<span class="kw">samples</span> <span class="kw">=</span> <span class="no">select_samples</span>)</pre></body></html></div>
<pre><code>## Warning: `select_()` is deprecated as of dplyr 0.7.0.
## Please use `select()` instead.
## This warning is displayed once every 8 hours.
## Call `lifecycle::last_warnings()` to see where this warning was generated.</code></pre>
<pre><code>## Warning: `filter_()` is deprecated as of dplyr 0.7.0.
## Please use `filter()` instead.
## See vignette('programming') for more help
## This warning is displayed once every 8 hours.
## Call `lifecycle::last_warnings()` to see where this warning was generated.</code></pre>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">embryo_data</span></pre></body></html></div>
<pre><code>## class: SingleCellExperiment 
## dim: 29452 14679 
## metadata(0):
## assays(1): counts
## rownames(29452): ENSMUSG00000051951 ENSMUSG00000089699 ...
##   ENSMUSG00000096730 ENSMUSG00000095742
## rowData names(2): ENSEMBL SYMBOL
## colnames(14679): cell_361 cell_362 ... cell_107079 cell_107080
## colData names(16): cell barcode ... celltype colour
## reducedDimNames(2): pca.corrected umap
## spikeNames(0):
## altExpNames(0):</code></pre>
</div>
<div id="visualize-the-data" class="section level1">
<h1 class="hasAnchor">
<a href="#visualize-the-data" class="anchor"></a>Visualize the data</h1>
<p>We recompute PCA and UMAP embedding for this subset of cells to visualize the data.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">embryo_data</span> <span class="kw">&lt;-</span> <span class="no">embryo_data</span>[,<span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span>(<span class="no">embryo_data</span>, <span class="st">"pca.corrected"</span>), <span class="fl">1</span>, <span class="kw">function</span>(<span class="no">x</span>) !<span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">x</span>)))]
<span class="no">embryo_data</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span>(<span class="no">embryo_data</span>, <span class="kw">dimred</span> <span class="kw">=</span> <span class="st">"pca.corrected"</span>, <span class="kw">name</span> <span class="kw">=</span> <span class="st">'umap'</span>)

<span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span>(<span class="no">embryo_data</span>, <span class="kw">colour_by</span><span class="kw">=</span><span class="st">"stage"</span>, <span class="kw">dimred</span> <span class="kw">=</span> <span class="st">"umap"</span>)</pre></body></html></div>
<p><img src="milo_gastrulation_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>We will test for significant differences in abundance of cells between these stages of development, and the associated gene signatures.</p>
</div>
<div id="differential-abundance-testing" class="section level1">
<h1 class="hasAnchor">
<a href="#differential-abundance-testing" class="anchor"></a>Differential abundance testing</h1>
<div id="create-a-milo-object" class="section level2">
<h2 class="hasAnchor">
<a href="#create-a-milo-object" class="anchor"></a>Create a Milo object</h2>
<p>For differential abundance analysis on graph neighbourhoods we first construct a <code>Milo</code> object. This extends the <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html"><code>SingleCellExperiment</code></a> class to store information about neighbourhoods on the KNN graph.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">embryo_milo</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/Milo.html">Milo</a></span>(<span class="no">embryo_data</span>)
<span class="no">embryo_milo</span></pre></body></html></div>
<pre><code>## class: Milo 
## dim: 29452 13002 
## metadata(0):
## assays(1): counts
## rownames(29452): ENSMUSG00000051951 ENSMUSG00000089699 ...
##   ENSMUSG00000096730 ENSMUSG00000095742
## rowData names(2): ENSEMBL SYMBOL
## colnames(13002): cell_361 cell_362 ... cell_107079 cell_107080
## colData names(16): cell barcode ... celltype colour
## reducedDimNames(2): pca.corrected umap
## spikeNames(0):
## altExpNames(0):
## nhoods dimensions(1): 0
## nhoodCounts dimensions(2): 1 1
## nhoodDistances dimension(1): 0
## graph names(0):
## nhoodIndex names(1): 0
## nhoodExpression dimension(2): 1 1
## nhoodReducedDim names(0):
## nhoodGraph names(0):
## nhoodAdjacency dimension(0):</code></pre>
</div>
<div id="construct-knn-graph" class="section level2">
<h2 class="hasAnchor">
<a href="#construct-knn-graph" class="anchor"></a>Construct KNN graph</h2>
<p>We need to add the KNN graph to the Milo object. This is stored in the <code>graph</code> slot, in <a href="https://igraph.org/r/"><code>igraph</code></a> format. The <code>miloR</code> package includes functionality to build and store the graph from the PCA dimensions stored in the <code>reducedDim</code> slot. In this case, we specify that we want to build the graph from the MNN corrected PCA dimensions.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">embryo_milo</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/buildGraph.html">buildGraph</a></span>(<span class="no">embryo_milo</span>, <span class="kw">k</span> <span class="kw">=</span> <span class="fl">30</span>, <span class="kw">d</span> <span class="kw">=</span> <span class="fl">30</span>, <span class="kw">reduced.dim</span> <span class="kw">=</span> <span class="st">"pca.corrected"</span>)</pre></body></html></div>
<p>Alternatively, one can add a precomputed KNN graph (for example constructed with Seurat or scanpy) to the <code>graph</code> slot using the adjacency matrix, through the helper function <code>buildFromAdjacency</code>.</p>
<!-- Alternatively, if you already have a KNN graph (for example constructed with Seurat/scanpy) you can add it from the adjacency matrix. -->
<!-- ```{r} -->
<!-- # ## Build up a mock SNN graph made with Seurat -->
<!-- # pca_df <- reducedDim(traj_milo, "PCA") -->
<!-- # rownames(pca_df) <- traj_milo$cell_id -->
<!-- # snn_graph <- FindNeighbors(pca_df)[["snn"]] -->
<!-- #  -->
<!-- # graph(traj_milo) <-  graph(buildFromAdjacency(snn_graph, k=10)) -->
<!-- ``` -->
</div>
<div id="defining-representative-neighbourhoods-on-the-knn-graph" class="section level2">
<h2 class="hasAnchor">
<a href="#defining-representative-neighbourhoods-on-the-knn-graph" class="anchor"></a>Defining representative neighbourhoods on the KNN graph</h2>
<p>We define the neighbourhood of a cell, the index, as the group of cells connected by an edge in the KNN graph to the index cell. For efficiency, we don’t test for DA in the neighbourhood of every cell, but we sample as indices a subset of representative cells, using a KNN sampling algorithm used by <a href="https://www.nature.com/articles/nmeth.3545">Gut et al. 2015</a>.</p>
<p>For sampling you need to define a few parameters:</p>
<ul>
<li>
<code>prop</code>: the proportion of cells to randomly sample to start with (usually 0.1 - 0.2 is sufficient)</li>
<li>
<code>k</code>: the k to use for KNN refinement (we recommend using the same k used for KNN graph building)</li>
<li>
<code>d</code>: the number of reduced dimensions to use for KNN refinement (we recommend using the same d used for KNN graph building)</li>
<li>
<code>refined</code>: indicates whether you want to use the sampling refinement algorith, or just pick cells at random. The default and recommended way to go is to use refinement. The only situation in which you might consider using <code>random</code> instead, is if you have batch corrected your data with a graph based correction algorithm, such as <a href="https://github.com/Teichlab/bbknn">BBKNN</a>, but the results of DA testing will be suboptimal.</li>
</ul>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">embryo_milo</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/makeNhoods.html">makeNhoods</a></span>(<span class="no">embryo_milo</span>, <span class="kw">prop</span> <span class="kw">=</span> <span class="fl">0.1</span>, <span class="kw">k</span> <span class="kw">=</span> <span class="fl">30</span>, <span class="kw">d</span><span class="kw">=</span><span class="fl">30</span>, <span class="kw">refined</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">reduced_dims</span> <span class="kw">=</span> <span class="st">"pca.corrected"</span>)</pre></body></html></div>
<p>Once we have defined neighbourhoods, it’s good to take a look at how big the neighbourhoods are (i.e. how many cells form each neighbourhood). This affects the power of DA testing. We can check this out using the <code>plotNhoodSizeHist</code> function. Empirically, we found it’s best to have a distribution peaking above 20. Otherwise you might consider rerunning <code>makeNhoods</code> increasing <code>k</code> and/or <code>prop</code>.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="fu"><a href="../reference/plotNhoodSizeHist.html">plotNhoodSizeHist</a></span>(<span class="no">embryo_milo</span>)</pre></body></html></div>
<p><img src="milo_gastrulation_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
<div id="counting-cells-in-neighbourhoods" class="section level2">
<h2 class="hasAnchor">
<a href="#counting-cells-in-neighbourhoods" class="anchor"></a>Counting cells in neighbourhoods</h2>
<p><em>Milo</em> leverages the variation in cell numbers between replicates for the same experimental condition to test for differential abundance. Therefore we have to count how many cells from each sample are in each neighbourhood. We need to use the cell metadata and specify which column contains the sample information.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">embryo_milo</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/countCells.html">countCells</a></span>(<span class="no">embryo_milo</span>, <span class="kw">meta.data</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="fu">colData</span>(<span class="no">embryo_milo</span>)), <span class="kw">sample</span><span class="kw">=</span><span class="st">"sample"</span>)</pre></body></html></div>
<p>This adds to the <code>Milo</code> object a <span class="math inline">\(n \times m\)</span> matrix, where <span class="math inline">\(n\)</span> is the number of neighbourhoods and <span class="math inline">\(m\)</span> is the number of experimental samples. Values indicate the number of cells from each sample counted in a neighbourhood. This count matrix will be used for DA testing.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/methods.html">nhoodCounts</a></span>(<span class="no">embryo_milo</span>))</pre></body></html></div>
<pre><code>## 6 x 8 sparse Matrix of class "dgCMatrix"
##   2  3  6 15 10 14 20 30
## 1 8 11 38  1  .  2 19  .
## 2 .  .  8 15 38  4  5 58
## 3 .  .  7  4 48  4  5 39
## 4 .  .  6 21  8 14  4 42
## 5 .  .  8 10  5  5  6 27
## 6 1  7 22  3  2  2 15  1</code></pre>
</div>
<div id="defining-experimental-design" class="section level2">
<h2 class="hasAnchor">
<a href="#defining-experimental-design" class="anchor"></a>Defining experimental design</h2>
<p>Now we are all set to test for differential abundance in neighbourhoods. We implement this hypothesis testing in a generalized linear model (GLM) framework, specifically using the Negative Binomial GLM implementation in <a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html"><code>edgeR</code></a>.</p>
<p>We first need to think about our experimental design. The design matrix should match each sample to the experimental condition of interest for DA testing. In this case, we want to detect DA between embryonic stages, stored in the <code>stage</code> column of the dataset <code>colData</code>. We also include the <code>sequencing.batch</code> column in the design matrix. This represents a known technical covariate that we want to account for in DA testing.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">embryo_design</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="fu">colData</span>(<span class="no">embryo_milo</span>))[,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"sample"</span>, <span class="st">"stage"</span>, <span class="st">"sequencing.batch"</span>)]
<span class="co">## Convert batch info from integer to factor</span>
<span class="no">embryo_design</span>$<span class="no">sequencing.batch</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>(<span class="no">embryo_design</span>$<span class="no">sequencing.batch</span>)
<span class="no">embryo_design</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html">distinct</a></span>(<span class="no">embryo_design</span>)
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">embryo_design</span>) <span class="kw">&lt;-</span> <span class="no">embryo_design</span>$<span class="no">sample</span>

<span class="no">embryo_design</span></pre></body></html></div>
<pre><code>##    sample stage sequencing.batch
## 2       2  E7.5                1
## 3       3  E7.5                1
## 6       6  E7.5                1
## 15     15  E7.0                2
## 10     10  E7.0                1
## 14     14  E7.0                2
## 20     20  E7.5                2
## 30     30  E7.0                3</code></pre>
</div>
<div id="computing-neighbourhood-connectivity" class="section level2">
<h2 class="hasAnchor">
<a href="#computing-neighbourhood-connectivity" class="anchor"></a>Computing neighbourhood connectivity</h2>
<p>Milo uses an adaptation of the Spatial FDR correction introduced by <a href="https://bioconductor.org/packages/release/bioc/html/cydar.html">cydar</a>, which accounts for the overlap between neighbourhoods. Specifically, each hypothesis test P-value is weighted by the reciprocal of the kth nearest neighbour distance. To use this statistic we first need to store the distances between nearest neighbors in the Milo object. This is done by the <code>calcNhoodDistance</code> function (N.B. this step is the most time consuming of the analysis workflow and might take a couple of minutes for large datasets).</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="no">embryo_milo</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/calcNhoodDistance.html">calcNhoodDistance</a></span>(<span class="no">embryo_milo</span>, <span class="kw">d</span><span class="kw">=</span><span class="fl">30</span>, <span class="kw">reduced.dim</span> <span class="kw">=</span> <span class="st">"pca.corrected"</span>)</pre></body></html></div>
</div>
<div id="testing" class="section level2">
<h2 class="hasAnchor">
<a href="#testing" class="anchor"></a>Testing</h2>
<p>Now we can do the DA test, explicitly defining our experimental design. In this case, we want to dest for differences between experimental stages, while accounting for the variability between technical batches (You can find more info on how to use formulas to define a testing design in R <a href="https://r4ds.had.co.nz/model-basics.html#formulas-and-model-families">here</a>)</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="no">da_results</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/testNhoods.html">testNhoods</a></span>(<span class="no">embryo_milo</span>, <span class="kw">design</span> <span class="kw">=</span> ~ <span class="no">sequencing.batch</span> + <span class="no">stage</span>, <span class="kw">design.df</span> <span class="kw">=</span> <span class="no">embryo_design</span>)</pre></body></html></div>
<p>This calculates a Fold-change and corrected P-value for each neighbourhood, which indicates wheather there is significant differential abundance between developmental stages.</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="no">da_results</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(<span class="no">SpatialFDR</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>()</pre></body></html></div>
<pre><code>##         logFC   logCPM        F       PValue         FDR Nhood  SpatialFDR
## 11   7.337916 19.56402 12.93403 0.0003272084 0.008580533    11 0.008109758
## 26  -7.480463 20.04784 12.81098 0.0003493595 0.008580533    26 0.008109758
## 44  -8.287875 20.48628 14.64008 0.0001324344 0.008580533    44 0.008109758
## 65   7.246433 19.67782 12.65635 0.0003793550 0.008580533    65 0.008109758
## 138  7.540027 19.63237 13.67156 0.0002211329 0.008580533   138 0.008109758
## 139 -7.396184 20.20320 12.51051 0.0004100255 0.008580533   139 0.008109758</code></pre>
</div>
</div>
<div id="inspecting-da-testing-results" class="section level1">
<h1 class="hasAnchor">
<a href="#inspecting-da-testing-results" class="anchor"></a>Inspecting DA testing results</h1>
<p>We can start inspecting the results of our DA analysis from a couple of standard diagnostic plots. We first inspect the distribution of uncorrected P values, to verify that the test was balanced.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="fu">ggplot</span>(<span class="no">da_results</span>, <span class="fu">aes</span>(<span class="no">PValue</span>)) + <span class="fu">geom_histogram</span>(<span class="kw">bins</span><span class="kw">=</span><span class="fl">50</span>)</pre></body></html></div>
<p><img src="milo_gastrulation_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>Then we visualize the test results with a volcano plot (remember that each point here represents a neighbourhood, <em>not</em> a cell).</p>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="fu">ggplot</span>(<span class="no">da_results</span>, <span class="fu">aes</span>(<span class="no">logFC</span>, -<span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="no">SpatialFDR</span>))) +
  <span class="fu">geom_point</span>() +
  <span class="fu">geom_hline</span>(<span class="kw">yintercept</span> <span class="kw">=</span> <span class="fl">1</span>) <span class="co">## Mark significance threshold (10% FDR)</span></pre></body></html></div>
<p><img src="milo_gastrulation_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>Looks like we have detected several neighbourhoods were there is a significant difference in cell abundances between developmental stages.</p>
<p>To visualize DA results relating them to the embedding of single cells, we can build an abstracted graph of neighbourhoods that we can superimpose on the single-cell embedding. Here each node represents a neighbourhood, while edges indicate how many cells two neighbourhoods have in common. Here the layout of nodes is determined by the position of the index cell in the UMAP embedding of all single-cells. The neighbourhoods displaying singificant DA are colored by their log-Fold Change.</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="no">embryo_milo</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/buildNhoodGraph.html">buildNhoodGraph</a></span>(<span class="no">embryo_milo</span>)

<span class="co">## Plot single-cell UMAP</span>
<span class="no">umap_pl</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span>(<span class="no">embryo_milo</span>, <span class="kw">dimred</span> <span class="kw">=</span> <span class="st">"umap"</span>, <span class="kw">colour_by</span><span class="kw">=</span><span class="st">"celltype"</span>, <span class="kw">text_by</span> <span class="kw">=</span> <span class="st">"celltype"</span>, <span class="kw">text_size</span> <span class="kw">=</span> <span class="fl">3</span>) +
  <span class="fu">guides</span>(<span class="kw">fill</span><span class="kw">=</span><span class="st">"none"</span>)

<span class="co">## Plot neighbourhood graph</span>
<span class="no">nh_graph_pl</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/plotNhoodGraphDA.html">plotNhoodGraphDA</a></span>(<span class="no">embryo_milo</span>, <span class="no">da_results</span>, <span class="kw">layout</span><span class="kw">=</span><span class="st">"umap"</span>,<span class="kw">alpha</span><span class="kw">=</span><span class="fl">0.05</span>)

<span class="no">umap_pl</span> + <span class="no">nh_graph_pl</span> +
  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span>(<span class="kw">guides</span><span class="kw">=</span><span class="st">"collect"</span>)</pre></body></html></div>
<p><img src="milo_gastrulation_files/figure-html/unnamed-chunk-16-1.png" width="1440"></p>
<p>We might also be interested in visualizing wheather DA is particularly evident in certain cell types. To do this, we assign a cell type label to each neighbourhood by finding the most abundant cell type within cells in each neighbourhood. We can label neighbourhoods in the results <code>data.frame</code> using the function <code>annotateNhoods</code>. This also saves the fraction of cells harbouring the label.</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="no">da_results</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/annotateNhoods.html">annotateNhoods</a></span>(<span class="no">embryo_milo</span>, <span class="no">da_results</span>, <span class="kw">coldata_col</span> <span class="kw">=</span> <span class="st">"celltype"</span>)
<span class="no">da_results</span></pre></body></html></div>
<pre><code>##           logFC   logCPM            F       PValue         FDR Nhood
## 1    4.56995163 19.83133 5.974100e+00 1.456749e-02 0.046812378     1
## 2   -2.47964941 20.32636 2.840380e+00 9.201339e-02 0.153296087     2
## 3   -2.32839718 20.09926 2.439441e+00 1.184106e-01 0.188141250     3
## 4   -1.98319255 19.98213 1.549825e+00 2.132462e-01 0.284057709     4
## 5   -0.57738664 19.45406 1.742434e-01 6.763939e-01 0.732760032     5
## 6    2.42924448 19.34933 2.821000e+00 9.312943e-02 0.154555223     6
## 7    1.23211902 20.34733 6.048114e-01 4.368017e-01 0.516220226     7
## 8    1.86195662 19.46433 1.908006e+00 1.672749e-01 0.240415292     8
## 9   -0.91699737 20.34798 3.899184e-01 5.323838e-01 0.607427238     9
## 10  -2.49959967 19.91675 3.128454e+00 7.702547e-02 0.135097888    10
## 11   7.33791587 19.56402 1.293403e+01 3.272084e-04 0.008580533    11
## 12   6.06309745 20.56560 8.103883e+00 4.443284e-03 0.023105078    12
## 13  -2.57638793 20.15233 2.949785e+00 8.597919e-02 0.144931530    13
## 14  -2.30864017 20.88676 2.095521e+00 1.478227e-01 0.220808574    14
## 15  -1.99446894 20.18261 1.911355e+00 1.669028e-01 0.240415292    15
## 16   2.84842324 19.27326 4.170392e+00 4.121257e-02 0.087525707    16
## 17  -3.52610682 19.98721 5.110783e+00 2.383997e-02 0.061460263    17
## 18  -2.94931722 20.64994 3.483052e+00 6.208550e-02 0.116818773    18
## 19   0.06319834 19.77200 2.098112e-03 9.634682e-01 0.970681111    19
## 20  -1.02687015 19.85519 5.707908e-01 4.499967e-01 0.526017946    20
## 21   2.10230541 20.23459 1.445534e+00 2.293289e-01 0.301785517    21
## 22  -2.09134919 20.31934 2.091260e+00 1.482351e-01 0.220808574    22
## 23  -2.52960926 19.69929 2.953869e+00 8.576241e-02 0.144850678    23
## 24   3.28061130 19.75816 3.434956e+00 6.391605e-02 0.118685887    24
## 25   5.17328533 19.89927 7.594009e+00 5.887181e-03 0.026876670    25
## 26  -7.48046310 20.04784 1.281098e+01 3.493595e-04 0.008580533    26
## 27  -0.86230391 19.59364 4.571946e-01 4.989836e-01 0.577770432    27
## 28  -3.67346274 19.49327 6.210807e+00 1.274411e-02 0.043219147    28
## 29   3.94377412 20.17261 4.671055e+00 3.074403e-02 0.074095057    29
## 30   0.44407217 20.37756 8.500117e-02 7.706486e-01 0.813304458    30
## 31  -6.78374855 20.63472 1.092711e+01 9.574141e-04 0.010808701    31
## 32  -2.33855027 19.38758 3.023853e+00 8.213799e-02 0.140974245    32
## 33  -2.92652852 20.11169 3.621685e+00 5.711467e-02 0.110075177    33
## 34   3.19456952 19.78358 4.387644e+00 3.627346e-02 0.082334990    34
## 35  -3.60260143 20.76621 4.747927e+00 2.940141e-02 0.071838769    35
## 36   2.31579273 19.86861 1.400517e+00 2.367190e-01 0.308707373    36
## 37   3.16946549 19.40642 3.395369e+00 6.546591e-02 0.119764935    37
## 38   1.09922943 19.36372 6.742765e-01 4.116216e-01 0.490515776    38
## 39   0.26083377 20.25515 3.334382e-02 8.551204e-01 0.888248550    39
## 40   4.18339632 19.12120 6.463638e+00 1.105398e-02 0.038870151    40
## 41   1.95927246 19.85042 1.849896e+00 1.738848e-01 0.247418147    41
## 42  -5.59024059 20.29044 9.394292e+00 2.193546e-03 0.016765965    42
## 43  -4.81617126 19.61913 8.885763e+00 2.894211e-03 0.019716613    43
## 44  -8.28787539 20.48628 1.464008e+01 1.324344e-04 0.008580533    44
## 45  -1.10312181 19.50629 6.921505e-01 4.054918e-01 0.485232834    45
## 46  -1.73114403 19.28018 1.778270e+00 1.824504e-01 0.255237430    46
## 47  -3.17983124 19.21452 5.357076e+00 2.069709e-02 0.056374928    47
## 48  -1.52238333 19.92462 1.071480e+00 3.006844e-01 0.378280367    48
## 49  -4.00775979 19.27602 7.919526e+00 4.918266e-03 0.024388207    49
## 50   3.25859215 19.71063 4.304079e+00 3.809543e-02 0.084517098    50
## 51  -1.32375635 19.15820 1.093678e+00 2.957315e-01 0.373143531    51
## 52   0.13267109 20.13239 7.761856e-03 9.298012e-01 0.945224469    52
## 53   3.81161381 19.86450 5.558571e+00 1.844625e-02 0.053112786    53
## 54  -0.33355601 20.35539 5.562088e-02 8.135703e-01 0.852311729    54
## 55  -0.23667453 20.04985 2.724977e-02 8.688946e-01 0.898206748    55
## 56   1.57542685 19.31478 1.100197e+00 2.942968e-01 0.371880147    56
## 57   2.27820012 19.73813 2.334512e+00 1.266268e-01 0.196466192    57
## 58  -1.08658266 20.19447 5.798643e-01 4.464178e-01 0.523976068    58
## 59  -6.72013308 20.50886 1.119035e+01 8.310757e-04 0.010320258    59
## 60  -4.43409639 19.41362 7.407837e+00 6.526827e-03 0.028571519    60
## 61   3.11310734 20.09125 3.677571e+00 5.523203e-02 0.108704173    61
## 62  -3.17918685 20.26928 4.403275e+00 3.594291e-02 0.082334990    62
## 63   0.43202596 20.37052 6.951342e-02 7.920624e-01 0.832830303    63
## 64  -3.90658303 19.61008 5.500311e+00 1.906976e-02 0.053485316    64
## 65   7.24643265 19.67782 1.265635e+01 3.793550e-04 0.008580533    65
## 66  -2.01061635 19.30080 2.436503e+00 1.186324e-01 0.188145333    66
## 67  -0.21820505 19.64945 2.475700e-02 8.749833e-01 0.901243343    67
## 68  -0.16641157 19.90809 1.331438e-02 9.081444e-01 0.928710237    68
## 69  -4.66046933 19.76472 9.046009e+00 2.651791e-03 0.018960308    69
## 70   3.26485664 19.53139 4.661456e+00 3.091612e-02 0.074095057    70
## 71  -3.12128390 20.42480 4.069640e+00 4.373829e-02 0.091530381    71
## 72   2.84728862 19.15529 3.139281e+00 7.651620e-02 0.134530532    72
## 73   3.76416906 19.39575 5.807504e+00 1.601023e-02 0.049313775    73
## 74   0.78063158 19.86736 2.789671e-01 5.974132e-01 0.666015009    74
## 75  -3.67518763 20.27760 5.263158e+00 2.184159e-02 0.058281510    75
## 76   0.54025509 19.41109 1.801582e-01 6.712639e-01 0.729043630    76
## 77   4.57410025 19.49217 6.863584e+00 8.835740e-03 0.033396760    77
## 78   0.41846882 19.54047 9.769429e-02 7.546333e-01 0.801330921    78
## 79  -1.83638702 19.98769 1.682970e+00 1.946174e-01 0.268028495    79
## 80  -4.21264183 20.79568 5.593154e+00 1.808613e-02 0.052888275    80
## 81  -3.92720991 19.84314 5.741425e+00 1.662257e-02 0.049693958    81
## 82  -1.83851063 19.82052 1.811773e+00 1.783846e-01 0.250497587    82
## 83   1.66752316 19.71209 1.211867e+00 2.710400e-01 0.348132167    83
## 84   4.12389555 19.14505 6.305288e+00 1.208344e-02 0.041636907    84
## 85  -5.68442950 19.58786 1.145975e+01 7.191854e-04 0.009232713    85
## 86  -2.61212294 20.81046 2.724775e+00 9.889354e-02 0.162574071    86
## 87   7.00617152 19.46943 1.171910e+01 6.258646e-04 0.009232713    87
## 88   5.32261949 19.51469 8.528592e+00 3.518881e-03 0.020971928    88
## 89  -2.72966220 19.84305 3.230285e+00 7.237599e-02 0.129372083    89
## 90  -4.36834254 20.10816 6.732507e+00 9.507431e-03 0.034860582    90
## 91  -3.88330522 20.45179 5.523546e+00 1.881852e-02 0.053112786    91
## 92   0.03983333 20.02578 7.986994e-04 9.774554e-01 0.979168312    92
## 93   5.21720973 19.49196 8.153607e+00 4.323366e-03 0.023093273    93
## 94  -3.92799243 20.81215 5.429137e+00 1.986116e-02 0.054732935    94
## 95  -3.88857315 19.76103 7.008771e+00 8.148220e-03 0.032069599    95
## 96   4.00235154 20.09935 5.265308e+00 2.181466e-02 0.058281510    96
## 97   2.84420576 19.95572 3.209570e+00 7.329681e-02 0.129983395    97
## 98  -2.52840648 19.95166 3.063335e+00 8.016652e-02 0.138396121    98
## 99   0.12613097 19.62620 1.015064e-02 9.197545e-01 0.939463509    99
## 100  1.12996601 20.35680 6.025443e-01 4.376624e-01 0.516525905   100
## 101 -0.65796475 20.60527 1.607878e-01 6.884572e-01 0.743014207   101
## 102 -3.98824988 19.47807 5.589624e+00 1.812256e-02 0.052888275   102
## 103 -5.35411289 20.68832 8.446133e+00 3.681623e-03 0.020971928   103
## 104 -3.28890715 19.73646 4.425639e+00 3.547545e-02 0.081603037   104
## 105  2.08203116 19.93202 1.900682e+00 1.680919e-01 0.241175346   105
## 106  3.26626182 20.62655 3.556788e+00 5.938675e-02 0.113230739   106
## 107  3.57287318 19.73928 4.396962e+00 3.607601e-02 0.082334990   107
## 108  3.04049304 20.10077 3.396421e+00 6.542422e-02 0.119764935   108
## 109 -0.05973637 19.13448 1.780421e-03 9.663457e-01 0.972010065   109
## 110  4.15074484 19.65841 5.983785e+00 1.448785e-02 0.046812378   110
## 111 -4.66979307 19.33705 8.432772e+00 3.708705e-03 0.020971928   111
## 112 -3.26689132 19.64948 5.216822e+00 2.243020e-02 0.059034077   112
## 113 -3.13034018 19.83534 4.238559e+00 3.959101e-02 0.085997682   113
## 114 -2.18359722 20.18469 2.112744e+00 1.461686e-01 0.218870237   114
## 115 -3.37824361 19.93166 4.789633e+00 2.869870e-02 0.070757128   115
## 116  2.01204146 19.86563 1.734292e+00 1.879517e-01 0.262215476   116
## 117 -1.41602067 20.00765 1.003369e+00 3.165673e-01 0.393074894   117
## 118 -5.27310454 19.30608 1.022393e+01 1.398903e-03 0.013412888   118
## 119  6.07591023 19.87952 9.430469e+00 2.150809e-03 0.016625174   119
## 120  1.21939449 19.93115 7.394367e-01 3.899014e-01 0.468537024   120
## 121  4.55668857 19.05468 6.963719e+00 8.355491e-03 0.032735212   121
## 122 -4.65797811 19.92358 7.155590e+00 7.508531e-03 0.030884633   122
## 123 -2.18978634 19.34287 2.717716e+00 9.933148e-02 0.162956806   123
## 124  0.17891144 19.92247 1.560636e-02 9.005899e-01 0.924289634   124
## 125 -2.25910310 20.28064 2.296145e+00 1.297876e-01 0.200283775   125
## 126 -4.04791371 19.32800 7.319318e+00 6.855440e-03 0.029152720   126
## 127 -2.76931021 19.64449 3.889414e+00 4.867155e-02 0.098578964   127
## 128 -3.27879320 19.77964 4.981409e+00 2.568542e-02 0.065201457   128
## 129 -3.41392854 19.86650 4.579927e+00 3.241891e-02 0.077051027   129
## 130  3.12942213 19.91943 2.643532e+00 1.040640e-01 0.169103999   130
## 131  5.75019370 19.70418 9.881560e+00 1.683681e-03 0.014507189   131
## 132  3.76625368 19.67670 5.043749e+00 2.477832e-02 0.063273219   132
## 133  0.62871125 19.30603 2.583985e-01 6.112552e-01 0.675846591   133
## 134  1.08981739 19.91667 5.714444e-01 4.497374e-01 0.526017946   134
## 135  4.48846923 19.58235 6.934483e+00 8.492878e-03 0.033122226   135
## 136  2.72440502 19.79801 2.173842e+00 1.404669e-01 0.212184160   136
## 137  2.03071116 20.01343 1.634670e+00 2.011447e-01 0.273939953   137
## 138  7.54002707 19.63237 1.367156e+01 2.211329e-04 0.008580533   138
## 139 -7.39618382 20.20320 1.251051e+01 4.100255e-04 0.008580533   139
## 140  3.52049887 19.81817 4.836698e+00 2.792666e-02 0.069452399   140
## 141 -1.70770153 19.08745 1.564031e+00 2.111613e-01 0.282206214   141
## 142 -1.14310236 19.46294 4.534813e-01 5.007317e-01 0.578233897   142
## 143 -3.45141104 19.73856 5.934432e+00 1.489840e-02 0.047697109   143
## 144  1.71748058 20.56586 1.063468e+00 3.024982e-01 0.379615380   144
## 145 -4.50353723 20.48492 7.092495e+00 7.776892e-03 0.031474403   145
## 146  1.78670527 20.10735 1.280890e+00 2.578128e-01 0.334144025   146
## 147 -1.68717232 19.68313 1.679871e+00 1.950288e-01 0.268164540   147
## 148  2.87208597 19.42979 3.434521e+00 6.393287e-02 0.118685887   148
## 149  0.96333216 20.20361 4.296400e-01 5.122096e-01 0.587534533   149
## 150 -3.39221621 18.92144 4.885176e+00 2.715402e-02 0.067758086   150
## 151  5.38603877 19.24733 9.643264e+00 1.915978e-03 0.015508575   151
## 152 -3.10656605 20.06323 4.301427e+00 3.815481e-02 0.084517098   152
## 153 -0.89901377 20.89290 3.264005e-01 5.678238e-01 0.638522663   153
## 154 -5.66091970 19.80774 1.009669e+01 1.498550e-03 0.013825330   154
## 155  1.01691084 19.46805 3.911491e-01 5.317376e-01 0.607427238   155
## 156 -2.10400748 20.39312 2.118689e+00 1.456025e-01 0.218403723   156
## 157 -1.61162575 19.29363 1.608678e+00 2.047637e-01 0.276672869   157
## 158 -3.24645844 20.05205 4.540041e+00 3.318169e-02 0.077999708   158
## 159  6.41047761 19.20771 1.081566e+01 1.016605e-03 0.011182657   159
## 160 -2.61907866 19.80559 2.729929e+00 9.857514e-02 0.162574071   160
## 161  0.07003897 20.23213 2.457134e-03 9.604683e-01 0.969616197   161
## 162  7.57124087 19.75426 1.306745e+01 3.047888e-04 0.008580533   162
## 163 -2.30403927 20.20853 2.422587e+00 1.196895e-01 0.189122630   163
## 164 -4.29233525 19.21286 7.593432e+00 5.889061e-03 0.026876670   164
## 165 -4.48885149 19.33696 8.137550e+00 4.361725e-03 0.023093273   165
## 166  1.93546598 19.88424 1.692577e+00 1.933490e-01 0.267139182   166
## 167  1.99795936 20.47698 1.625426e+00 2.024232e-01 0.274963821   167
## 168  4.16016367 19.49682 6.013154e+00 1.424909e-02 0.046309555   168
## 169  4.83060601 19.64257 7.738072e+00 5.436385e-03 0.025913437   169
## 170 -1.88679872 19.48694 2.072582e+00 1.500586e-01 0.222366683   170
## 171  6.98068099 19.23362 1.271453e+01 3.677757e-04 0.008580533   171
## 172 -1.49886361 19.77550 9.779295e-01 3.227802e-01 0.397910099   172
## 173  5.57737930 19.64146 9.612825e+00 1.947905e-03 0.015619648   173
## 174 -3.17129535 20.57331 3.943092e+00 4.714363e-02 0.096768508   174
## 175 -5.46265377 20.59640 8.453101e+00 3.667580e-03 0.020971928   175
## 176  0.88204898 19.61135 3.650432e-01 5.457585e-01 0.619392623   176
## 177 -3.21970653 19.69762 4.561719e+00 3.276481e-02 0.077444100   177
## 178 -2.95868708 19.69462 4.670288e+00 3.075774e-02 0.074095057   178
## 179  2.96799061 19.90220 3.677362e+00 5.523895e-02 0.108704173   179
## 180  5.13583728 19.29530 8.637922e+00 3.314324e-03 0.020698383   180
## 181 -6.02480564 19.73673 1.234971e+01 4.467534e-04 0.008819405   181
## 182 -0.73841837 19.76390 2.693758e-01 6.037842e-01 0.668863978   182
## 183 -2.53566692 19.67142 3.663925e+00 5.568559e-02 0.108834256   183
## 184 -2.85542933 19.74296 4.431977e+00 3.534412e-02 0.081583245   184
## 185 -1.81226712 19.74270 1.599775e+00 2.060210e-01 0.277497733   185
## 186 -4.94367276 20.54511 8.264784e+00 4.067024e-03 0.022226155   186
## 187 -6.19511196 20.44501 1.098762e+01 9.267515e-04 0.010602037   187
## 188 -4.69330959 19.48088 6.924437e+00 8.540621e-03 0.033156322   188
## 189 -3.97673947 20.88162 5.316989e+00 2.117781e-02 0.057320384   189
## 190  7.29854923 19.65435 1.298434e+01 3.185656e-04 0.008580533   190
## 191 -4.45050547 19.88862 6.872690e+00 8.790923e-03 0.033396760   191
## 192  4.75130114 19.65406 7.410582e+00 6.516897e-03 0.028571519   192
## 193 -2.91365610 19.38438 4.210517e+00 4.024981e-02 0.086335838   193
## 194 -4.98045604 19.78608 8.421072e+00 3.732586e-03 0.020971928   194
## 195  3.34663530 19.90248 3.082013e+00 7.925172e-02 0.137647716   195
## 196 -3.57256014 20.19552 5.428361e+00 1.986997e-02 0.054732935   196
## 197  4.99770461 19.11473 8.536103e+00 3.504425e-03 0.020971928   197
## 198  6.70156875 19.29377 1.156574e+01 6.794580e-04 0.009232713   198
## 199  1.72491431 19.75418 1.260156e+00 2.617001e-01 0.337790082   199
## 200 -6.04255797 20.41821 1.002714e+01 1.556031e-03 0.014076641   200
## 201 -4.61928947 19.48993 6.780622e+00 9.254994e-03 0.034503328   201
## 202  4.08881458 19.54428 6.232015e+00 1.259264e-02 0.043045762   202
## 203 -3.60008564 20.38600 5.302048e+00 2.135992e-02 0.057631488   203
## 204  3.60075888 19.89452 3.759805e+00 5.258058e-02 0.104916608   204
## 205 -0.38136296 19.52476 7.980205e-02 7.775829e-01 0.819614393   205
## 206 -3.97361902 20.11839 6.044493e+00 1.399878e-02 0.045865495   206
## 207  2.37713798 19.90141 2.238473e+00 1.347058e-01 0.205288829   207
## 208  1.70676586 19.62413 1.624603e+00 2.025375e-01 0.274963821   208
## 209 -3.55484633 20.53623 4.934693e+00 2.638774e-02 0.066200818   209
## 210 -6.04836411 20.00516 1.038169e+01 1.284613e-03 0.013279491   210
## 211  3.93741714 19.71079 5.066757e+00 2.445201e-02 0.062626339   211
## 212 -3.48096292 19.49421 5.903797e+00 1.515924e-02 0.048004723   212
## 213 -0.60402063 19.60764 2.072224e-01 6.489820e-01 0.709333194   213
## 214  2.72928982 19.81639 1.968613e+00 1.606857e-01 0.234869340   214
## 215 -4.93397087 20.25997 8.340958e+00 3.900369e-03 0.021709520   215
## 216 -3.96050646 19.99784 4.550225e+00 3.298517e-02 0.077750762   216
## 217 -6.58647575 20.14913 1.154074e+01 6.886227e-04 0.009232713   217
## 218 -0.34538210 19.94366 6.259654e-02 8.024531e-01 0.841692831   218
## 219 -0.73338402 19.68724 3.121075e-01 5.764262e-01 0.647347796   219
## 220 -1.74216577 19.59671 1.841791e+00 1.748305e-01 0.247532222   220
## 221 -0.10085837 20.18842 4.409028e-03 9.470628e-01 0.958945839   221
## 222 -2.80388048 19.53726 4.184150e+00 4.087977e-02 0.087202723   222
## 223  1.90339956 20.02019 1.621059e+00 2.030304e-01 0.275197671   223
## 224 -3.84865156 19.69373 6.600919e+00 1.023449e-02 0.036741406   224
## 225 -3.97123724 19.53830 7.335035e+00 6.795883e-03 0.029152720   225
## 226 -2.05627603 19.66883 2.513826e+00 1.129438e-01 0.181450231   226
## 227  1.60666441 20.20149 8.824752e-01 3.475908e-01 0.423626230   227
## 228  6.96900832 19.15089 1.269578e+01 3.714673e-04 0.008580533   228
## 229  7.13252438 19.31383 1.279522e+01 3.523040e-04 0.008580533   229
## 230 -3.81574177 19.74987 6.803846e+00 9.135616e-03 0.034246464   230
## 231  4.75095871 19.72563 7.495094e+00 6.218636e-03 0.027935025   231
## 232 -4.59762500 20.54248 7.037962e+00 8.016737e-03 0.031844261   232
## 233 -4.36461342 20.27269 6.275522e+00 1.228769e-02 0.042171362   233
## 234 -2.90034499 19.12434 3.618658e+00 5.721857e-02 0.110075177   234
## 235 -4.44992050 19.99661 6.555845e+00 1.049645e-02 0.037061549   235
## 236  1.94605493 20.66301 1.501239e+00 2.205649e-01 0.292948419   236
## 237 -3.17493000 19.65863 5.257033e+00 2.191847e-02 0.058281510   237
## 238 -4.41600255 20.17241 5.468883e+00 1.941511e-02 0.053909909   238
## 239 -4.47562445 19.75061 8.458847e+00 3.656041e-03 0.020971928   239
## 240 -3.22922194 20.15742 4.228598e+00 3.982370e-02 0.086085903   240
## 241 -0.55145346 20.14468 1.487425e-01 6.997639e-01 0.751436051   241
## 242 -0.80007429 19.36251 3.505150e-01 5.538608e-01 0.626929468   242
## 243  1.83385189 19.44432 1.603536e+00 2.054888e-01 0.277216044   243
## 244 -2.81496912 19.90435 3.349815e+00 6.729925e-02 0.122336357   244
## 245  3.50282048 19.34962 4.118427e+00 4.249552e-02 0.089365574   245
## 246 -3.45515921 20.07336 4.954533e+00 2.608706e-02 0.066025668   246
## 247  1.35191250 19.53597 9.126282e-01 3.394853e-01 0.415518429   247
## 248 -7.58204122 20.00448 1.363201e+01 2.258214e-04 0.008580533   248
## 249 -3.06219434 20.40320 3.930233e+00 4.750499e-02 0.097277523   249
## 250 -2.48911565 20.08888 2.867448e+00 9.047883e-02 0.151496301   250
## 251 -2.87486290 19.51535 4.361550e+00 3.683238e-02 0.082728217   251
## 252  6.89400472 19.35688 1.183461e+01 5.883343e-04 0.009232713   252
## 253 -0.27425413 20.67682 3.008578e-02 8.623060e-01 0.894629416   253
## 254  5.02815731 19.13675 8.631645e+00 3.325734e-03 0.020698383   254
## 255  1.39805989 19.14465 1.062886e+00 3.026304e-01 0.379615380   255
## 256 -1.28955928 19.82703 8.692878e-01 3.512181e-01 0.427439957   256
## 257  2.09690071 19.53558 1.895500e+00 1.686727e-01 0.241604625   257
## 258 -3.43021852 20.79600 4.305737e+00 3.805837e-02 0.084517098   258
## 259  4.04173853 19.45030 5.486048e+00 1.922569e-02 0.053731734   259
## 260 -3.89307396 19.90334 5.249051e+00 2.201908e-02 0.058309789   260
## 261  2.00519905 19.64686 2.211556e+00 1.370723e-01 0.208524885   261
## 262 -4.90963188 19.85931 8.122885e+00 4.397062e-03 0.023093273   262
## 263  0.93787135 20.61617 3.778668e-01 5.387879e-01 0.613917644   263
## 264 -6.74930819 19.52915 1.296893e+01 3.211886e-04 0.008580533   264
## 265 -1.91389599 19.96750 1.716239e+00 1.902656e-01 0.263896106   265
## 266  4.96398385 19.20626 8.330941e+00 3.921883e-03 0.021709520   266
## 267  0.01086739 19.45937 1.010680e-04 9.919794e-01 0.991979381   267
## 268  5.78146448 19.73796 9.764516e+00 1.793977e-03 0.015090514   268
## 269 -0.92339581 19.52024 5.200614e-01 4.708646e-01 0.549393080   269
## 270 -2.91226026 19.42869 3.992694e+00 4.577671e-02 0.095331113   270
## 271 -1.99093053 19.96621 1.879324e+00 1.705006e-01 0.243815895   271
## 272 -4.37529989 19.64419 8.482343e+00 3.609237e-03 0.020971928   272
## 273 -2.72154597 20.72672 2.865649e+00 9.057996e-02 0.151496301   273
## 274 -7.12966521 19.74051 1.405805e+01 1.801759e-04 0.008580533   274
## 275  4.67275709 19.78620 6.699808e+00 9.683008e-03 0.035203477   275
## 276 -1.39538269 19.87070 9.645600e-01 3.261099e-01 0.400289376   276
## 277 -4.62561185 20.15685 7.046124e+00 7.980364e-03 0.031844261   277
## 278 -5.02239579 19.55781 8.901961e+00 2.868716e-03 0.019716613   278
## 279 -3.78770971 19.25692 6.341603e+00 1.183896e-02 0.041124805   279
## 280  3.11708049 19.35933 4.283364e+00 3.856177e-02 0.084618927   280
## 281  0.67812018 19.28696 2.869819e-01 5.921958e-01 0.662456371   281
## 282  3.26855328 19.59718 3.792442e+00 5.156594e-02 0.103372848   282
## 283 -3.63084790 19.20541 5.549948e+00 1.853718e-02 0.053112786   283
## 284 -6.40414497 19.39478 1.104618e+01 8.980216e-04 0.010412196   284
## 285 -2.75935251 19.97166 3.558371e+00 5.933022e-02 0.113230739   285
## 286 -2.84478444 19.90608 2.533163e+00 1.115686e-01 0.179728155   286
## 287 -3.74631810 18.93113 5.692338e+00 1.709305e-02 0.050923058   287
## 288 -5.65721853 19.53375 1.110874e+01 8.683260e-04 0.010320258   288
## 289 -1.26889156 19.92824 8.179566e-01 3.658418e-01 0.443000392   289
## 290 -2.80650927 19.92527 3.682456e+00 5.507063e-02 0.108704173   290
## 291 -4.71168602 19.18850 7.891092e+00 4.996000e-03 0.024388207   291
## 292 -0.51933881 19.31568 1.642373e-01 6.853102e-01 0.740549263   292
## 293 -1.36537640 19.38772 7.648830e-01 3.818659e-01 0.459524451   293
## 294  3.41906063 19.95016 4.180870e+00 4.095885e-02 0.087202723   294
## 295  3.67681034 19.33234 5.797453e+00 1.610185e-02 0.049313775   295
## 296 -2.61565162 19.74032 2.977471e+00 8.452109e-02 0.143623196   296
## 297 -2.39113074 19.14626 3.428028e+00 6.418444e-02 0.118685887   297
## 298  2.22532604 20.06631 1.924028e+00 1.655034e-01 0.239463627   298
## 299 -3.61720793 20.36585 5.425466e+00 1.990289e-02 0.054732935   299
## 300  2.34989242 20.64465 2.055645e+00 1.517341e-01 0.224461876   300
## 301 -1.61697322 20.25013 1.109198e+00 2.923304e-01 0.370486683   301
## 302 -1.18656647 20.07652 6.968641e-01 4.038975e-01 0.484000083   302
## 303 -6.05301447 19.65973 1.300487e+01 3.151056e-04 0.008580533   303
## 304 -0.13173000 19.38142 8.620344e-03 9.260315e-01 0.944538307   304
## 305 -2.28836103 19.94405 2.443357e+00 1.181156e-01 0.188141250   305
## 306  2.97694845 20.18758 3.467814e+00 6.265934e-02 0.117383656   306
## 307 -1.98707722 20.13625 1.907942e+00 1.672820e-01 0.240415292   307
## 308 -3.49296991 19.53870 5.896879e+00 1.521880e-02 0.048006364   308
## 309 -0.46287366 19.00714 9.014417e-02 7.640121e-01 0.809286928   309
## 310 -0.18540440 19.95363 1.872537e-02 8.911648e-01 0.915711821   310
## 311  7.51378186 19.72687 1.315441e+01 2.910150e-04 0.008580533   311
## 312 -1.50161389 20.26580 1.248412e+00 2.639342e-01 0.340023398   312
## 313  6.89113291 19.30331 1.215323e+01 4.961718e-04 0.008907606   313
## 314 -3.46041851 19.56891 5.744033e+00 1.659795e-02 0.049693958   314
## 315  2.65246601 20.27729 2.538857e+00 1.111672e-01 0.179626156   315
## 316 -3.30694479 19.35643 5.499812e+00 1.907518e-02 0.053485316   316
## 317  1.96845857 19.53917 1.838101e+00 1.752629e-01 0.247735728   317
## 318 -1.69515090 20.17689 1.388189e+00 2.387928e-01 0.310901751   318
## 319  3.78461423 19.68181 4.663994e+00 3.087052e-02 0.074095057   319
## 320 -2.58766961 19.70350 3.647330e+00 5.624247e-02 0.109672816   320
## 321 -5.68639588 19.93028 1.002409e+01 1.558602e-03 0.014076641   321
## 322 -3.37926289 20.21357 4.480563e+00 3.435420e-02 0.079885323   322
## 323  3.32744442 19.10356 5.100514e+00 2.398129e-02 0.061604633   323
## 324 -2.52693195 20.04574 2.993956e+00 8.366567e-02 0.142714009   324
## 325  3.74990945 19.83442 4.134862e+00 4.208529e-02 0.088983760   325
## 326 -1.39545763 19.44592 1.023595e+00 3.117390e-01 0.388767538   326
## 327  3.25475677 20.31341 3.828603e+00 5.046590e-02 0.101404544   327
## 328 -0.65372228 19.46494 2.445671e-01 6.209585e-01 0.683054309   328
## 329  3.78132514 19.39605 5.829228e+00 1.581403e-02 0.049313775   329
## 330 -1.28719431 20.22260 8.094445e-01 3.683477e-01 0.445129986   330
## 331 -0.82831319 19.58078 4.596185e-01 4.978480e-01 0.577234593   331
## 332  0.63818848 19.95000 1.973010e-01 6.569363e-01 0.714386974   332
## 333  1.26217867 20.20237 6.651063e-01 4.148196e-01 0.493641052   333
## 334 -3.42114810 19.53761 6.057010e+00 1.390008e-02 0.045865495   334
## 335  0.18195858 20.81155 1.444912e-02 9.043282e-01 0.925911192   335
## 336 -5.62635035 19.77586 1.017267e+01 1.438212e-03 0.013412888   336
## 337 -1.26909545 19.40796 4.722134e-01 4.920171e-01 0.572019895   337
## 338  2.35368501 19.19914 2.462617e+00 1.166766e-01 0.186421760   338
## 339  0.77975707 20.17849 2.004756e-01 6.543655e-01 0.713450227   339
## 340  1.90734169 19.83316 1.588339e+00 2.076493e-01 0.278501294   340
## 341 -6.58988976 19.35029 1.273290e+01 3.641949e-04 0.008580533   341
## 342 -2.61865189 20.13544 1.945449e+00 1.631683e-01 0.237285489   342
## 343  0.75399435 20.27806 2.465526e-01 6.195449e-01 0.682566036   343
## 344  1.76231400 20.43101 1.209169e+00 2.715738e-01 0.348296448   344
## 345 -2.86497299 21.02878 2.935341e+00 8.675068e-02 0.145659658   345
## 346 -4.00575374 20.50016 5.546537e+00 1.857328e-02 0.053112786   346
## 347  0.43630310 19.23830 1.166309e-01 7.327386e-01 0.780980992   347
## 348 -1.10579519 19.84288 7.457483e-02 7.848051e-01 0.826212033   348
## 349  2.82519598 19.78017 3.539292e+00 6.001563e-02 0.114176082   349
## 350 -0.76631096 19.28409 2.463074e-01 6.197190e-01 0.682566036   350
## 351  1.39720045 19.61280 1.142954e+00 2.851042e-01 0.362936780   351
## 352 -1.75117942 19.81915 1.675888e+00 1.955589e-01 0.268463220   352
## 353  2.96483227 19.37687 3.125738e+00 7.715381e-02 0.135097888   353
## 354  1.71836628 18.82361 9.750558e-01 3.234921e-01 0.398215543   354
## 355 -6.30305922 19.24681 1.108799e+01 8.780639e-04 0.010320258   355
## 356  7.15962953 19.49054 1.216224e+01 4.937909e-04 0.008907606   356
## 357 -1.57642414 20.04693 9.019287e-01 3.423320e-01 0.417810559   357
## 358 -3.65292105 20.19836 5.552063e+00 1.851483e-02 0.053112786   358
## 359  3.46217574 19.76891 3.915327e+00 4.792751e-02 0.097909065   359
## 360 -0.34721387 20.07261 4.349886e-02 8.348011e-01 0.871361784   360
## 361 -1.93542888 19.84228 1.685991e+00 1.942175e-01 0.267907747   361
## 362  3.28686750 19.50456 4.379358e+00 3.644995e-02 0.082517309   362
## 363  3.54846805 19.76163 4.125048e+00 4.232975e-02 0.089235691   363
## 364  0.65361654 20.34594 1.685238e-01 6.814527e-01 0.737309470   364
## 365 -1.85096850 19.79932 1.818843e+00 1.775400e-01 0.249720206   365
## 366  3.79645855 19.90430 5.129921e+00 2.357889e-02 0.061305112   366
## 367 -1.60525286 20.04373 1.263728e+00 2.610253e-01 0.337790082   367
## 368 -2.36280038 19.67394 3.189411e+00 7.420496e-02 0.131003822   368
## 369 -1.26575996 19.52268 9.677154e-01 3.253200e-01 0.399891864   369
## 370 -4.54504356 19.64417 7.962963e+00 4.801885e-03 0.024300237   370
## 371  0.99597116 19.83082 4.335213e-01 5.103103e-01 0.586139536   371
## 372  2.87082995 19.93144 3.419000e+00 6.453600e-02 0.119079327   372
## 373  0.58862370 19.55367 1.527675e-01 6.959288e-01 0.748254310   373
## 374 -2.67891857 20.15524 3.222108e+00 7.273798e-02 0.129748832   374
## 375  0.47360197 19.65396 9.930710e-02 7.526819e-01 0.800249151   375
## 376 -3.72749755 19.77207 4.361950e+00 3.682374e-02 0.082728217   376
## 377 -1.50053808 20.16338 9.942762e-01 3.187698e-01 0.394667328   377
## 378 -4.30182470 19.74121 7.885371e+00 5.011791e-03 0.024388207   378
## 379 -2.32478718 20.03502 2.511059e+00 1.131420e-01 0.181450231   379
## 380  4.80268686 19.86667 6.572745e+00 1.039744e-02 0.036938430   380
## 381  2.31732935 20.17341 1.979160e+00 1.595695e-01 0.234435974   381
## 382  5.77686036 19.41640 9.913698e+00 1.654616e-03 0.014507189   382
## 383 -2.68981985 19.28532 3.979601e+00 4.613341e-02 0.095790467   383
## 384 -2.33836816 19.42543 2.977234e+00 8.453347e-02 0.143623196   384
## 385 -1.24722709 19.41075 8.592386e-01 3.540170e-01 0.430235987   385
## 386 -6.31473382 18.86880 1.193830e+01 5.565841e-04 0.009232713   386
## 387 -0.17169160 19.96767 1.458594e-02 9.038785e-01 0.925911192   387
## 388  3.66974298 19.71786 5.255284e+00 2.194048e-02 0.058281510   388
## 389 -4.50905316 19.15155 7.655305e+00 5.690887e-03 0.026828466   389
## 390  7.04065633 19.15146 1.285339e+01 3.415599e-04 0.008580533   390
## 391 -2.79154459 19.79556 3.966409e+00 4.649580e-02 0.096128671   391
## 392 -4.44298666 19.49106 8.636842e+00 3.316285e-03 0.020698383   392
## 393 -3.30702469 19.06265 3.954309e+00 4.683081e-02 0.096356924   393
## 394  2.21397405 19.41585 1.411415e+00 2.349039e-01 0.307706169   394
## 395 -4.29079727 20.08111 6.735818e+00 9.489839e-03 0.034860582   395
## 396 -2.92624514 19.51250 4.254464e+00 3.922239e-02 0.085413221   396
## 397  2.95126262 19.95280 2.306205e+00 1.289504e-01 0.199350397   397
## 398 -2.29385881 19.82818 2.800874e+00 9.430406e-02 0.156202472   398
## 399  2.25542716 20.57636 1.859494e+00 1.727726e-01 0.246244036   399
## 400 -1.68583878 19.67714 1.731537e+00 1.883027e-01 0.262278721   400
## 401 -4.69616526 19.77893 8.488587e+00 3.596901e-03 0.020971928   401
## 402  3.41213293 19.76020 3.641927e+00 5.642504e-02 0.109779334   402
## 403 -3.18928150 20.21065 3.887913e+00 4.871501e-02 0.098578964   403
## 404 -5.04876521 19.71484 9.825433e+00 1.735686e-03 0.014744742   404
## 405 -1.34060783 19.53820 1.102946e+00 2.936944e-01 0.371666362   405
## 406 -4.33165201 19.67966 7.606807e+00 5.845633e-03 0.026876670   406
## 407  3.78019165 19.36953 5.979201e+00 1.452549e-02 0.046812378   407
## 408 -3.02751539 20.25253 3.532472e+00 6.026265e-02 0.114224973   408
## 409 -1.81012859 19.14752 1.969763e+00 1.605635e-01 0.234869340   409
## 410  2.02270478 19.48939 2.201074e+00 1.380064e-01 0.209204109   410
## 411  3.13376968 19.28639 4.294144e+00 3.831836e-02 0.084517098   411
## 412  3.86365600 20.07680 3.627449e+00 5.691738e-02 0.109988986   412
## 413 -3.38872851 19.56998 4.633968e+00 3.141449e-02 0.075079764   413
## 414 -5.06953393 19.66509 9.435726e+00 2.144670e-03 0.016625174   414
## 415  0.79262793 19.95207 2.761674e-01 5.992583e-01 0.666015009   415
## 416  2.12111916 19.90126 1.936458e+00 1.641438e-01 0.238300151   416
## 417  3.42222914 19.45255 4.810221e+00 2.835828e-02 0.070119331   417
## 418 -5.15815711 20.42360 8.767409e+00 3.087622e-03 0.020580064   418
## 419 -2.31992474 19.78222 2.936630e+00 8.668154e-02 0.145659658   419
## 420  1.13941855 18.99623 5.815424e-01 4.457608e-01 0.523921580   420
## 421  0.86380663 19.39502 4.421155e-01 5.061479e-01 0.582137881   421
## 422 -0.61345167 20.42304 1.764569e-01 6.744623e-01 0.731591184   422
## 423 -2.54012037 19.52979 3.470768e+00 6.254766e-02 0.117383656   423
## 424 -1.79627465 19.37906 1.946495e+00 1.630553e-01 0.237285489   424
## 425  0.12017218 19.91586 8.160612e-03 9.280254e-01 0.944538307   425
## 426  2.96331986 20.57140 3.212763e+00 7.315405e-02 0.129983395   426
## 427 -6.76170597 19.82145 1.148730e+01 7.086388e-04 0.009232713   427
## 428 -4.65533102 19.25527 8.987321e+00 2.738084e-03 0.019256359   428
## 429  4.26669455 19.57723 6.094104e+00 1.361175e-02 0.045297288   429
## 430  1.38852999 19.13198 1.110593e+00 2.920274e-01 0.370486683   430
## 431 -4.30506564 19.93211 7.332976e+00 6.803654e-03 0.029152720   431
## 432 -3.12854090 19.65230 5.199374e+00 2.265609e-02 0.059446250   432
## 433  1.20626571 20.22626 6.576964e-01 4.174306e-01 0.496060226   433
## 434  7.02118851 19.11687 1.275077e+01 3.607443e-04 0.008580533   434
## 435  2.43337374 20.35926 2.362573e+00 1.243693e-01 0.194724154   435
## 436 -4.57366405 19.24757 7.692994e+00 5.573523e-03 0.026420348   436
## 437  4.04982813 19.61795 5.603843e+00 1.797630e-02 0.052820760   437
## 438 -4.43146219 19.18581 7.083085e+00 7.817749e-03 0.031491214   438
## 439 -4.26609998 20.14212 7.170709e+00 7.445643e-03 0.030861649   439
## 440 -1.31503936 19.94108 9.075269e-01 3.408385e-01 0.416580434   440
## 441 -3.31148702 19.13124 4.612883e+00 3.180245e-02 0.075795834   441
## 442 -1.77134818 20.35764 1.479571e+00 2.239253e-01 0.296222964   442
## 443 -1.05606805 19.71431 4.554548e-01 4.998013e-01 0.577937390   443
## 444 -1.36033366 19.70560 1.006568e+00 3.157971e-01 0.392686866   444
## 445  0.85675519 20.08012 3.265641e-01 5.677267e-01 0.638522663   445
## 446 -0.48934246 19.61326 1.265748e-01 7.220320e-01 0.773412607   446
## 447  1.91500726 19.27855 2.030729e+00 1.542376e-01 0.227772496   447
## 448 -3.07661442 19.46694 4.134006e+00 4.210653e-02 0.088983760   448
## 449  2.83304829 20.15053 2.441532e+00 1.182530e-01 0.188141250   449
## 450 -2.03400531 20.17919 1.594834e+00 2.067227e-01 0.278006356   450
## 451  0.04821066 20.07273 7.586966e-04 9.780271e-01 0.979168312   451
## 452  2.75500770 19.72417 3.357907e+00 6.696963e-02 0.121995641   452
## 453  0.85647171 19.85199 2.764879e-01 5.990464e-01 0.666015009   453
## 454 -3.36180385 19.59388 4.935698e+00 2.637243e-02 0.066200818   454
## 455  1.07231486 19.78072 2.772619e-01 5.985355e-01 0.666015009   455
## 456  5.32335808 20.07513 7.009611e+00 8.144407e-03 0.032069599   456
## 457 -1.14238956 19.22185 8.171811e-01 3.660691e-01 0.443000392   457
## 458  1.55244718 19.45106 1.055887e+00 3.042275e-01 0.381061604   458
## 459 -3.62652493 19.68997 5.797878e+00 1.609796e-02 0.049313775   459
## 460 -7.69749879 19.88232 1.440233e+01 1.501682e-04 0.008580533   460
## 461 -1.08842676 19.83296 5.856346e-01 4.441648e-01 0.522761854   461
## 462  1.01767480 19.62790 4.948951e-01 4.817995e-01 0.560900923   462
## 463  1.49290331 19.60034 1.180854e+00 2.772586e-01 0.354527406   463
## 464  3.24682772 19.71881 4.264915e+00 3.898212e-02 0.085132539   464
## 465 -2.60368782 19.34723 2.846613e+00 9.165756e-02 0.153000357   465
## 466 -5.13511837 19.12347 1.017547e+01 1.436038e-03 0.013412888   466
## 467 -0.23369790 20.03322 2.816066e-02 8.667416e-01 0.897061836   467
## 468 -4.82738530 20.03564 7.134971e+00 7.595172e-03 0.030884633   468
## 469 -0.08241422 18.96723 4.323278e-03 9.475794e-01 0.958945839   469
## 470 -6.38827906 20.72758 1.026736e+01 1.366443e-03 0.013412888   470
## 471 -1.98474049 19.42772 1.715302e+00 1.903866e-01 0.263896106   471
## 472 -2.21205729 19.57964 2.829259e+00 9.265204e-02 0.154060944   472
## 473 -1.06491855 18.86811 6.876822e-01 4.070116e-01 0.486373134   473
## 474  6.88696582 19.37399 1.166588e+01 6.439621e-04 0.009232713   474
## 475 -3.15674492 19.28917 5.344261e+00 2.084952e-02 0.056610414   475
## 476  3.11457545 19.89092 2.348195e+00 1.255204e-01 0.195811778   476
## 477  2.13757547 19.51036 2.401013e+00 1.213489e-01 0.190342555   477
## 478  2.24915592 19.34115 1.657274e+00 1.980585e-01 0.271027479   478
## 479  1.90618600 19.43401 1.867715e+00 1.718265e-01 0.245303104   479
## 480 -2.67910685 19.74418 3.721949e+00 5.378399e-02 0.106930107   480
## 481 -1.72174032 19.79017 1.705126e+00 1.917065e-01 0.265297115   481
## 482 -3.37204321 19.14234 6.052387e+00 1.393645e-02 0.045865495   482
## 483 -1.93206314 20.33787 1.720682e+00 1.896930e-01 0.263787072   483
## 484 -4.97133576 19.18455 9.179028e+00 2.466283e-03 0.018061923   484
## 485 -6.17564835 19.79247 1.232668e+01 4.522772e-04 0.008819405   485
## 486  5.49545240 19.88924 8.659715e+00 3.275018e-03 0.020698383   486
## 487 -5.19347157 19.90347 8.417997e+00 3.738890e-03 0.020971928   487
## 488 -3.07572970 19.51087 4.826337e+00 2.809474e-02 0.069668462   488
## 489 -4.42734619 20.10089 4.219607e+00 4.003499e-02 0.086275821   489
## 490 -0.52604127 20.20686 1.242081e-01 7.245361e-01 0.775127176   490
## 491 -2.15395659 19.93968 2.339630e+00 1.262116e-01 0.196176802   491
## 492  2.14498176 19.96565 2.007134e+00 1.566516e-01 0.230940030   492
## 493 -5.97973229 19.74768 1.176907e+01 6.093392e-04 0.009232713   493
## 494 -2.93339973 20.20796 3.208963e+00 7.332397e-02 0.129983395   494
## 495 -4.28282031 20.01119 6.885984e+00 8.725909e-03 0.033396760   495
## 496  7.51698504 19.63628 1.362369e+01 2.268201e-04 0.008580533   496
## 497 -3.80701969 20.12229 5.781824e+00 1.624538e-02 0.049313775   497
## 498  1.44768092 19.83069 8.268519e-01 3.632483e-01 0.440830295   498
## 499  0.22085101 19.01426 2.862146e-02 8.656660e-01 0.897030688   499
## 500  4.47198309 19.75738 6.588209e+00 1.030768e-02 0.036849940   500
## 501 -2.80864656 20.25895 3.509921e+00 6.108729e-02 0.115446903   501
## 502  2.13819404 19.14535 2.561736e+00 1.095703e-01 0.177715206   502
## 503 -0.72530665 19.02351 3.324966e-01 5.642305e-01 0.636149529   503
## 504 -3.20010166 19.22480 4.486445e+00 3.423634e-02 0.079885323   504
## 505  2.35071955 19.80704 2.284345e+00 1.307774e-01 0.201087813   505
## 506 -2.52195923 19.63671 3.335074e+00 6.790417e-02 0.122656378   506
## 507 -5.92857583 19.97840 9.675310e+00 1.882938e-03 0.015508575   507
## 508 -2.68693855 20.03967 3.345653e+00 6.746948e-02 0.122386503   508
## 509 -0.86404142 19.39091 4.488068e-01 5.029472e-01 0.580011702   509
## 510  1.71327591 18.94396 1.409072e+00 2.352926e-01 0.307745528   510
## 511  2.80766027 19.75593 2.342527e+00 1.259774e-01 0.196168038   511
## 512  5.10739276 19.27327 8.643255e+00 3.304661e-03 0.020698383   512
## 513 -2.34664492 19.54873 3.115801e+00 7.762525e-02 0.135370860   513
## 514 -1.62628639 19.51551 1.553308e+00 2.127329e-01 0.283864395   514
## 515  5.29765253 19.78087 8.500397e+00 3.573687e-03 0.020971928   515
## 516  0.22496180 19.19670 2.533213e-02 8.735516e-01 0.900850117   516
## 517  5.03085040 19.26446 8.417577e+00 3.739749e-03 0.020971928   517
## 518 -0.40355089 19.83392 8.833720e-02 7.663198e-01 0.809731992   518
## 519  1.96714328 20.17044 1.644998e+00 1.997277e-01 0.272442560   519
## 520 -1.42344980 19.33039 1.152239e+00 2.831565e-01 0.360992921   520
## 521  0.24666813 20.06725 2.659095e-02 8.704750e-01 0.898757608   521
## 522  4.35588479 19.73863 6.222035e+00 1.266368e-02 0.043116824   522
## 523 -2.60857347 19.36161 3.364833e+00 6.668884e-02 0.121742607   523
## 524  2.83734110 19.88596 3.429469e+00 6.412851e-02 0.118685887   524
## 525 -3.99158321 20.18663 6.314243e+00 1.202267e-02 0.041594553   525
## 526  4.99698114 19.48642 7.434942e+00 6.429456e-03 0.028571519   526
## 527 -2.68580974 19.69667 3.787902e+00 5.170583e-02 0.103411657   527
## 528 -2.14830378 19.69746 2.652178e+00 1.034999e-01 0.168506494   528
## 529 -5.52370698 20.05535 9.542117e+00 2.024170e-03 0.016080904   529
## 530  3.69563815 19.36412 5.752542e+00 1.651788e-02 0.049693958   530
## 531 -0.44886716 20.00542 1.000840e-01 7.517481e-01 0.800248019   531
## 532 -4.13078513 19.57578 7.595373e+00 5.882737e-03 0.026876670   532
## 533 -2.66288233 19.96984 3.307058e+00 6.907002e-02 0.124500164   533
## 534 -0.07092338 19.46597 2.443822e-03 9.605755e-01 0.969616197   534
## 535 -3.41211159 19.58357 5.478308e+00 1.931086e-02 0.053794546   535
## 536 -2.37897467 19.95800 2.724530e+00 9.890870e-02 0.162574071   536
## 537  5.25830413 19.97391 7.136232e+00 7.589842e-03 0.030884633   537
## 538  1.34322303 19.72741 1.021050e+00 3.123412e-01 0.388953245   538
## 539  7.51610553 19.77942 1.265011e+01 3.806198e-04 0.008580533   539
## 540 -2.43529165 19.64159 3.403472e+00 6.514544e-02 0.119689044   540
## 541 -3.09746444 19.70579 4.390332e+00 3.621639e-02 0.082334990   541
## 542 -6.24717749 19.97404 1.083886e+01 1.003984e-03 0.011182657   542
## 543 -3.49427205 19.41300 5.915836e+00 1.505618e-02 0.048004723   543
## 544 -0.25876357 19.46669 3.833484e-02 8.447839e-01 0.878575263   544
## 545 -2.90116012 20.36774 3.687962e+00 5.488933e-02 0.108704173   545
## 546  2.52529975 20.15763 2.195795e+00 1.384796e-01 0.209551191   546
## 547 -5.08786515 20.22420 8.769134e+00 3.084710e-03 0.020580064   547
## 548 -3.24218265 19.47468 5.580559e+00 1.821644e-02 0.052982041   548
## 549 -1.88176348 19.52411 1.926697e+00 1.652103e-01 0.239443351   549
## 550  2.38063183 19.21353 2.166313e+00 1.411558e-01 0.212850011   550
## 551  3.96934745 19.43941 5.109800e+00 2.385346e-02 0.061460263   551
## 552  0.63396336 19.94133 2.079724e-01 6.483901e-01 0.709333194   552
## 553 -1.79764977 19.55381 1.400346e+00 2.367476e-01 0.308707373   553
## 554  4.20583273 20.14955 5.840398e+00 1.571412e-02 0.049206989   554
## 555  0.65104640 19.74301 2.432115e-01 6.219277e-01 0.683244545   555
## 556  2.17298694 19.78882 1.843661e+00 1.746117e-01 0.247532222   556
## 557 -7.37812247 19.78242 1.442326e+01 1.485148e-04 0.008580533   557
## 558  4.05625665 19.52126 5.617144e+00 1.784059e-02 0.052602163   558
## 559 -2.92194082 19.25760 4.368164e+00 3.668985e-02 0.082728217   559
## 560 -6.73359172 19.45728 1.152293e+01 6.952315e-04 0.009232713   560
## 561  3.10907230 20.10437 3.628250e+00 5.689002e-02 0.109988986   561
## 562 -0.05050059 19.83117 1.258200e-03 9.717061e-01 0.975373920   562
## 563  7.88626147 20.14688 1.172096e+01 6.252422e-04 0.009232713   563
## 564 -3.30091293 19.50273 3.282083e+00 7.012735e-02 0.125877123   564
## 565  1.56815959 19.21423 1.259591e+00 2.618070e-01 0.337790082   565
## 566 -2.17696410 19.77963 2.551070e+00 1.103116e-01 0.178579937   566
## 567 -2.54557978 19.83347 3.023560e+00 8.215282e-02 0.140974245   567
## 568 -3.73445586 20.08303 5.790635e+00 1.616431e-02 0.049313775   568
## 569 -3.27377921 19.80800 4.438473e+00 3.521005e-02 0.081583245   569
## 570  0.27788085 19.71816 4.258213e-02 8.365263e-01 0.872101589   570
## 571  4.52384406 19.21196 6.719137e+00 9.578827e-03 0.034972907   571
## 572  5.48482270 19.39268 9.280220e+00 2.334029e-03 0.017263766   572
## 573 -4.24190544 19.72695 6.036934e+00 1.405874e-02 0.045865495   573
## 574  2.87890648 19.31929 3.335282e+00 6.789563e-02 0.122656378   574
## 575 -6.58132966 19.37489 1.257413e+01 3.963496e-04 0.008580533   575
## 576  5.67096671 19.49240 9.657903e+00 1.900813e-03 0.015508575   576
## 577  4.55152587 19.58993 7.175919e+00 7.424100e-03 0.030861649   577
## 578 -2.56138119 20.31399 3.076722e+00 7.950968e-02 0.137816777   578
## 579 -4.63513924 19.88266 8.119074e+00 4.406292e-03 0.023093273   579
## 580 -0.69629841 19.35507 1.973299e-01 6.569128e-01 0.714386974   580
## 581  5.00119183 19.35153 7.868177e+00 5.059558e-03 0.024388207   581
## 582 -2.84084923 19.43867 4.038130e+00 4.456123e-02 0.093025628   582
## 583 -0.10488501 19.51022 5.609704e-03 9.403003e-01 0.954766427   583
## 584  6.49519040 18.96044 1.111971e+01 8.632228e-04 0.010320258   584
## 585  1.88971649 19.43210 1.947206e+00 1.629786e-01 0.237285489   585
## 586 -2.33161896 20.34581 2.532022e+00 1.116493e-01 0.179728155   586
## 587  3.30746329 19.41065 4.510020e+00 3.376819e-02 0.079161495   587
## 588 -4.87867116 19.53633 8.742400e+00 3.130144e-03 0.020580064   588
## 589  6.94962525 19.24578 1.259106e+01 3.927891e-04 0.008580533   589
## 590  5.36163286 19.54505 8.737819e+00 3.137996e-03 0.020580064   590
## 591 -4.30595204 19.26081 7.945124e+00 4.849336e-03 0.024300237   591
## 592 -3.99461648 20.48515 5.851162e+00 1.561846e-02 0.049086577   592
## 593  0.92698167 19.33836 4.691763e-01 4.934126e-01 0.572866067   593
## 594 -2.79111531 19.84653 3.592266e+00 5.813298e-02 0.111584116   594
## 595  5.33646461 20.09056 7.414598e+00 6.502399e-03 0.028571519   595
## 596 -6.01856253 19.97462 1.044621e+01 1.240637e-03 0.012981303   596
## 597 -0.06191417 19.70391 2.049555e-03 9.638931e-01 0.970681111   597
## 598 -5.03051967 18.82559 7.054965e+00 7.941153e-03 0.031838827   598
## 599  3.92049919 20.00946 5.232607e+00 2.222786e-02 0.058681550   599
## 600 -2.48749389 19.41580 3.197050e+00 7.385942e-02 0.130662638   600
## 601  1.66280337 20.21515 1.170192e+00 2.794378e-01 0.356782241   601
## 602 -1.32042816 19.73845 1.033776e+00 3.093451e-01 0.386906810   602
## 603 -2.70089484 19.51598 3.668909e+00 5.551947e-02 0.108834256   603
## 604 -7.56753958 19.96554 1.343113e+01 2.512274e-04 0.008580533   604
## 605 -6.08213828 19.80503 1.149229e+01 7.067483e-04 0.009232713   605
## 606 -5.80002283 20.28377 8.115857e+00 4.414099e-03 0.023093273   606
## 607  0.72654942 19.73530 2.004181e-01 6.544118e-01 0.713450227   607
## 608  2.68093340 19.70091 2.737455e+00 9.811219e-02 0.162197024   608
## 609  2.13079953 19.52828 1.660915e+00 1.975666e-01 0.270786211   609
## 610 -6.35459567 19.66361 9.873758e+00 1.690815e-03 0.014507189   610
## 611 -1.28448363 20.26293 7.817564e-01 3.766664e-01 0.454542610   611
## 612 -5.53798890 20.02534 9.356435e+00 2.239191e-03 0.016765965   612
## 613 -3.75049543 19.77699 6.123159e+00 1.339021e-02 0.044978015   613
## 614 -5.25211392 19.30073 1.029942e+01 1.342977e-03 0.013398534   614
## 615 -2.81743252 19.65373 4.264385e+00 3.899427e-02 0.085132539   615
## 616 -1.63752011 19.71526 1.587706e+00 2.077399e-01 0.278501294   616
## 617  3.05578183 19.18287 3.585008e+00 5.838718e-02 0.111821880   617
## 618  7.20470464 19.32828 1.288746e+01 3.354213e-04 0.008580533   618
## 619 -3.67547484 20.07603 5.525028e+00 1.880261e-02 0.053112786   619
## 620 -2.56476666 20.00298 2.954469e+00 8.573062e-02 0.144850678   620
## 621  4.82658064 19.14775 7.894947e+00 4.985388e-03 0.024388207   621
## 622  1.90367790 19.09438 1.825836e+00 1.767092e-01 0.249369271   622
## 623  4.60545068 19.64387 6.872823e+00 8.790266e-03 0.033396760   623
## 624 -1.35008540 19.79833 1.028997e+00 3.104658e-01 0.387743357   624
## 625 -2.67580049 19.40284 3.666450e+00 5.560137e-02 0.108834256   625
## 626 -4.45190687 19.54643 8.727106e+00 3.156441e-03 0.020580064   626
## 627  2.59369970 19.34217 3.452563e+00 6.323933e-02 0.118212084   627
## 628 -3.34765560 19.33334 4.940554e+00 2.629855e-02 0.066200818   628
## 629 -4.53702668 19.19126 8.180931e+00 4.258878e-03 0.022981870   629
## 630 -3.34258780 19.69219 5.759731e+00 1.645055e-02 0.049693958   630
## 631  6.91480966 19.05801 1.239718e+01 4.355772e-04 0.008819405   631
## 632 -0.10204274 19.83656 4.291963e-03 9.477693e-01 0.958945839   632
## 633 -3.52642407 19.12942 5.358525e+00 2.067994e-02 0.056374928   633
## 634 -0.57809236 19.78304 1.583351e-01 6.907187e-01 0.744518445   634
## 635 -2.06309706 19.70182 2.410921e+00 1.205836e-01 0.189836265   635
## 636  7.51683597 19.68413 1.344207e+01 2.497716e-04 0.008580533   636
## 637 -0.05488352 19.93571 1.235404e-03 9.719635e-01 0.975373920   637
## 638 -2.44563698 19.87718 2.997578e+00 8.347897e-02 0.142679199   638
## 639 -2.96149745 19.43731 4.480460e+00 3.435628e-02 0.079885323   639
## 640 -2.14266394 19.94383 2.125146e+00 1.449904e-01 0.217866527   640
## 641  3.86723186 19.46472 5.009519e+00 2.527219e-02 0.064342851   641
## 642  3.89502618 19.75869 6.036901e+00 1.405900e-02 0.045865495   642
## 643 -4.68242253 19.67627 6.119221e+00 1.342001e-02 0.044978015   643
## 644  6.09814975 19.81835 9.947054e+00 1.624988e-03 0.014507189   644
## 645 -0.91564075 19.94786 4.441060e-01 5.051921e-01 0.581818520   645
## 646 -2.40506680 20.20967 2.694485e+00 1.007876e-01 0.164715800   646
## 647 -4.56932593 20.17454 7.445466e+00 6.392051e-03 0.028564478   647
## 648 -1.83939761 19.18542 2.087174e+00 1.486319e-01 0.221015921   648
## 649 -1.71189400 19.98406 1.465209e+00 2.261865e-01 0.298566205   649
## 650  1.63474441 19.59388 1.478672e+00 2.240661e-01 0.296222964   650
## 651  0.35827855 19.89324 6.334387e-02 8.013019e-01 0.841514085   651
## 652 -2.66804082 19.35177 3.851257e+00 4.978941e-02 0.100280084   652
## 653 -2.04982020 19.94566 1.488461e+00 2.225392e-01 0.295113857   653
## 654  4.45013719 20.35031 4.884387e+00 2.716641e-02 0.067758086   654
## 655  3.42131649 19.44182 4.430641e+00 3.537176e-02 0.081583245   655
## 656 -4.77427016 19.82773 7.205687e+00 7.302211e-03 0.030712240   656
## 657 -2.26454784 19.59932 3.029781e+00 8.183865e-02 0.140974245   657
## 658 -5.02871878 20.25099 7.936899e+00 4.871376e-03 0.024300237   658
## 659  0.92971668 19.10269 3.956478e-01 5.293874e-01 0.605619179   659
## 660 -3.22654998 19.82597 3.958407e+00 4.671706e-02 0.096353937   660
## 661  6.36609659 19.45423 9.349884e+00 2.247186e-03 0.016765965   661
## 662  6.56196826 19.06185 1.145514e+01 7.209694e-04 0.009232713   662
## 663 -2.48034270 19.33237 2.659893e+00 1.029994e-01 0.168010452   663
## 664 -1.01916073 19.40207 4.093996e-01 5.223171e-01 0.598328547   664
## 665  7.22982142 19.37365 1.267081e+01 3.764432e-04 0.008580533   665
## 666 -1.92383190 20.16850 1.777817e+00 1.825061e-01 0.255237430   666
## 667 -5.23562003 19.80279 9.888094e+00 1.677730e-03 0.014507189   667
## 668 -2.81889385 20.48920 3.410249e+00 6.487871e-02 0.119454786   668
## 669 -1.12260958 19.42971 5.191016e-01 4.712743e-01 0.549393080   669
## 670  2.46248950 20.12940 2.500216e+00 1.139229e-01 0.182361604   670
## 671 -2.40337674 19.55552 2.957839e+00 8.555224e-02 0.144850678   671
## 672  0.90540042 19.10213 3.700204e-01 5.430334e-01 0.617116095   672
## 673  6.13589140 19.79727 1.021045e+01 1.409128e-03 0.013412888   673
## 674 -2.31903356 19.54919 2.329525e+00 1.270328e-01 0.196740268   674
## 675 -1.66733241 20.12202 1.329608e+00 2.489552e-01 0.323152156   675
## 676  1.57944261 20.14849 1.134753e+00 2.868387e-01 0.364603846   676
## 677 -5.93033879 20.41671 1.032072e+01 1.327614e-03 0.013398534   677
## 678 -4.59724827 19.33523 7.942360e+00 4.856733e-03 0.024300237   678
## 679 -5.99174443 20.10229 1.064381e+01 1.115213e-03 0.011812999   679
## 680 -5.44844508 19.21645 6.916421e+00 8.578908e-03 0.033156322   680
## 681 -2.95295388 19.55807 2.291838e+00 1.301479e-01 0.200479197   681
## 682  4.84579943 19.41574 7.819518e+00 5.197269e-03 0.024912047   682
## 683 -4.31822073 19.54999 8.059421e+00 4.553379e-03 0.023254759   683
## 684  2.03486868 19.33405 2.353494e+00 1.250947e-01 0.195503269   684
## 685 -3.41287460 19.42639 3.067456e+00 7.996371e-02 0.138324317   685
## 686  6.59917929 19.19883 1.154257e+01 6.879478e-04 0.009232713   686
## 687  0.77205868 19.65477 2.185948e-01 6.401424e-01 0.702355690   687
## 688  0.89516428 19.83688 3.392388e-01 5.603072e-01 0.633390749   688
## 689 -4.04098288 19.47565 7.147174e+00 7.543775e-03 0.030884633   689
## 690 -2.84107663 19.61391 4.392826e+00 3.616351e-02 0.082334990   690
## 691 -4.86118480 19.39067 8.283890e+00 4.024553e-03 0.022135040   691
## 692  0.83639394 19.74592 3.714739e-01 5.422423e-01 0.617034347   692
## 693 -3.65725318 19.26353 6.382661e+00 1.156866e-02 0.040349228   693
## 694 -5.44610936 20.26607 8.721484e+00 3.166164e-03 0.020580064   694
## 695 -2.96780020 20.01490 3.879800e+00 4.895068e-02 0.098822784   695
## 696 -4.69562422 20.03365 7.172733e+00 7.437265e-03 0.030861649   696
## 697 -1.97228472 19.56481 2.266252e+00 1.323113e-01 0.202371273   697
## 698  2.25457594 19.90352 2.133120e+00 1.442387e-01 0.217117140   698
## 699 -1.68626205 20.23843 1.363625e+00 2.429913e-01 0.315888676   699
## 700 -5.19171296 19.35251 9.477333e+00 2.096701e-03 0.016504307   700
## 701 -4.11379570 19.78250 7.334161e+00 6.799181e-03 0.029152720   701
## 702  4.27541932 19.10289 6.870703e+00 8.800681e-03 0.033396760   702
## 703 -2.16441146 20.18533 2.204998e+00 1.376559e-01 0.209042054   703
## 704 -0.43176668 20.31076 8.868563e-02 7.658728e-01 0.809731992   704
## 705  3.78749292 19.22836 5.402485e+00 2.016619e-02 0.055279830   705
## 706  6.92460487 19.20394 1.259629e+01 3.916956e-04 0.008580533   706
## 707  4.11195260 19.39730 5.796214e+00 1.611318e-02 0.049313775   707
## 708  1.72996981 19.46712 1.461639e+00 2.267528e-01 0.298853857   708
## 709 -3.53570915 19.19290 6.802915e+00 9.140373e-03 0.034246464   709
## 710  5.39957845 19.41232 8.936717e+00 2.814780e-03 0.019634807   710
## 711 -0.78180394 19.92119 2.876096e-01 5.917912e-01 0.662456371   711
## 712 -4.31749678 19.60464 8.088819e+00 4.480278e-03 0.023157102   712
## 713 -0.82352017 19.28346 2.843363e-01 5.939076e-01 0.663506125   713
## 714 -3.29278729 20.18048 3.637207e+00 5.658507e-02 0.109841598   714
## 715  1.53476004 19.34166 1.214436e+00 2.705327e-01 0.348001593   715
## 716 -4.55685112 20.04053 6.754051e+00 9.393540e-03 0.034739902   716
## 717 -5.36886661 20.32428 9.365122e+00 2.228632e-03 0.016765965   717
## 718 -3.47238175 19.03902 5.783750e+00 1.622763e-02 0.049313775   718
## 719 -6.04804373 20.03510 1.018499e+01 1.428663e-03 0.013412888   719
## 720 -3.01503527 19.28113 5.188291e+00 2.280080e-02 0.059643560   720
## 721 -0.62774549 19.16630 2.481545e-01 6.184095e-01 0.682566036   721
## 722 -3.78274857 19.91472 5.903438e+00 1.516233e-02 0.048004723   722
## 723 -4.16872287 20.00997 6.569131e+00 1.041853e-02 0.036938430   723
## 724 -4.47159883 19.81460 7.602333e+00 5.860123e-03 0.026876670   724
## 725 -1.19466773 19.15661 6.161734e-01 4.325270e-01 0.511873365   725
## 726 -0.48972309 20.02931 1.202843e-01 7.287476e-01 0.778661786   726
## 727 -2.10958970 19.82650 2.265368e+00 1.323868e-01 0.202371273   727
## 728 -2.98670438 19.82737 4.228232e+00 3.983229e-02 0.086085903   728
## 729 -2.88044659 20.21107 3.720243e+00 5.383893e-02 0.106930107   729
## 730 -5.57633838 20.04668 8.996658e+00 2.724167e-03 0.019256359   730
## 731 -7.87460069 20.32750 1.358583e+01 2.314224e-04 0.008580533   731
## 732  1.99184122 19.28419 2.263393e+00 1.325555e-01 0.202371273   732
## 733 -4.85915115 19.76790 8.511734e+00 3.551546e-03 0.020971928   733
## 734 -2.91196861 19.32923 3.299046e+00 6.940733e-02 0.124845895   734
## 735 -1.34175400 19.56245 1.073956e+00 3.001266e-01 0.378133046   735
## 736 -4.75346165 19.74869 8.616606e+00 3.353234e-03 0.020698383   736
## 737  6.59481443 19.19314 1.153809e+01 6.896028e-04 0.009232713   737
## 738 -2.33075395 19.09646 2.699264e+00 1.004862e-01 0.164536578   738
## 739 -7.48401952 19.97523 1.342179e+01 2.524759e-04 0.008580533   739
## 740  5.15584353 19.14128 9.050593e+00 2.645168e-03 0.018960308   740
## 741 -3.21666590 19.49239 5.112567e+00 2.381550e-02 0.061460263   741
## 742 -0.30687805 20.00783 4.093115e-02 8.396829e-01 0.874329972   742
## 743  4.49583944 19.71382 6.627441e+00 1.008350e-02 0.036351432   743
## 744 -1.73061913 19.96146 1.438093e+00 2.305310e-01 0.302902882   744
## 745  7.62353425 19.65364 1.388041e+01 1.979542e-04 0.008580533   745
## 746  7.08299008 19.40791 1.210653e+01 5.087095e-04 0.008907606   746
## 747 -0.52036419 19.56004 1.391586e-01 7.091419e-01 0.760554725   747
## 748 -2.38654410 20.05418 2.402847e+00 1.212068e-01 0.190342555   748
## 749 -4.25268267 19.67964 8.251242e+00 4.097400e-03 0.022250438   749
## 750 -2.75907003 19.96739 3.531235e+00 6.030759e-02 0.114224973   750
## 751 -2.49168170 19.81094 3.111279e+00 7.784084e-02 0.135471484   751
## 752  2.87034556 19.74814 3.902892e+00 4.828304e-02 0.098401071   752
## 753  1.61457442 19.90388 1.203250e+00 2.727501e-01 0.349282993   753
## 754 -5.36187637 19.59350 1.075161e+01 1.052281e-03 0.011428573   754
## 755  2.13833409 19.33385 1.992442e+00 1.581764e-01 0.232787916   755
## 756 -7.15155788 19.70624 1.335335e+01 2.618213e-04 0.008580533   756
## 757  0.22804948 20.00845 2.300058e-02 8.794645e-01 0.904772810   757
## 758  6.54094556 19.11284 1.140374e+01 7.411206e-04 0.009351198   758
## 759 -2.49250610 19.44334 3.121679e+00 7.734602e-02 0.135158620   759
## 760  1.28784563 19.89283 7.113482e-01 3.990554e-01 0.478866463   760
## 761  0.64677377 19.08039 2.688163e-01 6.041604e-01 0.668863978   761
## 762 -1.85012389 19.70999 1.971381e+00 1.603919e-01 0.234869340   762
## 763  5.12684611 19.21708 8.884982e+00 2.895447e-03 0.019716613   763
## 764 -1.41468380 19.51334 9.911448e-01 3.195329e-01 0.395042144   764
## 765  2.18483548 19.38151 2.422730e+00 1.196786e-01 0.189122630   765
## 766 -3.40280302 19.62469 5.674949e+00 1.726300e-02 0.051251401   766
## 767  7.44399929 19.80029 1.219530e+01 4.851448e-04 0.008907606   767
## 768 -1.13798949 19.94870 6.352681e-01 4.254847e-01 0.504931992   768
## 769  1.93540426 19.68479 1.579237e+00 2.089561e-01 0.279694742   769
## 770 -2.70827395 19.20734 3.448500e+00 6.339480e-02 0.118245082   770
## 771 -1.94141669 20.12034 1.822708e+00 1.770803e-01 0.249482520   771
## 772 -1.70074590 19.59667 1.646917e+00 1.994657e-01 0.272442560   772
## 773 -3.19821724 19.56592 4.764403e+00 2.912168e-02 0.071389729   773
## 774  2.98689933 19.47985 3.242182e+00 7.185276e-02 0.128704949   774
## 775  6.79558929 19.34258 1.159764e+01 6.679401e-04 0.009232713   775
## 776 -1.74683095 19.40113 1.776628e+00 1.826524e-01 0.255237430   776
## 777  3.37901709 19.63746 4.743777e+00 2.947232e-02 0.071838769   777
## 778 -3.27849522 19.60804 5.272342e+00 2.172684e-02 0.058281510   778
## 779 -3.28667920 19.29004 5.643738e+00 1.757242e-02 0.051990140   779
## 780 -7.60961021 19.86327 1.323927e+01 2.781813e-04 0.008580533   780
## 781 -2.30460882 19.40082 2.279611e+00 1.311768e-01 0.201341195   781
## 782 -6.19396369 19.73451 1.257167e+01 3.968701e-04 0.008580533   782
## 783 -1.35024448 19.87981 6.773198e-01 4.105683e-01 0.489941083   783
## 784 -7.39019607 19.71523 1.541852e+01 8.783236e-05 0.008580533   784
## 785  3.23163959 19.75587 4.673893e+00 3.069334e-02 0.074095057   785
## 786  1.39252017 20.35990 7.757119e-01 3.785175e-01 0.456134828   786
## 787  6.33725401 19.01508 1.066170e+01 1.104513e-03 0.011812999   787
## 788 -3.21280784 19.82246 4.709517e+00 3.006444e-02 0.073074481   788
## 789 -6.59250983 19.52282 1.212974e+01 5.024398e-04 0.008907606   789
## 790 -4.41973281 19.95661 7.317212e+00 6.863461e-03 0.029152720   790
## 791 -0.83080554 20.15322 3.330585e-01 5.639015e-01 0.636149529   791
## 792 -0.46082774 19.75480 1.167664e-01 7.325893e-01 0.780980992   792
## 793 -2.96825593 19.59523 4.780948e+00 2.884358e-02 0.070910570   793
## 794 -2.82032553 18.91781 3.976418e+00 4.622058e-02 0.095790467   794
## 795 -4.12512545 19.57341 7.549829e+00 6.032937e-03 0.027243475   795
## 796  0.47577633 19.28697 9.600499e-02 7.566963e-01 0.802528379   796
## 797 -0.89299486 19.91729 2.939600e-01 5.877292e-01 0.659178597   797
## 798  2.68047137 19.57405 2.416440e+00 1.201597e-01 0.189516557   798
## 799 -7.85104868 19.99472 1.415336e+01 1.713105e-04 0.008580533   799
## 800 -4.30713488 20.37013 6.773994e+00 9.289358e-03 0.034503328   800
## 801  0.67209529 19.58181 2.141779e-01 6.435417e-01 0.705183573   801
## 802 -2.39609182 19.75186 3.003293e+00 8.318529e-02 0.142461044   802
## 803 -5.21695267 19.24902 1.034179e+01 1.312591e-03 0.013398534   803
## 804 -2.81126881 19.41080 4.089026e+00 4.323994e-02 0.090708734   804
## 805 -3.37658208 19.95624 4.571854e+00 3.257180e-02 0.077200560   805
## 806  0.11142691 19.29980 8.315585e-03 9.273471e-01 0.944538307   806
## 807 -2.72040027 19.48341 3.184145e+00 7.444419e-02 0.131156294   807
## 808 -4.86157370 20.38161 7.382994e+00 6.617394e-03 0.028820936   808
## 809  0.97043856 19.29765 5.778160e-01 4.472219e-01 0.524202695   809
## 810  2.28413186 20.48006 1.918693e+00 1.660908e-01 0.239908918   810
## 811 -6.36643272 19.12450 1.191172e+01 5.645526e-04 0.009232713   811
## 812 -0.27731097 19.59284 4.477211e-02 8.324362e-01 0.869951589   812
## 813 -3.43884395 19.61030 5.531819e+00 1.872989e-02 0.053112786   813
## 814 -0.79608985 19.67846 3.549405e-01 5.513690e-01 0.624933385   814
## 815  3.40929828 19.05865 4.215950e+00 4.012127e-02 0.086275821   815
## 816 -1.82571640 20.07816 1.547838e+00 2.135399e-01 0.284057709   816
## 817 -3.90864918 19.53658 7.578543e+00 5.937793e-03 0.026955693   817
## 818 -7.36846694 19.36749 1.527022e+01 9.497151e-05 0.008580533   818
## 819  4.04611413 20.15924 5.539969e+00 1.864300e-02 0.053112786   819
## 820 -7.29960728 19.30718 1.530573e+01 9.321064e-05 0.008580533   820
## 821 -4.57630881 19.68007 9.165863e+00 2.484041e-03 0.018061923   821
## 822 -2.51834466 19.13557 1.845036e+00 1.744511e-01 0.247532222   822
## 823 -4.40126309 20.17228 7.225230e+00 7.223305e-03 0.030530029   823
## 824 -1.35257592 19.85465 9.822374e-01 3.217169e-01 0.397169967   824
## 825 -0.30221187 19.97764 4.787908e-02 8.268088e-01 0.865124353   825
## 826  1.10473753 19.84480 6.267330e-01 4.286109e-01 0.507939446   826
## 827  1.93027878 20.14403 1.614034e+00 2.040118e-01 0.276091637   827
## 828  4.85524236 19.53066 8.076471e+00 4.510837e-03 0.023175436   828
## 829 -3.09563018 19.53149 3.897625e+00 4.843447e-02 0.098475778   829
## 830 -4.39956410 19.69339 7.632942e+00 5.761717e-03 0.026876670   830
## 831 -3.61955189 19.83718 5.779645e+00 1.626550e-02 0.049313775   831
## 832 -3.53552457 20.07514 5.168837e+00 2.305712e-02 0.060130737   832
## 833  6.56751179 19.06974 1.147663e+01 7.127072e-04 0.009232713   833
## 834 -0.51831691 19.48513 1.537048e-01 6.950441e-01 0.748240745   834
## 835 -5.87990355 20.39951 9.650078e+00 1.908904e-03 0.015508575   835
## 836  1.27666437 19.47186 9.400498e-01 3.323334e-01 0.407345748   836
## 837 -2.91947978 18.98809 4.283558e+00 3.855737e-02 0.084618927   837
## 838 -4.73968500 19.61017 8.600613e+00 3.382733e-03 0.020731321   838
## 839 -4.78071783 20.12262 7.868596e+00 5.058388e-03 0.024388207   839
## 840  1.94198575 20.16743 1.413137e+00 2.346185e-01 0.307706169   840
## 841 -1.01009671 19.24129 5.895033e-01 4.426641e-01 0.521711246   841
## 842 -5.51973458 19.06452 1.111288e+01 8.663942e-04 0.010320258   842
## 843  1.56070692 20.24663 9.983247e-01 3.177867e-01 0.394018734   843
## 844  4.20325358 19.63480 6.180128e+00 1.296653e-02 0.043800310   844
## 845 -2.87425028 20.12538 3.497678e+00 6.153998e-02 0.116046822   845
## 846 -1.86114696 19.68211 2.092337e+00 1.481308e-01 0.220808574   846
## 847 -5.27525873 20.12667 8.627942e+00 3.332484e-03 0.020698383   847
## 848  3.16199774 19.49216 4.191750e+00 4.069711e-02 0.087077617   848
## 849 -3.67866556 19.54043 6.674647e+00 9.820372e-03 0.035552233   849
## 850 -0.82309152 20.19243 2.707319e-01 6.028748e-01 0.668863978   850
## 851 -3.29788104 20.53504 4.298270e+00 3.822561e-02 0.084517098   851
## 852  7.25127002 19.41179 1.253489e+01 4.047298e-04 0.008580533   852
## 853 -3.35701621 19.65281 5.528444e+00 1.876600e-02 0.053112786   853
## 854 -2.96477502 19.77452 4.351178e+00 3.705704e-02 0.083015506   854
## 855 -2.90296981 19.74993 4.307708e+00 3.801435e-02 0.084517098   855
## 856  1.92632996 19.58990 2.080321e+00 1.493001e-01 0.221625367   856
## 857 -3.84257728 19.32121 6.454876e+00 1.110852e-02 0.038902489   857
## 858  4.90027281 20.07636 6.092919e+00 1.362086e-02 0.045297288   858
##      SpatialFDR                       celltype celltype_fraction
## 1   0.046017136                     Mesenchyme         0.9873418
## 2   0.152022140                       Epiblast         0.9921875
## 3   0.186739541                       Epiblast         0.9906542
## 4   0.282380484               Nascent mesoderm         0.9578947
## 5   0.733497068      Anterior Primitive Streak         0.6721311
## 6   0.153305386               Nascent mesoderm         0.6792453
## 7   0.514354330                   ExE ectoderm         1.0000000
## 8   0.238257909               Nascent mesoderm         1.0000000
## 9   0.605774299                   ExE ectoderm         1.0000000
## 10  0.133970213               Primitive Streak         0.9777778
## 11  0.008109758           Rostral neurectoderm         0.8307692
## 12  0.022592635           Rostral neurectoderm         0.7551020
## 13  0.143903143                       Epiblast         1.0000000
## 14  0.219445538                   ExE ectoderm         1.0000000
## 15  0.238257909               Nascent mesoderm         0.9911504
## 16  0.086409634               Nascent mesoderm         0.8800000
## 17  0.060701283               Primitive Streak         0.9587629
## 18  0.115567277                       Epiblast         0.9939394
## 19  0.971502649                   ExE endoderm         1.0000000
## 20  0.524188043                   ExE endoderm         1.0000000
## 21  0.300133752 Haematoendothelial progenitors         0.8947368
## 22  0.219445538               Primitive Streak         0.6904762
## 23  0.143890220                   ExE endoderm         1.0000000
## 24  0.117424127            Blood progenitors 1         0.8947368
## 25  0.026143404           Rostral neurectoderm         0.8352941
## 26  0.008109758                       Epiblast         1.0000000
## 27  0.575366051               Primitive Streak         0.6285714
## 28  0.042271585               Primitive Streak         0.9838710
## 29  0.073124971                   ExE endoderm         1.0000000
## 30  0.814783482                   ExE ectoderm         1.0000000
## 31  0.010480226                       Epiblast         0.9871795
## 32  0.139942231               Primitive Streak         1.0000000
## 33  0.108891817                       Epiblast         0.9908257
## 34  0.081149377                   ExE ectoderm         1.0000000
## 35  0.070952068               Primitive Streak         0.8135593
## 36  0.307224692 Haematoendothelial progenitors         0.9880952
## 37  0.118381784               Nascent mesoderm         0.5454545
## 38  0.489432156                   ExE endoderm         1.0000000
## 39  0.888906705                   ExE endoderm         1.0000000
## 40  0.038020035 Haematoendothelial progenitors         0.8372093
## 41  0.245519585                   ExE ectoderm         1.0000000
## 42  0.016081159                       Epiblast         1.0000000
## 43  0.019152008               Nascent mesoderm         1.0000000
## 44  0.008109758                       Epiblast         1.0000000
## 45  0.484009881                   ExE endoderm         0.7538462
## 46  0.253507721               Nascent mesoderm         0.6153846
## 47  0.055543022                   ExE endoderm         1.0000000
## 48  0.376545331               Nascent mesoderm         0.9890110
## 49  0.023702221                       Epiblast         1.0000000
## 50  0.083265413           Rostral neurectoderm         0.5540541
## 51  0.371408023               Nascent mesoderm         0.9361702
## 52  0.945884493                   ExE endoderm         1.0000000
## 53  0.052244047                   ExE ectoderm         1.0000000
## 54  0.853688998                   ExE endoderm         1.0000000
## 55  0.898401508                   ExE ectoderm         1.0000000
## 56  0.370111461                  Def. endoderm         0.8867925
## 57  0.195063089                   ExE ectoderm         1.0000000
## 58  0.522194441               Primitive Streak         1.0000000
## 59  0.009985810                       Epiblast         1.0000000
## 60  0.027847708      Anterior Primitive Streak         0.9152542
## 61  0.107338242                   ExE ectoderm         1.0000000
## 62  0.081149377                       Epiblast         1.0000000
## 63  0.834263293                   ExE ectoderm         1.0000000
## 64  0.052599134                       Epiblast         0.9850746
## 65  0.008109758                 Mixed mesoderm         0.4117647
## 66  0.186794231                 Mixed mesoderm         0.5849057
## 67  0.902002637              Visceral endoderm         1.0000000
## 68  0.928906837                   ExE endoderm         1.0000000
## 69  0.018370520               Primitive Streak         0.9875000
## 70  0.073124971                   ExE ectoderm         1.0000000
## 71  0.090346795                       Epiblast         0.9928058
## 72  0.133319591                  Def. endoderm         0.7111111
## 73  0.048503872                     Mesenchyme         0.9285714
## 74  0.665676693                   ExE endoderm         1.0000000
## 75  0.057412556                       Epiblast         0.9837398
## 76  0.729798262                   ExE endoderm         1.0000000
## 77  0.032647682                   ExE mesoderm         0.5833333
## 78  0.803261776               Nascent mesoderm         0.8000000
## 79  0.266436149               Primitive Streak         1.0000000
## 80  0.052142116                       Epiblast         0.8907104
## 81  0.048944514                   ExE endoderm         1.0000000
## 82  0.248743967      Anterior Primitive Streak         0.6941176
## 83  0.346458124                   ExE ectoderm         1.0000000
## 84  0.040709021                 Mixed mesoderm         0.4545455
## 85  0.008861509                       Epiblast         0.9855072
## 86  0.161414645                       Epiblast         0.7634409
## 87  0.008861509                 Mixed mesoderm         0.6949153
## 88  0.020453593                     Mesenchyme         0.9508197
## 89  0.128353618                   ExE endoderm         1.0000000
## 90  0.033951871                       Epiblast         1.0000000
## 91  0.052244047                       Epiblast         1.0000000
## 92  0.979026214                   ExE endoderm         1.0000000
## 93  0.022582426                     Mesenchyme         1.0000000
## 94  0.053871099                       Epiblast         1.0000000
## 95  0.031052104                       Epiblast         0.8750000
## 96  0.057412556                   ExE endoderm         1.0000000
## 97  0.128692947                       Epiblast         0.9361702
## 98  0.137461748               Primitive Streak         0.6702128
## 99  0.939786278               Nascent mesoderm         0.9710145
## 100 0.514656237                   ExE ectoderm         1.0000000
## 101 0.743999527               Nascent mesoderm         0.9220779
## 102 0.052142116                   ExE ectoderm         1.0000000
## 103 0.020453593               Primitive Streak         0.9532164
## 104 0.080393222                       Epiblast         1.0000000
## 105 0.239015527                       Epiblast         0.8876404
## 106 0.112189505                   ExE ectoderm         1.0000000
## 107 0.081149377 Haematoendothelial progenitors         0.6233766
## 108 0.118381784                   ExE ectoderm         1.0000000
## 109 0.972721560               Primitive Streak         0.6521739
## 110 0.046009368                   ExE mesoderm         0.6617647
## 111 0.020453593               Primitive Streak         0.7500000
## 112 0.058059779                       Epiblast         1.0000000
## 113 0.084816677               Primitive Streak         0.9885057
## 114 0.217618893                       Epiblast         0.9565217
## 115 0.069992136                   ExE endoderm         1.0000000
## 116 0.260440683                   ExE ectoderm         1.0000000
## 117 0.391390097                       Epiblast         0.7676768
## 118 0.012925756               Nascent mesoderm         1.0000000
## 119 0.015996213                       Epiblast         0.5000000
## 120 0.467403834                   ExE ectoderm         1.0000000
## 121 0.031729734                 Mixed mesoderm         0.5384615
## 122 0.030013712                       Epiblast         1.0000000
## 123 0.161795151               Primitive Streak         0.6785714
## 124 0.924741725                   ExE ectoderm         1.0000000
## 125 0.199176852               Nascent mesoderm         0.8943089
## 126 0.028316464               Primitive Streak         0.6603774
## 127 0.097252869                   ExE endoderm         1.0000000
## 128 0.064545618                       Epiblast         0.7272727
## 129 0.076013350                       Epiblast         0.9767442
## 130 0.167775296            Blood progenitors 2         1.0000000
## 131 0.013987599           Rostral neurectoderm         0.7123288
## 132 0.062581148 Haematoendothelial progenitors         0.8611111
## 133 0.675259928                       Epiblast         0.6153846
## 134 0.524188043                   ExE endoderm         1.0000000
## 135 0.032165090                            Gut         0.8000000
## 136 0.211013101 Haematoendothelial progenitors         0.8375000
## 137 0.272552200                   ExE ectoderm         1.0000000
## 138 0.008109758           Rostral neurectoderm         0.9104478
## 139 0.008109758                       Epiblast         1.0000000
## 140 0.068636000                   ExE endoderm         1.0000000
## 141 0.280277030               Nascent mesoderm         0.8409091
## 142 0.575777052                   ExE ectoderm         1.0000000
## 143 0.046839313                       Epiblast         0.9746835
## 144 0.378010744                   ExE endoderm         1.0000000
## 145 0.030516916                       Epiblast         0.5000000
## 146 0.332255041                   ExE endoderm         1.0000000
## 147 0.266484025                       Epiblast         1.0000000
## 148 0.117424127               Nascent mesoderm         0.9473684
## 149 0.585390951                   ExE endoderm         1.0000000
## 150 0.066981608               Primitive Streak         0.5675676
## 151 0.014880099                Caudal epiblast         0.8297872
## 152 0.083265413               Primitive Streak         0.5490196
## 153 0.637621568                   ExE ectoderm         1.0000000
## 154 0.013328022                       Epiblast         0.9024390
## 155 0.605774299 Haematoendothelial progenitors         0.9833333
## 156 0.217211940                   ExE ectoderm         1.0000000
## 157 0.274868648                       Epiblast         1.0000000
## 158 0.076964221               Primitive Streak         0.9797980
## 159 0.010802425           Rostral neurectoderm         0.7173913
## 160 0.161414645              Visceral endoderm         0.9761905
## 161 0.970118911                   ExE endoderm         1.0000000
## 162 0.008109758           Rostral neurectoderm         0.8648649
## 163 0.187630597                       Epiblast         1.0000000
## 164 0.026143404                       Epiblast         0.9200000
## 165 0.022582426                       Epiblast         1.0000000
## 166 0.265442850                   ExE ectoderm         1.0000000
## 167 0.273589271                   ExE ectoderm         1.0000000
## 168 0.045482071                Caudal epiblast         0.5593220
## 169 0.025293849                     Mesenchyme         1.0000000
## 170 0.220873255                   ExE endoderm         1.0000000
## 171 0.008109758                 Mixed mesoderm         0.8723404
## 172 0.396484980               Nascent mesoderm         0.9506173
## 173 0.014963267                   ExE endoderm         1.0000000
## 174 0.095469548                       Epiblast         0.7243590
## 175 0.020453593                       Epiblast         0.9113924
## 176 0.618000614                   ExE endoderm         1.0000000
## 177 0.076366398               Nascent mesoderm         0.9866667
## 178 0.073124971               Primitive Streak         0.8947368
## 179 0.107338242                   ExE ectoderm         1.0000000
## 180 0.020131039          Intermediate mesoderm         0.2857143
## 181 0.008362158                       Epiblast         1.0000000
## 182 0.668348397                   ExE endoderm         1.0000000
## 183 0.107645394              Visceral endoderm         0.9864865
## 184 0.080393222               Primitive Streak         1.0000000
## 185 0.275610949                       Epiblast         0.9125000
## 186 0.021703820                       Epiblast         1.0000000
## 187 0.010276393                       Epiblast         0.9117647
## 188 0.032249479                       Epiblast         1.0000000
## 189 0.056486216                       Epiblast         1.0000000
## 190 0.008109758                 Mixed mesoderm         0.5147059
## 191 0.032625503               Nascent mesoderm         0.9777778
## 192 0.027847708                     Mesenchyme         0.9852941
## 193 0.085141020      Anterior Primitive Streak         0.8771930
## 194 0.020453593               Primitive Streak         0.9125000
## 195 0.136765712            Blood progenitors 1         0.7674419
## 196 0.053871099                   ExE ectoderm         1.0000000
## 197 0.020453593                     Mesenchyme         0.5348837
## 198 0.008861509               Nascent mesoderm         0.4489796
## 199 0.336058307                   ExE ectoderm         1.0000000
## 200 0.013531434                       Epiblast         1.0000000
## 201 0.033649483                       Epiblast         1.0000000
## 202 0.042081646                   ExE ectoderm         1.0000000
## 203 0.056785977               Primitive Streak         0.9402985
## 204 0.103692399            Blood progenitors 1         0.6511628
## 205 0.821086936                   ExE ectoderm         1.0000000
## 206 0.044998595                       Epiblast         0.9816514
## 207 0.204144978                   ExE ectoderm         1.0000000
## 208 0.273589271                   ExE ectoderm         1.0000000
## 209 0.065428362               Primitive Streak         0.9933775
## 210 0.012860653                       Epiblast         1.0000000
## 211 0.061899511                  Def. endoderm         0.7866667
## 212 0.047135669               Primitive Streak         0.9062500
## 213 0.709828734                   ExE endoderm         0.6619718
## 214 0.233092965            Blood progenitors 2         0.5000000
## 215 0.021196259                       Epiblast         1.0000000
## 216 0.076692375                       Epiblast         1.0000000
## 217 0.008861509                       Epiblast         1.0000000
## 218 0.843077643                       Epiblast         0.9673913
## 219 0.646549094               Nascent mesoderm         0.9054054
## 220 0.245628579               Primitive Streak         0.9714286
## 221 0.959259089                   ExE endoderm         1.0000000
## 222 0.086082663                       Epiblast         0.7846154
## 223 0.273670931                   ExE ectoderm         1.0000000
## 224 0.035939034                       Epiblast         1.0000000
## 225 0.028316464      Anterior Primitive Streak         0.6363636
## 226 0.180139330               Nascent mesoderm         0.6216216
## 227 0.422347870                   ExE endoderm         1.0000000
## 228 0.008109758          Intermediate mesoderm         0.7380952
## 229 0.008109758                Caudal epiblast         0.6666667
## 230 0.033441509               Nascent mesoderm         0.9875000
## 231 0.027128510                       Epiblast         0.5600000
## 232 0.030813538                       Epiblast         1.0000000
## 233 0.041229699                       Epiblast         0.9674797
## 234 0.108910306 Haematoendothelial progenitors         0.9111111
## 235 0.036219442                       Epiblast         0.5454545
## 236 0.291223521                       Epiblast         0.7839506
## 237 0.057412556               Primitive Streak         0.7567568
## 238 0.053142967                       Epiblast         0.9818182
## 239 0.020453593               Primitive Streak         0.9000000
## 240 0.084824257                       Epiblast         1.0000000
## 241 0.752454028                   ExE ectoderm         1.0000000
## 242 0.626049548                   ExE ectoderm         1.0000000
## 243 0.275440163               Nascent mesoderm         0.7413793
## 244 0.120964039               Nascent mesoderm         0.9444444
## 245 0.088319364                     Mesenchyme         0.9615385
## 246 0.065343205               Primitive Streak         0.7238095
## 247 0.414470812                 Mixed mesoderm         0.8906250
## 248 0.008109758                       Epiblast         1.0000000
## 249 0.095887084                       Epiblast         1.0000000
## 250 0.150285709                       Epiblast         0.9905660
## 251 0.081608840               Nascent mesoderm         0.5384615
## 252 0.008861509                            Gut         0.8888889
## 253 0.895113620                       Epiblast         0.9940828
## 254 0.020131039           Rostral neurectoderm         0.8409091
## 255 0.378010744               Nascent mesoderm         0.4000000
## 256 0.426416504              Visceral endoderm         0.8117647
## 257 0.239480690                   ExE endoderm         1.0000000
## 258 0.083265413                   ExE ectoderm         1.0000000
## 259 0.052889335                 Mixed mesoderm         0.3859649
## 260 0.057451727               Nascent mesoderm         0.9890110
## 261 0.207322806                       Epiblast         0.9722222
## 262 0.022582426                       Epiblast         1.0000000
## 263 0.612079789                   ExE ectoderm         1.0000000
## 264 0.008109758               Nascent mesoderm         0.9846154
## 265 0.262259359                   ExE endoderm         1.0000000
## 266 0.021198857                     Mesenchyme         0.6956522
## 267 0.991979381               Primitive Streak         0.6229508
## 268 0.014502836                   ExE endoderm         1.0000000
## 269 0.547528876               Nascent mesoderm         0.8125000
## 270 0.094099706                   ExE ectoderm         1.0000000
## 271 0.241569277                       Epiblast         1.0000000
## 272 0.020453593               Nascent mesoderm         1.0000000
## 273 0.150285709                       Epiblast         1.0000000
## 274 0.008109758                       Epiblast         0.9367089
## 275 0.034360657                     Mesenchyme         1.0000000
## 276 0.399177190                       Epiblast         0.9662921
## 277 0.030813538                   ExE ectoderm         1.0000000
## 278 0.019152008                       Epiblast         1.0000000
## 279 0.040175261                       Epiblast         1.0000000
## 280 0.083453311                     Mesenchyme         0.4814815
## 281 0.661734478                       Epiblast         0.9215686
## 282 0.102026537            Blood progenitors 1         0.8030303
## 283 0.052244047                       Epiblast         1.0000000
## 284 0.010091012                       Epiblast         0.9636364
## 285 0.112189505               Nascent mesoderm         1.0000000
## 286 0.178531715                   ExE ectoderm         1.0000000
## 287 0.050101678                       Epiblast         0.9736842
## 288 0.010004306                       Epiblast         0.6060606
## 289 0.442021247                       Epiblast         1.0000000
## 290 0.107338242               Primitive Streak         0.9000000
## 291 0.023761232                       Epiblast         0.9782609
## 292 0.741381811                   ExE endoderm         1.0000000
## 293 0.458351935                 Mixed mesoderm         0.9137931
## 294 0.086082663                   ExE endoderm         1.0000000
## 295 0.048503872                     Mesenchyme         0.6037736
## 296 0.142523090                   ExE ectoderm         1.0000000
## 297 0.117424127               Primitive Streak         0.9130435
## 298 0.237393182                   ExE ectoderm         1.0000000
## 299 0.053871099                       Epiblast         1.0000000
## 300 0.222831729                   ExE ectoderm         1.0000000
## 301 0.368500647                   ExE ectoderm         1.0000000
## 302 0.482536188               Primitive Streak         0.5619048
## 303 0.008109758                       Epiblast         1.0000000
## 304 0.945074026              Parietal endoderm         1.0000000
## 305 0.186739541               Primitive Streak         0.6129032
## 306 0.116011341                   ExE ectoderm         1.0000000
## 307 0.238257909               Primitive Streak         0.9633028
## 308 0.047135669                   ExE ectoderm         1.0000000
## 309 0.811528075              Parietal endoderm         1.0000000
## 310 0.916594758                   ExE endoderm         1.0000000
## 311 0.008109758                 Mixed mesoderm         0.4507042
## 312 0.338283008               Primitive Streak         0.9918033
## 313 0.008471652                Caudal epiblast         0.5490196
## 314 0.048944514                       Epiblast         0.7205882
## 315 0.178379881                       Epiblast         0.8319328
## 316 0.052599134               Nascent mesoderm         1.0000000
## 317 0.245948164                 Mixed mesoderm         0.8593750
## 318 0.309266399                   ExE ectoderm         1.0000000
## 319 0.073124971                  Def. endoderm         0.7397260
## 320 0.108426801                       Epiblast         0.8961039
## 321 0.013531434               Primitive Streak         0.7956989
## 322 0.078686047                       Epiblast         1.0000000
## 323 0.060837265               Primitive Streak         0.6279070
## 324 0.141617906                   ExE ectoderm         1.0000000
## 325 0.087843436                   ExE endoderm         1.0000000
## 326 0.387367891                 Mixed mesoderm         0.5573770
## 327 0.100027705                   ExE endoderm         1.0000000
## 328 0.683329204                   ExE endoderm         1.0000000
## 329 0.048410910               Primitive Streak         0.5087719
## 330 0.444027294                       Epiblast         1.0000000
## 331 0.574800748                   ExE ectoderm         1.0000000
## 332 0.715124971                   ExE endoderm         1.0000000
## 333 0.492342590                   ExE ectoderm         1.0000000
## 334 0.044998595                       Epiblast         1.0000000
## 335 0.926035762                   ExE ectoderm         1.0000000
## 336 0.012925756                       Epiblast         0.9620253
## 337 0.569742351                       Epiblast         0.6607143
## 338 0.185094547                            PGC         0.6808511
## 339 0.713850979                   ExE ectoderm         1.0000000
## 340 0.276490283               Nascent mesoderm         0.5542169
## 341 0.008109758                       Epiblast         0.8392857
## 342 0.235418266                       Epiblast         0.9615385
## 343 0.682705871                   ExE endoderm         1.0000000
## 344 0.346599824                   ExE endoderm         1.0000000
## 345 0.144597406                       Epiblast         0.7222222
## 346 0.052244047                   ExE ectoderm         1.0000000
## 347 0.782674806               Primitive Streak         0.7346939
## 348 0.827771073                   ExE ectoderm         1.0000000
## 349 0.113074468                   ExE ectoderm         1.0000000
## 350 0.682705871              Parietal endoderm         1.0000000
## 351 0.361157171                   ExE endoderm         1.0000000
## 352 0.266953064              Visceral endoderm         1.0000000
## 353 0.133981640 Haematoendothelial progenitors         0.8181818
## 354 0.397021669                            PGC         0.7575758
## 355 0.010004306                       Epiblast         1.0000000
## 356 0.008471652           Rostral neurectoderm         0.6101695
## 357 0.416533989                   ExE ectoderm         1.0000000
## 358 0.052244047                   ExE ectoderm         1.0000000
## 359 0.096548412            Blood progenitors 1         0.6133333
## 360 0.872402546                   ExE ectoderm         1.0000000
## 361 0.266310752                   ExE endoderm         1.0000000
## 362 0.081355040                   ExE endoderm         1.0000000
## 363 0.088151418                 Mixed mesoderm         0.8051948
## 364 0.737833303                   ExE ectoderm         1.0000000
## 365 0.247929968      Anterior Primitive Streak         0.9756098
## 366 0.060417298               Nascent mesoderm         0.5294118
## 367 0.335916836               Nascent mesoderm         1.0000000
## 368 0.129677005                   ExE ectoderm         1.0000000
## 369 0.398894415              Visceral endoderm         1.0000000
## 370 0.023640628                       Epiblast         0.7808219
## 371 0.584079854                   ExE endoderm         1.0000000
## 372 0.117797740                       Epiblast         0.7444444
## 373 0.749530968                  Def. endoderm         0.7727273
## 374 0.128662978                       Epiblast         1.0000000
## 375 0.802167701 Haematoendothelial progenitors         0.9295775
## 376 0.081608840                       Epiblast         0.8846154
## 377 0.393012839                   ExE endoderm         1.0000000
## 378 0.023761232      Anterior Primitive Streak         0.6582278
## 379 0.180139330                       Epiblast         1.0000000
## 380 0.036122367           Rostral neurectoderm         0.6951220
## 381 0.232584496                   ExE endoderm         1.0000000
## 382 0.013987599           Rostral neurectoderm         0.6428571
## 383 0.094579452               Nascent mesoderm         0.8490566
## 384 0.142523090                       Epiblast         0.8833333
## 385 0.429185871                   ExE ectoderm         1.0000000
## 386 0.008861509                       Epiblast         0.8857143
## 387 0.926035762                   ExE ectoderm         1.0000000
## 388 0.057412556               Surface ectoderm         0.4722222
## 389 0.026143404                       Epiblast         1.0000000
## 390 0.008109758                Caudal epiblast         0.8372093
## 391 0.094857108                   ExE ectoderm         1.0000000
## 392 0.020131039               Primitive Streak         0.8412698
## 393 0.095116859      Anterior Primitive Streak         0.9523810
## 394 0.306060399                 Mixed mesoderm         0.6315789
## 395 0.033951871                       Epiblast         1.0000000
## 396 0.084268946                   ExE endoderm         1.0000000
## 397 0.198194105            Blood progenitors 2         0.9882353
## 398 0.155000548                   ExE endoderm         1.0000000
## 399 0.244175644                   ExE endoderm         1.0000000
## 400 0.260600244                   ExE endoderm         1.0000000
## 401 0.020453593              Visceral endoderm         1.0000000
## 402 0.108608384            Blood progenitors 1         0.8947368
## 403 0.097252869               Primitive Streak         0.9385965
## 404 0.014180189               Primitive Streak         0.7564103
## 405 0.369725340               Primitive Streak         1.0000000
## 406 0.026143404                       Epiblast         1.0000000
## 407 0.046009368                       Epiblast         0.8000000
## 408 0.113084683                   ExE ectoderm         1.0000000
## 409 0.233092965                   ExE ectoderm         1.0000000
## 410 0.207946949                   ExE ectoderm         1.0000000
## 411 0.083328041                 Mixed mesoderm         0.5400000
## 412 0.108837251                     Mesenchyme         0.9591837
## 413 0.074067989                   ExE ectoderm         1.0000000
## 414 0.015996213                   ExE ectoderm         1.0000000
## 415 0.665676693                   ExE endoderm         1.0000000
## 416 0.236355106                   ExE ectoderm         1.0000000
## 417 0.069324256                Caudal epiblast         0.6379310
## 418 0.019972578                       Epiblast         1.0000000
## 419 0.144597406               Primitive Streak         0.7804878
## 420 0.522194441               Nascent mesoderm         1.0000000
## 421 0.580092628               Nascent mesoderm         1.0000000
## 422 0.732086324                       Epiblast         0.9927536
## 423 0.116011341               Primitive Streak         0.5303030
## 424 0.235418266               Primitive Streak         0.8771930
## 425 0.945074026                   ExE ectoderm         1.0000000
## 426 0.128692947                   ExE ectoderm         1.0000000
## 427 0.008861509               Primitive Streak         0.8536585
## 428 0.018641166               Primitive Streak         0.6274510
## 429 0.044372908                     Mesenchyme         1.0000000
## 430 0.368500647               Nascent mesoderm         0.8636364
## 431 0.028316464               Primitive Streak         1.0000000
## 432 0.058514503               Primitive Streak         1.0000000
## 433 0.494598013                   ExE ectoderm         1.0000000
## 434 0.008109758                 Mixed mesoderm         0.4285714
## 435 0.193268864                   ExE ectoderm         1.0000000
## 436 0.025833459                       Epiblast         1.0000000
## 437 0.052033086                  Def. endoderm         0.5588235
## 438 0.030516916                       Epiblast         0.9795918
## 439 0.029980708                       Epiblast         0.9716981
## 440 0.415450701                   ExE ectoderm         1.0000000
## 441 0.074828761      Anterior Primitive Streak         0.6888889
## 442 0.294481131               Primitive Streak         1.0000000
## 443 0.575572707                   ExE ectoderm         1.0000000
## 444 0.391135923                   ExE ectoderm         1.0000000
## 445 0.637621568               Nascent mesoderm         0.9801980
## 446 0.774970694                   ExE endoderm         1.0000000
## 447 0.226245833                            PGC         0.4117647
## 448 0.087843436                   ExE ectoderm         1.0000000
## 449 0.186739541                   ExE mesoderm         0.4038462
## 450 0.275964283                       Epiblast         0.9739130
## 451 0.979026214                   ExE ectoderm         1.0000000
## 452 0.120588859           Rostral neurectoderm         0.5066667
## 453 0.665676693                   ExE ectoderm         1.0000000
## 454 0.065428362                 Mixed mesoderm         0.5362319
## 455 0.665676693 Haematoendothelial progenitors         1.0000000
## 456 0.031052104                     Mesenchyme         0.9693878
## 457 0.442021247                   ExE endoderm         0.9800000
## 458 0.379492319                   ExE ectoderm         1.0000000
## 459 0.048503872      Anterior Primitive Streak         0.9733333
## 460 0.008109758                       Epiblast         1.0000000
## 461 0.521161189                   ExE endoderm         1.0000000
## 462 0.558751824                   ExE ectoderm         1.0000000
## 463 0.352902435              Visceral endoderm         0.9848485
## 464 0.083953708                     Mesenchyme         0.7260274
## 465 0.151789218                   ExE endoderm         1.0000000
## 466 0.012925756               Primitive Streak         0.9555556
## 467 0.897434209                   ExE endoderm         1.0000000
## 468 0.030013712               Nascent mesoderm         0.9900000
## 469 0.959259089               Primitive Streak         0.7435897
## 470 0.012925756                       Epiblast         0.9600000
## 471 0.262259359 Haematoendothelial progenitors         0.6949153
## 472 0.152760001               Primitive Streak         0.9565217
## 473 0.485227972               Nascent mesoderm         0.9714286
## 474 0.008861509                 Mixed mesoderm         0.5471698
## 475 0.055844084              Visceral endoderm         1.0000000
## 476 0.194479605                  Def. endoderm         0.7093023
## 477 0.188978978               Primitive Streak         0.6875000
## 478 0.269679487                  Def. endoderm         0.9259259
## 479 0.243206108                            PGC         0.3728814
## 480 0.105654408               Primitive Streak         1.0000000
## 481 0.263769787                   ExE endoderm         1.0000000
## 482 0.044998595              Visceral endoderm         1.0000000
## 483 0.261974096                       Epiblast         1.0000000
## 484 0.017443954                   ExE ectoderm         1.0000000
## 485 0.008362158                       Epiblast         0.9259259
## 486 0.020131039                   ExE endoderm         1.0000000
## 487 0.020453593                       Epiblast         0.8804348
## 488 0.068836228               Primitive Streak         0.9538462
## 489 0.085049993                       Epiblast         0.9900000
## 490 0.776483192                       Epiblast         1.0000000
## 491 0.194861578               Primitive Streak         1.0000000
## 492 0.229300269                   ExE ectoderm         1.0000000
## 493 0.008861509                       Epiblast         0.9750000
## 494 0.128692947                       Epiblast         0.8750000
## 495 0.032625503               Primitive Streak         0.7575758
## 496 0.008109758                Caudal epiblast         0.7761194
## 497 0.048503872                       Epiblast         1.0000000
## 498 0.439789169               Nascent mesoderm         0.5555556
## 499 0.897434209                   ExE endoderm         1.0000000
## 500 0.036036078                   ExE endoderm         1.0000000
## 501 0.114265690                   ExE ectoderm         1.0000000
## 502 0.176452439                   ExE ectoderm         1.0000000
## 503 0.635341592               Primitive Streak         0.7317073
## 504 0.078686047                       Epiblast         0.9583333
## 505 0.199902707                   ExE ectoderm         1.0000000
## 506 0.121467516              Visceral endoderm         0.6164384
## 507 0.014880099                       Epiblast         1.0000000
## 508 0.121032286               Primitive Streak         0.8514851
## 509 0.577765276              Visceral endoderm         0.8448276
## 510 0.306060399                       Epiblast         0.6216216
## 511 0.194861578            Blood progenitors 2         0.9863014
## 512 0.020131039           Rostral neurectoderm         0.8571429
## 513 0.134299430              Visceral endoderm         1.0000000
## 514 0.282053776                   ExE endoderm         1.0000000
## 515 0.020453593                 Mixed mesoderm         0.6842105
## 516 0.901260574                 Mixed mesoderm         0.6250000
## 517 0.020453593                      Notochord         0.5600000
## 518 0.811528075                   ExE ectoderm         1.0000000
## 519 0.271166525                   ExE ectoderm         1.0000000
## 520 0.359096212                   ExE ectoderm         1.0000000
## 521 0.899020441                   ExE endoderm         1.0000000
## 522 0.042155319                     Mesenchyme         0.9041096
## 523 0.120333244               Nascent mesoderm         1.0000000
## 524 0.117424127                       Epiblast         0.5465116
## 525 0.040629126                       Epiblast         0.9912281
## 526 0.027755402               Surface ectoderm         0.7457627
## 527 0.102116635               Nascent mesoderm         0.8157895
## 528 0.167114520                       Epiblast         0.9342105
## 529 0.015393029                   ExE ectoderm         1.0000000
## 530 0.048944514                   ExE mesoderm         0.7777778
## 531 0.801846517                   ExE ectoderm         1.0000000
## 532 0.026143404               Primitive Streak         0.9275362
## 533 0.123306675               Nascent mesoderm         1.0000000
## 534 0.970118911                   ExE ectoderm         1.0000000
## 535 0.053017240      Anterior Primitive Streak         0.9264706
## 536 0.161414645                   ExE ectoderm         1.0000000
## 537 0.030013712                   ExE endoderm         1.0000000
## 538 0.387471022                   ExE endoderm         1.0000000
## 539 0.008109758            Pharyngeal mesoderm         0.5066667
## 540 0.118326037               Primitive Streak         0.9726027
## 541 0.081149377                 Mixed mesoderm         0.7012987
## 542 0.010792246                       Epiblast         1.0000000
## 543 0.047135669               Primitive Streak         0.7894737
## 544 0.879137538               Nascent mesoderm         1.0000000
## 545 0.107338242                       Epiblast         0.9696970
## 546 0.208306359                   ExE endoderm         1.0000000
## 547 0.019972578                       Epiblast         0.5423729
## 548 0.052244047              Visceral endoderm         0.9677419
## 549 0.237393182               Primitive Streak         1.0000000
## 550 0.211867322                  Def. endoderm         0.6666667
## 551 0.060701283                     Mesenchyme         1.0000000
## 552 0.709677766                       Epiblast         0.9891304
## 553 0.307224692               Nascent mesoderm         0.8208955
## 554 0.048272520                       Epiblast         0.7047619
## 555 0.683329204                       Epiblast         0.9746835
## 556 0.245628579                   ExE endoderm         1.0000000
## 557 0.008109758                       Epiblast         0.9268293
## 558 0.051759779                     Mesenchyme         1.0000000
## 559 0.081608840      Anterior Primitive Streak         0.9607843
## 560 0.008861509               Primitive Streak         0.7777778
## 561 0.108837251                   ExE ectoderm         1.0000000
## 562 0.975728156                       Epiblast         0.9529412
## 563 0.008861509           Rostral neurectoderm         0.7641509
## 564 0.124754328                       Epiblast         0.9836066
## 565 0.336058307                   ExE ectoderm         1.0000000
## 566 0.177363135               Nascent mesoderm         1.0000000
## 567 0.139942231                       Epiblast         0.8705882
## 568 0.048503872                   ExE ectoderm         1.0000000
## 569 0.080393222                       Epiblast         1.0000000
## 570 0.872789244                   ExE ectoderm         1.0000000
## 571 0.034107391               Surface ectoderm         0.7872340
## 572 0.016586836                Caudal epiblast         0.6851852
## 573 0.044998595                   ExE ectoderm         1.0000000
## 574 0.121467516            Blood progenitors 2         0.9800000
## 575 0.008109758                       Epiblast         0.9649123
## 576 0.014880099           Rostral neurectoderm         0.5500000
## 577 0.029980708                            Gut         0.7692308
## 578 0.136906320               Primitive Streak         0.7096774
## 579 0.022582426               Primitive Streak         1.0000000
## 580 0.715124971              Parietal endoderm         1.0000000
## 581 0.023761232                 Mixed mesoderm         0.7735849
## 582 0.091789106               Primitive Streak         0.8524590
## 583 0.955204818                   ExE ectoderm         1.0000000
## 584 0.010004306               Somitic mesoderm         0.3714286
## 585 0.235418266                   ExE ectoderm         1.0000000
## 586 0.178531715               Nascent mesoderm         1.0000000
## 587 0.078084909                   ExE ectoderm         1.0000000
## 588 0.019972578                       Epiblast         0.9531250
## 589 0.008109758               Nascent mesoderm         0.3191489
## 590 0.019972578                 Mixed mesoderm         0.6031746
## 591 0.023640628                       Epiblast         1.0000000
## 592 0.048138729                       Epiblast         0.9931034
## 593 0.570612616                   ExE ectoderm         1.0000000
## 594 0.110470830               Nascent mesoderm         0.9651163
## 595 0.027847708           Rostral neurectoderm         0.7800000
## 596 0.012589080                       Epiblast         0.7083333
## 597 0.971502649      Anterior Primitive Streak         0.6184211
## 598 0.030813538               Primitive Streak         1.0000000
## 599 0.057793868                   ExE endoderm         1.0000000
## 600 0.129339186                   ExE ectoderm         1.0000000
## 601 0.354998300                       Epiblast         0.9827586
## 602 0.385465362                   ExE endoderm         1.0000000
## 603 0.107630971               Nascent mesoderm         0.8769231
## 604 0.008109758               Primitive Streak         0.7187500
## 605 0.008861509                       Epiblast         1.0000000
## 606 0.022582426                       Epiblast         1.0000000
## 607 0.713850979                   ExE ectoderm         1.0000000
## 608 0.161031282                 Mixed mesoderm         0.5972222
## 609 0.269299999                   ExE endoderm         1.0000000
## 610 0.013987599                       Epiblast         1.0000000
## 611 0.453178277                       Epiblast         0.9918033
## 612 0.016081159                       Epiblast         1.0000000
## 613 0.044024284               Primitive Streak         0.9878049
## 614 0.012925756               Nascent mesoderm         0.9811321
## 615 0.083953708               Primitive Streak         0.9324324
## 616 0.276490283               Primitive Streak         0.9473684
## 617 0.110783538               Surface ectoderm         0.4565217
## 618 0.008109758          Intermediate mesoderm         0.7400000
## 619 0.052244047               Primitive Streak         1.0000000
## 620 0.143890220                   ExE endoderm         1.0000000
## 621 0.023761232           Rostral neurectoderm         0.5681818
## 622 0.247583848               Nascent mesoderm         0.4761905
## 623 0.032625503                     Mesenchyme         1.0000000
## 624 0.386302193               Nascent mesoderm         0.9518072
## 625 0.107630971      Anterior Primitive Streak         0.9491525
## 626 0.019972578                       Epiblast         1.0000000
## 627 0.116909567                   ExE ectoderm         1.0000000
## 628 0.065428362               Primitive Streak         0.6607143
## 629 0.022462770                   ExE endoderm         1.0000000
## 630 0.048884942               Primitive Streak         0.8400000
## 631 0.008362158          Intermediate mesoderm         0.7179487
## 632 0.959259089                   ExE ectoderm         1.0000000
## 633 0.055543022               Nascent mesoderm         0.9782609
## 634 0.745690252                   ExE endoderm         1.0000000
## 635 0.188386937               Primitive Streak         0.9078947
## 636 0.008109758                 Mixed mesoderm         0.5652174
## 637 0.975728156                   ExE ectoderm         1.0000000
## 638 0.141617906               Primitive Streak         1.0000000
## 639 0.078686047                       Epiblast         1.0000000
## 640 0.216729524                       Epiblast         1.0000000
## 641 0.063677150                 Mixed mesoderm         0.6101695
## 642 0.044998595           Rostral neurectoderm         0.6578947
## 643 0.044024284               Primitive Streak         0.8285714
## 644 0.013983553                Caudal epiblast         0.5875000
## 645 0.579745957                   ExE endoderm         1.0000000
## 646 0.163468906                       Epiblast         1.0000000
## 647 0.027697428                       Epiblast         0.9912281
## 648 0.219662034               Nascent mesoderm         0.8125000
## 649 0.296655010                       Epiblast         1.0000000
## 650 0.294481131                   ExE endoderm         1.0000000
## 651 0.843077643                   ExE ectoderm         1.0000000
## 652 0.098931831               Primitive Streak         1.0000000
## 653 0.293451737      Anterior Primitive Streak         0.9139785
## 654 0.066981608                     Mesenchyme         0.8606557
## 655 0.080393222              Visceral endoderm         0.9824561
## 656 0.029829461               Nascent mesoderm         0.9879518
## 657 0.139942231                       Epiblast         1.0000000
## 658 0.023640628                       Epiblast         1.0000000
## 659 0.603925127                  Def. endoderm         0.5681818
## 660 0.095080105                       Epiblast         0.9756098
## 661 0.016081159           Rostral neurectoderm         0.9298246
## 662 0.008861509                 Mixed mesoderm         0.5500000
## 663 0.166664742                   ExE ectoderm         1.0000000
## 664 0.596351413 Haematoendothelial progenitors         1.0000000
## 665 0.008109758            Caudal neurectoderm         0.3518519
## 666 0.253507721               Primitive Streak         1.0000000
## 667 0.013987599                   ExE ectoderm         1.0000000
## 668 0.118131847                       Epiblast         1.0000000
## 669 0.547528876               Nascent mesoderm         0.9666667
## 670 0.180934524                   ExE ectoderm         1.0000000
## 671 0.143890220                 Mixed mesoderm         0.6417910
## 672 0.615720970                            PGC         0.6046512
## 673 0.012925756           Rostral neurectoderm         0.8076923
## 674 0.195478423      Anterior Primitive Streak         0.9090909
## 675 0.321429539               Primitive Streak         0.9905660
## 676 0.362675680                   ExE ectoderm         1.0000000
## 677 0.012925756                       Epiblast         0.9926471
## 678 0.023640628                       Epiblast         0.9464286
## 679 0.011452349                       Epiblast         1.0000000
## 680 0.032272515                       Epiblast         0.8695652
## 681 0.199388405                       Epiblast         0.9531250
## 682 0.024281437           Rostral neurectoderm         0.8035714
## 683 0.022763797                       Epiblast         0.9846154
## 684 0.194024824           Rostral neurectoderm         0.6415094
## 685 0.137447839                       Epiblast         0.8750000
## 686 0.008861509                 Mixed mesoderm         0.5777778
## 687 0.702395138                   ExE ectoderm         1.0000000
## 688 0.632468614                   ExE endoderm         1.0000000
## 689 0.030013712               Primitive Streak         0.9841270
## 690 0.081149377               Primitive Streak         0.9436620
## 691 0.021618319      Anterior Primitive Streak         0.9482759
## 692 0.615345916                   ExE endoderm         1.0000000
## 693 0.039457147               Primitive Streak         0.7647059
## 694 0.019972578                       Epiblast         1.0000000
## 695 0.097516998               Nascent mesoderm         1.0000000
## 696 0.029980708                       Epiblast         0.9607843
## 697 0.201305561               Primitive Streak         1.0000000
## 698 0.215995232                   ExE ectoderm         1.0000000
## 699 0.314208653               Primitive Streak         0.9487179
## 700 0.015817664                       Epiblast         0.9814815
## 701 0.028316464                       Epiblast         1.0000000
## 702 0.032625503              Visceral endoderm         0.9761905
## 703 0.207833667               Primitive Streak         1.0000000
## 704 0.811528075                   ExE ectoderm         1.0000000
## 705 0.054431819                   ExE endoderm         1.0000000
## 706 0.008109758           Rostral neurectoderm         1.0000000
## 707 0.048503872                      Notochord         0.5000000
## 708 0.297042334               Nascent mesoderm         0.6779661
## 709 0.033441509                   ExE ectoderm         1.0000000
## 710 0.019024921                 Mixed mesoderm         0.8750000
## 711 0.661734478                   ExE ectoderm         1.0000000
## 712 0.022646710               Nascent mesoderm         1.0000000
## 713 0.663143354              Parietal endoderm         1.0000000
## 714 0.108666096                       Epiblast         0.8990826
## 715 0.346346483 Haematoendothelial progenitors         0.9814815
## 716 0.033854539                       Epiblast         0.9803922
## 717 0.016081159                       Epiblast         1.0000000
## 718 0.048503872               Nascent mesoderm         1.0000000
## 719 0.012925756                       Epiblast         1.0000000
## 720 0.058802142                   ExE endoderm         1.0000000
## 721 0.682564563               Nascent mesoderm         0.7826087
## 722 0.047135669               Primitive Streak         0.9777778
## 723 0.036122367               Primitive Streak         1.0000000
## 724 0.026143404                   ExE ectoderm         1.0000000
## 725 0.510096242                   ExE endoderm         1.0000000
## 726 0.779925400                   ExE ectoderm         1.0000000
## 727 0.201305561               Nascent mesoderm         1.0000000
## 728 0.084824257                       Epiblast         0.9882353
## 729 0.105654408               Primitive Streak         0.9741379
## 730 0.018641166                       Epiblast         0.9693878
## 731 0.008109758                       Epiblast         1.0000000
## 732 0.201305561               Nascent mesoderm         0.7600000
## 733 0.020453593                       Epiblast         0.7804878
## 734 0.123739197                 Mixed mesoderm         0.5454545
## 735 0.376356929                       Epiblast         0.8235294
## 736 0.020131039               Primitive Streak         0.9620253
## 737 0.008861509           Rostral neurectoderm         0.5869565
## 738 0.163337791                       Epiblast         1.0000000
## 739 0.008109758                       Epiblast         0.8085106
## 740 0.018370520               Somitic mesoderm         0.3095238
## 741 0.060701283              Visceral endoderm         1.0000000
## 742 0.874778629                   ExE ectoderm         1.0000000
## 743 0.035580846                 Mixed mesoderm         0.7945205
## 744 0.301218229               Primitive Streak         0.9684211
## 745 0.008109758           Rostral neurectoderm         0.7391304
## 746 0.008471652           Rostral neurectoderm         0.7500000
## 747 0.761877218                   ExE endoderm         1.0000000
## 748 0.188978978      Anterior Primitive Streak         0.9405941
## 749 0.021703820                       Epiblast         1.0000000
## 750 0.113084683               Primitive Streak         1.0000000
## 751 0.134497293              Visceral endoderm         1.0000000
## 752 0.097028531                       Epiblast         0.6883117
## 753 0.347451382                   ExE ectoderm         1.0000000
## 754 0.011089843                   ExE endoderm         1.0000000
## 755 0.230975933                   ExE ectoderm         1.0000000
## 756 0.008109758                       Epiblast         1.0000000
## 757 0.905336931                   ExE ectoderm         1.0000000
## 758 0.009060419                      Notochord         0.6511628
## 759 0.133983402                       Epiblast         1.0000000
## 760 0.477562234                   ExE endoderm         1.0000000
## 761 0.668348397                       Epiblast         0.9285714
## 762 0.233092965               Primitive Streak         1.0000000
## 763 0.019152008               Nascent mesoderm         0.4000000
## 764 0.393516274               Nascent mesoderm         0.9843750
## 765 0.187630597                   ExE endoderm         1.0000000
## 766 0.050377817                       Epiblast         1.0000000
## 767 0.008471652           Rostral neurectoderm         0.9743590
## 768 0.503376109               Primitive Streak         1.0000000
## 769 0.277767158                   ExE endoderm         1.0000000
## 770 0.116979224                 Mixed mesoderm         0.4897959
## 771 0.247606330                       Epiblast         1.0000000
## 772 0.271166525               Nascent mesoderm         0.8714286
## 773 0.070514618                       Epiblast         0.9855072
## 774 0.127632169            Blood progenitors 2         0.9824561
## 775 0.008861509                 Mixed mesoderm         0.6470588
## 776 0.253507721               Nascent mesoderm         1.0000000
## 777 0.070952068               Nascent mesoderm         0.7794118
## 778 0.057412556                   ExE ectoderm         1.0000000
## 779 0.051105716               Primitive Streak         1.0000000
## 780 0.008109758                       Epiblast         1.0000000
## 781 0.200191846                   ExE ectoderm         1.0000000
## 782 0.008109758                       Epiblast         1.0000000
## 783 0.488761025                   ExE ectoderm         1.0000000
## 784 0.008109758                       Epiblast         1.0000000
## 785 0.073124971                   ExE ectoderm         1.0000000
## 786 0.454878519                 Mixed mesoderm         0.8031496
## 787 0.011452349                Caudal epiblast         0.4358974
## 788 0.072127791                       Epiblast         1.0000000
## 789 0.008471652               Primitive Streak         0.8636364
## 790 0.028316464                       Epiblast         0.9894737
## 791 0.635341592               Primitive Streak         1.0000000
## 792 0.782674806              Visceral endoderm         0.9750000
## 793 0.070078877                       Epiblast         0.9428571
## 794 0.094579452               Primitive Streak         0.8378378
## 795 0.026478582                       Epiblast         1.0000000
## 796 0.804425973                   ExE ectoderm         1.0000000
## 797 0.658601156      Anterior Primitive Streak         0.7444444
## 798 0.188091050 Haematoendothelial progenitors         0.8923077
## 799 0.008109758                       Epiblast         0.9693878
## 800 0.033649483               Primitive Streak         0.8496241
## 801 0.705406535                 Mixed mesoderm         0.8676471
## 802 0.141454628                   ExE endoderm         1.0000000
## 803 0.012925756               Primitive Streak         0.5800000
## 804 0.089576294                   ExE ectoderm         1.0000000
## 805 0.076094054                       Epiblast         0.9270833
## 806 0.945074026               Primitive Streak         1.0000000
## 807 0.129853990               Primitive Streak         0.8852459
## 808 0.028064616                       Epiblast         0.6567164
## 809 0.522446634                   ExE endoderm         1.0000000
## 810 0.237688068                   ExE ectoderm         1.0000000
## 811 0.008861509                       Epiblast         1.0000000
## 812 0.871141177                   ExE endoderm         1.0000000
## 813 0.052244047                 Mixed mesoderm         0.5492958
## 814 0.623783258 Haematoendothelial progenitors         0.6081081
## 815 0.085056562               Nascent mesoderm         0.5641026
## 816 0.282445263                   ExE endoderm         1.0000000
## 817 0.026233512               Primitive Streak         0.5156250
## 818 0.008109758                       Epiblast         1.0000000
## 819 0.052244047                   ExE ectoderm         1.0000000
## 820 0.008109758                       Epiblast         1.0000000
## 821 0.017456375               Primitive Streak         0.4864865
## 822 0.245628579                   ExE ectoderm         1.0000000
## 823 0.029637209                       Epiblast         1.0000000
## 824 0.395757069               Nascent mesoderm         0.9425287
## 825 0.866450851                   ExE endoderm         1.0000000
## 826 0.506112315                   ExE ectoderm         1.0000000
## 827 0.274421892                   ExE ectoderm         1.0000000
## 828 0.022678279           Rostral neurectoderm         0.5645161
## 829 0.097102847                   ExE ectoderm         1.0000000
## 830 0.026143404               Nascent mesoderm         0.8552632
## 831 0.048503872                   ExE ectoderm         1.0000000
## 832 0.059230670                       Epiblast         1.0000000
## 833 0.008861509                 Mixed mesoderm         0.5641026
## 834 0.749155665                       Epiblast         0.9047619
## 835 0.014880099                       Epiblast         0.9318182
## 836 0.406185932               Primitive Streak         1.0000000
## 837 0.083453311               Primitive Streak         1.0000000
## 838 0.020185776               Nascent mesoderm         0.8857143
## 839 0.023761232                       Epiblast         0.9814815
## 840 0.306033739                   ExE ectoderm         1.0000000
## 841 0.520010335      Anterior Primitive Streak         0.4800000
## 842 0.010004306                   ExE endoderm         1.0000000
## 843 0.392219165                   ExE ectoderm         1.0000000
## 844 0.042875425                     Mesenchyme         0.8970588
## 845 0.114832044                   ExE ectoderm         1.0000000
## 846 0.219445538               Primitive Streak         1.0000000
## 847 0.020131039               Nascent mesoderm         1.0000000
## 848 0.085942769                     Mesenchyme         0.7666667
## 849 0.034760723              Visceral endoderm         1.0000000
## 850 0.668348397                   ExE ectoderm         1.0000000
## 851 0.083265413                       Epiblast         1.0000000
## 852 0.008109758            Caudal neurectoderm         0.3750000
## 853 0.052244047                       Epiblast         1.0000000
## 854 0.081850949               Primitive Streak         0.7560976
## 855 0.083265413               Primitive Streak         1.0000000
## 856 0.220089537                   ExE ectoderm         1.0000000
## 857 0.038024801               Primitive Streak         1.0000000
## 858 0.044372908                   ExE endoderm         1.0000000</code></pre>
<p>While neighbourhoods tend to be homogeneous, we can define a threshold for <code>celltype_fraction</code> to exclude neighbourhoods that are a mix of cell types.</p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="fu">ggplot</span>(<span class="no">da_results</span>, <span class="fu">aes</span>(<span class="no">celltype_fraction</span>)) + <span class="fu">geom_histogram</span>(<span class="kw">bins</span><span class="kw">=</span><span class="fl">50</span>)</pre></body></html></div>
<p><img src="milo_gastrulation_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="no">da_results</span>$<span class="no">celltype</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="no">da_results</span>$<span class="no">celltype_fraction</span> <span class="kw">&lt;</span> <span class="fl">0.7</span>, <span class="st">"Mixed"</span>, <span class="no">da_results</span>$<span class="no">celltype</span>)</pre></body></html></div>
<p>Now we can visualize the distribution of DA Fold Changes in different cell types</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="fu"><a href="../reference/plotDAbeeswarm.html">plotDAbeeswarm</a></span>(<span class="no">da_results</span>, <span class="kw">group.by</span> <span class="kw">=</span> <span class="st">"celltype"</span>)</pre></body></html></div>
<p><img src="milo_gastrulation_files/figure-html/unnamed-chunk-20-1.png" width="672"></p>
<p>This is already quite informative: we can see that certain early development cell types, such as epiblast and primitive streak, are enriched in the earliest time stage, while others are enriched later in development, such as ectoderm cells. Interestingly, we also see plenty of DA neighbourhood with a mixed label. This could indicate that transitional states show changes in abundance in time.</p>
</div>
<div id="identifying-signatures-of-da-subpopulations" class="section level1">
<h1 class="hasAnchor">
<a href="#identifying-signatures-of-da-subpopulations" class="anchor"></a>Identifying signatures of DA subpopulations</h1>
<p>Once we have identified neighbourhoods with significant DA, we might be interested in learning more about the gene expression signatures that define these subpopulations. Here the analyst might get creative, depending on the specific characteristics of their dataset and the biological question of interest. In the <code>miloR</code> package, we provide functionality to explore neighbourhood markers through the function <code>findNhoodMarkers</code>. This groups significantly DA neighbourhoods that show concordant Fold-Change and performs a test for differential expression between cells in those neighbourhoods.</p>
<p>In practice, it might be convenient to subset a selected number of neighbourhoods of interest for gene-level downstream analysis. For the sake of demonstration, here we focus on identifying signatures of DA subpopulations in the endoderm development lineage.</p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">logcounts</a></span>(<span class="no">embryo_milo</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log1p</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span>(<span class="no">embryo_milo</span>))
<span class="no">dge_smp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/findNhoodMarkers.html">findNhoodMarkers</a></span>(<span class="no">embryo_milo</span>, <span class="no">da_results</span>,
                                     <span class="kw">assay</span> <span class="kw">=</span> <span class="st">"counts"</span>, <span class="kw">gene.offset</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">da.fdr</span> <span class="kw">=</span> <span class="fl">0.1</span>,
                                     <span class="kw">aggregate.samples</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">sample_col</span> <span class="kw">=</span> <span class="st">"sample"</span>,
                                     <span class="kw">subset.nhoods</span> <span class="kw">=</span> <span class="no">da_results</span>$<span class="no">celltype</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Anterior Primitive Streak"</span>, <span class="st">"Def. endoderm"</span>, <span class="st">"Gut"</span>, <span class="st">"Visceral endoderm"</span>)
                                     )

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">dge_smp</span>)</pre></body></html></div>
<pre><code>##                      logFC_1 adj.P.Val_1   logFC_2 adj.P.Val_2  logFC_3
## ENSMUSG00000051951  0.000000  1.00000000  0.000000    1.000000 0.000000
## ENSMUSG00000089699  0.000000  1.00000000  0.000000    1.000000 0.000000
## ENSMUSG00000102343  0.000000  1.00000000  0.000000    1.000000 0.000000
## ENSMUSG00000025900  0.000000  1.00000000  0.000000    1.000000 0.000000
## ENSMUSG00000025902 -3.431228  0.06766283 -1.674657    0.382225 2.362171
## ENSMUSG00000104328  0.000000  1.00000000  0.000000    1.000000 0.000000
##                     adj.P.Val_3   logFC_5 adj.P.Val_5             GeneID
## ENSMUSG00000051951 1.0000000000 0.0000000           1 ENSMUSG00000051951
## ENSMUSG00000089699 1.0000000000 0.0000000           1 ENSMUSG00000089699
## ENSMUSG00000102343 1.0000000000 0.0000000           1 ENSMUSG00000102343
## ENSMUSG00000025900 1.0000000000 0.0000000           1 ENSMUSG00000025900
## ENSMUSG00000025902 0.0009905911 0.1180312           1 ENSMUSG00000025902
## ENSMUSG00000104328 1.0000000000 0.0000000           1 ENSMUSG00000104328</code></pre>
<p>This identifies n marker genes at FDR 10% that distinguish two main groups within the epiblast neighbourhoods, one significantly depleted in the early stage and one significantly enriched. We can visualize expression of the detected marker genes using the function <code>plotNhoodExpressionDA</code>. This shows the average expression in each neighbourhood, ranked by log-Fold Change in the DA test. Note that the gene x nhood expression matrix can be pre-computed and stored using the <code>calcNhoodExpression</code> function, to avoid repeating the computation every time you need to plot.</p>
<p>In this case we mainly identified negative markers of the epiblast neighbourhoods enriched with age.</p>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="no">markers</span> <span class="kw">&lt;-</span> <span class="no">dge_smp</span>[<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">dge_smp</span>$<span class="no">adj.P.Val_1</span> <span class="kw">&lt;</span> <span class="fl">0.1</span> ), <span class="st">"GeneID"</span>]
<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">logcounts</a></span>(<span class="no">embryo_milo</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log1p</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span>(<span class="no">embryo_milo</span>))
<span class="no">embryo_milo</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/calcNhoodExpression.html">calcNhoodExpression</a></span>(<span class="no">embryo_milo</span>, <span class="kw">subset.row</span><span class="kw">=</span><span class="no">markers</span>)

<span class="fu"><a href="../reference/plotNhoodExpressionDA.html">plotNhoodExpressionDA</a></span>(<span class="no">embryo_milo</span>, <span class="no">da_results</span>, <span class="kw">features</span> <span class="kw">=</span> <span class="no">markers</span>,
                      <span class="kw">subset.nhoods</span> <span class="kw">=</span> <span class="no">da_results</span>$<span class="no">celltype</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Anterior Primitive Streak"</span>, <span class="st">"Def. endoderm"</span>, <span class="st">"Gut"</span>, <span class="st">"Visceral endoderm"</span>),
                      <span class="kw">assay</span><span class="kw">=</span><span class="st">"logcounts"</span>,
                      <span class="kw">scale_to_1</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">cluster_features</span> <span class="kw">=</span> <span class="fl">TRUE</span>
                      )</pre></body></html></div>
<p><img src="milo_gastrulation_files/figure-html/unnamed-chunk-22-1.png" width="864"></p>
<details><p><summary><strong>Session Info</strong></summary></p>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</pre></body></html></div>
<pre><code>## R version 3.6.3 (2020-02-29)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.3 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
## LAPACK: /usr/lib/x86_64-linux-gnu/openblas/liblapack.so.3
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] MouseGastrulationData_1.0.0 patchwork_1.0.1            
##  [3] dplyr_1.0.2                 scater_1.14.6              
##  [5] ggplot2_3.3.2               SingleCellExperiment_1.8.0 
##  [7] SummarizedExperiment_1.16.1 DelayedArray_0.12.3        
##  [9] BiocParallel_1.20.1         matrixStats_0.56.0         
## [11] Biobase_2.46.0              GenomicRanges_1.38.0       
## [13] GenomeInfoDb_1.22.1         IRanges_2.20.2             
## [15] S4Vectors_0.24.4            BiocGenerics_0.32.0        
## [17] miloR_0.1.0                 edgeR_3.28.1               
## [19] limma_3.42.2                BiocStyle_2.14.4           
## 
## loaded via a namespace (and not attached):
##   [1] ggbeeswarm_0.6.0              colorspace_1.4-1             
##   [3] ellipsis_0.3.1                rprojroot_1.3-2              
##   [5] XVector_0.26.0                BiocNeighbors_1.4.2          
##   [7] fs_1.5.0                      rstudioapi_0.11              
##   [9] farver_2.0.3                  graphlayouts_0.7.0           
##  [11] ggrepel_0.8.2                 bit64_4.0.2                  
##  [13] AnnotationDbi_1.48.0          interactiveDisplayBase_1.24.0
##  [15] splines_3.6.3                 codetools_0.2-16             
##  [17] knitr_1.29                    polyclip_1.10-0              
##  [19] dbplyr_1.4.4                  uwot_0.1.8                   
##  [21] ggforce_0.3.2                 shiny_1.5.0                  
##  [23] BiocManager_1.30.10           compiler_3.6.3               
##  [25] httr_1.4.2                    backports_1.1.9              
##  [27] fastmap_1.0.1                 assertthat_0.2.1             
##  [29] Matrix_1.2-18                 later_1.1.0.1                
##  [31] tweenr_1.0.1                  BiocSingular_1.2.2           
##  [33] htmltools_0.5.0               tools_3.6.3                  
##  [35] rsvd_1.0.3                    igraph_1.2.5                 
##  [37] gtable_0.3.0                  glue_1.4.1                   
##  [39] GenomeInfoDbData_1.2.2        rappdirs_0.3.1               
##  [41] Rcpp_1.0.5                    pkgdown_1.5.1                
##  [43] vctrs_0.3.2                   ExperimentHub_1.12.0         
##  [45] DelayedMatrixStats_1.8.0      ggraph_2.0.3                 
##  [47] xfun_0.16                     stringr_1.4.0                
##  [49] mime_0.9                      lifecycle_0.2.0              
##  [51] irlba_2.3.3                   gtools_3.8.2                 
##  [53] statmod_1.4.34                AnnotationHub_2.18.0         
##  [55] zlibbioc_1.32.0               MASS_7.3-51.5                
##  [57] scales_1.1.1                  tidygraph_1.2.0              
##  [59] promises_1.1.1                yaml_2.2.1                   
##  [61] curl_4.3                      memoise_1.1.0                
##  [63] gridExtra_2.3                 stringi_1.4.6                
##  [65] RSQLite_2.2.0                 BiocVersion_3.10.1           
##  [67] desc_1.2.0                    rlang_0.4.7                  
##  [69] pkgconfig_2.0.3               bitops_1.0-6                 
##  [71] evaluate_0.14                 lattice_0.20-40              
##  [73] purrr_0.3.4                   labeling_0.3                 
##  [75] cowplot_1.0.0                 bit_4.0.4                    
##  [77] tidyselect_1.1.0              RcppAnnoy_0.0.16             
##  [79] magrittr_1.5                  bookdown_0.20                
##  [81] R6_2.4.1                      generics_0.0.2               
##  [83] DBI_1.1.0                     pillar_1.4.6                 
##  [85] withr_2.2.0                   RCurl_1.98-1.2               
##  [87] tibble_3.0.3                  crayon_1.3.4                 
##  [89] BiocFileCache_1.10.2          rmarkdown_2.3                
##  [91] viridis_0.5.1                 locfit_1.5-9.4               
##  [93] grid_3.6.3                    blob_1.2.1                   
##  [95] digest_0.6.25                 xtable_1.8-4                 
##  [97] tidyr_1.1.1                   httpuv_1.5.4                 
##  [99] munsell_0.5.0                 beeswarm_0.2.3               
## [101] viridisLite_0.3.0             vipor_0.4.5</code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Mike Morgan, Emma Dann.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
